// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/extensions/browser/webRecommendations.js

import { __decorate, __param } from '../../../../../../../external/tslib/tslib.es6.js';
import { ExtensionRecommendations } from './extensionRecommendations.js';
import { IProductService } from '../../../../platform/product/common/productService.service.js';
import { ExtensionRecommendationReason } from '../../../services/extensionRecommendations/common/extensionRecommendations.js';
import { localize } from '../../../../nls.js';
import { IExtensionManagementServerService } from '../../../services/extensionManagement/common/extensionManagement.service.js';

let WebRecommendations = class WebRecommendations extends ExtensionRecommendations {
    get recommendations() { return this._recommendations; }
    constructor(productService, extensionManagementServerService) {
        super();
        this.productService = productService;
        this.extensionManagementServerService = extensionManagementServerService;
        this._recommendations = [];
    }
    async doActivate() {
        const isOnlyWeb = this.extensionManagementServerService.webExtensionManagementServer && !this.extensionManagementServerService.localExtensionManagementServer && !this.extensionManagementServerService.remoteExtensionManagementServer;
        if (isOnlyWeb && Array.isArray(this.productService.webExtensionTips)) {
            this._recommendations = ( this.productService.webExtensionTips.map((extensionId) => ({
                extension: extensionId.toLowerCase(),
                reason: {
                    reasonId: ExtensionRecommendationReason.Application,
                    reasonText: ( localize(
                        7863,
                        "This extension is recommended for {0} for the Web",
                        this.productService.nameLong
                    ))
                }
            })));
        }
    }
};
WebRecommendations = ( __decorate([
    ( __param(0, IProductService)),
    ( __param(1, IExtensionManagementServerService))
], WebRecommendations));

export { WebRecommendations };
