// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/tasks/common/taskService.js

import { localize } from '../../../../nls.js';
import { RawContextKey, ContextKeyExpr } from '../../../../platform/contextkey/common/contextkey.js';

const CustomExecutionSupportedContext = ( new RawContextKey('customExecutionSupported', false, ( localize(
  11475,
  "Whether CustomExecution tasks are supported. Consider using in the when clause of a \'taskDefinition\' contribution."
))));
const ShellExecutionSupportedContext = ( new RawContextKey('shellExecutionSupported', false, ( localize(
  11476,
  "Whether ShellExecution tasks are supported. Consider using in the when clause of a \'taskDefinition\' contribution."
))));
const TaskCommandsRegistered = ( new RawContextKey('taskCommandsRegistered', false, ( localize(11477, "Whether the task commands have been registered yet"))));
const ProcessExecutionSupportedContext = ( new RawContextKey('processExecutionSupported', false, ( localize(
  11478,
  "Whether ProcessExecution tasks are supported. Consider using in the when clause of a \'taskDefinition\' contribution."
))));
const ServerlessWebContext = ( new RawContextKey('serverlessWebContext', false, ( localize(11479, "True when in the web with no remote authority."))));
const TasksAvailableContext = ( new RawContextKey('tasksAvailable', false, ( localize(11480, "Whether any tasks are available in the workspace."))));
const TaskExecutionSupportedContext = ( ContextKeyExpr.or(( ContextKeyExpr.and(ShellExecutionSupportedContext, ProcessExecutionSupportedContext)), CustomExecutionSupportedContext));

export { CustomExecutionSupportedContext, ProcessExecutionSupportedContext, ServerlessWebContext, ShellExecutionSupportedContext, TaskCommandsRegistered, TaskExecutionSupportedContext, TasksAvailableContext };
