// monaco-languageclient/lib/debugger/index.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) 2024 TypeFox and others.
 * Licensed under the MIT License. See LICENSE in the package root for license information.
 * ------------------------------------------------------------------------------------------ */
import { RegisteredMemoryFile } from '@codingame/monaco-vscode-files-service-override';
import * as vscode from 'vscode';
import { Uri } from 'vscode';
export const provideDebuggerExtensionConfig = (config) => {
    const filesOrContents = new Map();
    filesOrContents.set('./extension.js', '// nothing');
    return {
        config: {
            name: config.extensionName,
            publisher: config.publisher,
            version: config.version,
            engines: {
                vscode: '*'
            },
            // A browser field is mandatory for the extension to be flagged as `web`
            browser: 'extension.js',
            contributes: {
                debuggers: [
                    {
                        type: config.languageId,
                        label: 'Test',
                        languages: [config.languageId]
                    }
                ],
                breakpoints: [
                    {
                        language: config.languageId
                    }
                ]
            },
            activationEvents: [
                'onDebug'
            ]
        },
        filesOrContents
    };
};
export const configureDebugging = async (api, config, logger) => {
    class WebsocketDebugAdapter {
        websocket;
        constructor(websocket) {
            this.websocket = websocket;
            this.websocket.onmessage = (message) => {
                this._onDidSendMessage.fire(JSON.parse(message.data));
            };
        }
        _onDidSendMessage = new api.EventEmitter();
        onDidSendMessage = this._onDidSendMessage.event;
        handleMessage(message) {
            // path with on Windows (Chrome/Firefox) arrive here with \\ and not like expected with /
            // Chrome on Ubuntu behaves as expected
            const msg = JSON.stringify(message).replaceAll('\\\\', '/');
            this.websocket.send(msg);
        }
        dispose() {
            this.websocket.close();
        }
    }
    api.debug.registerDebugAdapterDescriptorFactory(config.languageId, {
        async createDebugAdapterDescriptor() {
            const websocket = new WebSocket(`${config.protocol}://${config.hostname}:${config.port}`);
            await new Promise((resolve, reject) => {
                websocket.onopen = resolve;
                websocket.onerror = () => reject(new Error(`Unable to connect to debugger server. Run "${config.helpContainerCmd}"`));
            });
            const adapter = new WebsocketDebugAdapter(websocket);
            const initMessage = {
                id: 'init',
                files: {},
                // the default file is the one that will be used by the debugger
                defaultFile: config.defaultFile,
                debuggerExecCall: config.debuggerExecCall
            };
            for (const [name, fileDef] of config.files.entries()) {
                logger?.info(`Found: ${name} Sending file: ${fileDef.path}`);
                initMessage.files[name] = {
                    path: fileDef.path,
                    code: fileDef.code,
                    uri: fileDef.uri
                };
            }
            websocket.send(JSON.stringify(initMessage));
            // oxlint-disable-next-line @typescript-eslint/no-explicit-any
            adapter.onDidSendMessage((message) => {
                if (message.type === 'event' && message.event === 'output') {
                    logger?.info('OUTPUT', message.body.output);
                }
            });
            return new api.DebugAdapterInlineImplementation(adapter);
        }
    });
};
export const createDebugLaunchConfigFile = (workspacePath, type) => {
    return new RegisteredMemoryFile(Uri.file(`${workspacePath}/.vscode/launch.json`), JSON.stringify({
        version: '0.2.0',
        configurations: [
            {
                name: 'Debugger: Lauch',
                type,
                request: 'attach',
            }
        ]
    }, null, 2));
};
