// @codingame/monaco-vscode-keybindings-service-override/index.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { SyncDescriptor } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/descriptors';
import { WorkbenchKeybindingService } from './vscode/src/vs/workbench/services/keybinding/browser/keybindingService.js';
import { IKeybindingService } from '@codingame/monaco-vscode-api/vscode/vs/platform/keybinding/common/keybinding.service';
import { VSBuffer } from '@codingame/monaco-vscode-api/vscode/vs/base/common/buffer';
import { IUserDataProfilesService } from '@codingame/monaco-vscode-api/vscode/vs/platform/userDataProfile/common/userDataProfile.service';
import { IKeyboardLayoutService } from '@codingame/monaco-vscode-api/vscode/vs/platform/keyboardLayout/common/keyboardLayout.service';
import { BrowserKeyboardLayoutService } from './vscode/src/vs/workbench/services/keybinding/browser/keyboardLayoutService.js';
import { IFileService } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/common/files.service';
import { ICommandService } from '@codingame/monaco-vscode-api/vscode/vs/platform/commands/common/commands.service';
import { CommandService } from './vscode/src/vs/workbench/services/commands/common/commandService.js';
import { DisposableStore, toDisposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import '@codingame/monaco-vscode-api/vscode/vs/platform/keybinding/common/keybindingResolver';
import { IContextKeyService } from '@codingame/monaco-vscode-api/vscode/vs/platform/contextkey/common/contextkey.service';
import { IUriIdentityService } from '@codingame/monaco-vscode-api/vscode/vs/platform/uriIdentity/common/uriIdentity.service';
import { ITelemetryService } from '@codingame/monaco-vscode-api/vscode/vs/platform/telemetry/common/telemetry.service';
import { INotificationService } from '@codingame/monaco-vscode-api/vscode/vs/platform/notification/common/notification.service';
import { IUserDataProfileService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/userDataProfile/common/userDataProfile.service';
import { IHostService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/host/browser/host.service';
import { IExtensionService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/extensions/common/extensions.service';
import { ILogService } from '@codingame/monaco-vscode-api/vscode/vs/platform/log/common/log.service';
import { Schemas } from '@codingame/monaco-vscode-api/vscode/vs/base/common/network';
import { URI } from '@codingame/monaco-vscode-api/vscode/vs/base/common/uri';
import getServiceOverride$1, { initFile } from '@codingame/monaco-vscode-files-service-override';
import '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/workbench.contribution';
import './vscode/src/vs/workbench/contrib/keybindings/browser/keybindings.contribution.js';
import './vscode/src/vs/workbench/contrib/preferences/browser/keybindingsEditorContribution.js';
import './vscode/src/vs/workbench/contrib/commands/common/commands.contribution.js';
import { getService } from '@codingame/monaco-vscode-api/services';

const defaultUserKeybindindsFile = URI.from({
    scheme: Schemas.vscodeUserData,
    path: '/User/keybindings.json'
});
async function initUserKeybindings(configurationJson, options, file = defaultUserKeybindindsFile) {
    await initFile(file, configurationJson, options);
}
async function updateUserKeybindings(keybindingsJson) {
    const userDataProfilesService = await getService(IUserDataProfilesService);
    const fileService = await getService(IFileService);
    await fileService.writeFile(userDataProfilesService.defaultProfile.keybindingsResource, VSBuffer.fromString(keybindingsJson));
}
let DynamicWorkbenchKeybindingService = class DynamicWorkbenchKeybindingService extends WorkbenchKeybindingService {
    constructor(shouldUseGlobalKeybindings, _lockKeyCodesOverride, contextKeyService, commandService, telemetryService, notificationService, userDataProfileService, hostService, extensionService, fileService, uriIdentityService, logService, keyboardLayoutService) {
        super(contextKeyService, commandService, telemetryService, notificationService, userDataProfileService, hostService, extensionService, fileService, uriIdentityService, logService, keyboardLayoutService);
        this.shouldUseGlobalKeybindings = shouldUseGlobalKeybindings;
        this._lockKeyCodesOverride = _lockKeyCodesOverride;
        this.keybindingProviders = [];
        this.authorizedContainers = [];
    }
    registerKeybindingProvider(provider) {
        this.keybindingProviders.push(provider);
        this.updateResolver();
        const store = new DisposableStore();
        store.add(provider.onDidChangeKeybindings(() => {
            this.updateResolver();
        }));
        store.add(toDisposable(() => {
            const idx = this.keybindingProviders.indexOf(provider);
            if (idx >= 0) {
                this.keybindingProviders.splice(idx, 1);
                this.updateResolver();
            }
        }));
        return store;
    }
    _getResolver() {
        return super._getResolver();
    }
    registerContainer(container) {
        const disposableStore = new DisposableStore();
        disposableStore.add(super.registerContainer(container));
        this.authorizedContainers.push(container);
        disposableStore.add(toDisposable(() => {
            this.authorizedContainers = this.authorizedContainers.filter((c) => c != container);
        }));
        return disposableStore;
    }
    _dispatch(e, target) {
        if (!this.shouldUseGlobalKeybindings() &&
            this.authorizedContainers.every((container) => !container.contains(e.target))) {
            return false;
        }
        return super._dispatch(e, target);
    }
    getUserKeybindingItems() {
        return [
            ...super.getUserKeybindingItems(),
            ...this.keybindingProviders.flatMap((provider) => provider.provideKeybindings())
        ];
    }
    lockKeyCodes(keyCodes) {
        this._lockKeyCodesOverride(keyCodes, (keyCodes) => super.lockKeyCodes(keyCodes));
    }
};
DynamicWorkbenchKeybindingService = __decorate([
    __param(2, IContextKeyService),
    __param(3, ICommandService),
    __param(4, ITelemetryService),
    __param(5, INotificationService),
    __param(6, IUserDataProfileService),
    __param(7, IHostService),
    __param(8, IExtensionService),
    __param(9, IFileService),
    __param(10, IUriIdentityService),
    __param(11, ILogService),
    __param(12, IKeyboardLayoutService)
], DynamicWorkbenchKeybindingService);
function getServiceOverride({ shouldUseGlobalKeybindings = () => false, lockKeyCodesOverride = (keyCodes, next) => {
    next(keyCodes);
} } = {}) {
    return {
        ...getServiceOverride$1(),
        [IKeybindingService.toString()]: new SyncDescriptor(DynamicWorkbenchKeybindingService, [shouldUseGlobalKeybindings, lockKeyCodesOverride], false),
        [IKeyboardLayoutService.toString()]: new SyncDescriptor(BrowserKeyboardLayoutService, [], true),
        [ICommandService.toString()]: new SyncDescriptor(CommandService, [], true)
    };
}

export { getServiceOverride as default, defaultUserKeybindindsFile, initUserKeybindings, updateUserKeybindings };
