// @codingame/monaco-vscode-extensions-service-override/vscode/src/vs/workbench/api/browser/mainThreadChatContext.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { Disposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import { extHostNamedCustomer } from '../../services/extensions/common/extHostCustomers.js';
import { ExtHostContext, MainContext } from '@codingame/monaco-vscode-api/vscode/vs/workbench/api/common/extHost.protocol';
import { IChatContextService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/chat/browser/contextContrib/chatContextService.service';

let MainThreadChatContext = class MainThreadChatContext extends Disposable {
    constructor(extHostContext, _chatContextService) {
        super();
        this._chatContextService = _chatContextService;
        this._providers = ( new Map());
        this._proxy = ( extHostContext.getProxy(ExtHostContext.ExtHostChatContext));
    }
    $registerChatContextProvider(handle, id, selector, _options, support) {
        this._providers.set(handle, { selector, support, id });
        this._chatContextService.registerChatContextProvider(id, selector, {
            provideChatContext: (token) => {
                return this._proxy.$provideChatContext(handle, token);
            },
            resolveChatContext: support.supportsResolve ? (context, token) => {
                return this._proxy.$resolveChatContext(handle, context, token);
            } : undefined,
            provideChatContextForResource: support.supportsResource ? (resource, withValue, token) => {
                return this._proxy.$provideChatContextForResource(handle, { resource, withValue }, token);
            } : undefined
        });
    }
    $unregisterChatContextProvider(handle) {
        const provider = this._providers.get(handle);
        if (!provider) {
            return;
        }
        this._chatContextService.unregisterChatContextProvider(provider.id);
        this._providers.delete(handle);
    }
    $updateWorkspaceContextItems(handle, items) {
        const provider = this._providers.get(handle);
        if (!provider) {
            return;
        }
        this._chatContextService.updateWorkspaceContextItems(provider.id, items);
    }
};
MainThreadChatContext = __decorate([
    extHostNamedCustomer(MainContext.MainThreadChatContext),
    ( __param(1, IChatContextService))
], MainThreadChatContext);

export { MainThreadChatContext };
