// @codingame/monaco-vscode-api/services.js

import { __decorate, __param } from './external/tslib/tslib.es6.js';
import './missing-services.js';
import './vscode/src/vs/workbench/api/common/jsonValidationExtensionPoint.js';
import './vscode/src/vs/workbench/services/themes/common/colorExtensionPoint.js';
import './vscode/src/vs/workbench/services/themes/common/iconExtensionPoint.js';
import './vscode/src/vs/platform/actions/common/actions.contribution.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/workbenchReferenceSearch.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/menuPreventer.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/diffEditorHelper.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/largeFileOptimizations.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/saveParticipants.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/toggleMinimap.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/toggleMultiCursorModifier.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/toggleRenderControlCharacter.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/toggleWordWrap.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/toggleRenderWhitespace.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/editorLineNumberMenu.js';
import './vscode/src/vs/workbench/contrib/format/browser/formatActionsMultiple.js';
import './vscode/src/vs/workbench/contrib/format/browser/formatActionsNone.js';
import './vscode/src/vs/workbench/contrib/format/browser/formatModified.js';
import './vscode/src/vs/workbench/contrib/folding/browser/folding.contribution.js';
import './vscode/src/vs/workbench/contrib/inlayHints/browser/inlayHintsAccessibilty.js';
import './vscode/src/vs/workbench/contrib/codeActions/browser/codeActions.contribution.js';
import './vscode/src/vs/workbench/contrib/list/browser/list.contribution.js';
import './vscode/src/vs/workbench/contrib/codeEditor/browser/editorFeatures.js';
import './vscode/src/vs/workbench/browser/actions/developerActions.js';
import './vscode/src/vs/workbench/contrib/dropOrPasteInto/browser/dropOrPasteInto.contribution.js';
import './vscode/src/vs/editor/browser/coreCommands.js';
import './vscode/src/vs/editor/browser/widget/codeEditor/codeEditorWidget.js';
import './vscode/src/vs/editor/browser/widget/diffEditor/diffEditor.contribution.js';
import './vscode/src/vs/editor/contrib/anchorSelect/browser/anchorSelect.js';
import './vscode/src/vs/editor/contrib/bracketMatching/browser/bracketMatching.js';
import './vscode/src/vs/editor/contrib/caretOperations/browser/caretOperations.js';
import './vscode/src/vs/editor/contrib/caretOperations/browser/transpose.js';
import './vscode/src/vs/editor/contrib/clipboard/browser/clipboard.js';
import './vscode/src/vs/editor/contrib/codeAction/browser/codeActionContributions.js';
import './vscode/src/vs/editor/contrib/codelens/browser/codelensController.js';
import './vscode/src/vs/editor/contrib/colorPicker/browser/colorPickerContribution.js';
import './vscode/src/vs/editor/contrib/comment/browser/comment.js';
import './vscode/src/vs/editor/contrib/contextmenu/browser/contextmenu.js';
import './vscode/src/vs/editor/contrib/cursorUndo/browser/cursorUndo.js';
import './vscode/src/vs/editor/contrib/dnd/browser/dnd.js';
import './vscode/src/vs/editor/contrib/dropOrPasteInto/browser/copyPasteContribution.js';
import './vscode/src/vs/editor/contrib/dropOrPasteInto/browser/dropIntoEditorContribution.js';
import './vscode/src/vs/editor/contrib/find/browser/findController.js';
import './vscode/src/vs/editor/contrib/folding/browser/folding.js';
import './vscode/src/vs/editor/contrib/fontZoom/browser/fontZoom.js';
import './vscode/src/vs/editor/contrib/format/browser/formatActions.js';
import './vscode/src/vs/editor/contrib/documentSymbols/browser/documentSymbols.js';
import './vscode/src/vs/editor/contrib/inlineCompletions/browser/inlineCompletions.contribution.js';
import './vscode/src/vs/editor/contrib/inlineProgress/browser/inlineProgress.js';
import './vscode/src/vs/editor/contrib/gotoSymbol/browser/goToCommands.js';
import './vscode/src/vs/editor/contrib/gotoSymbol/browser/link/goToDefinitionAtPosition.js';
import './vscode/src/vs/editor/contrib/gotoError/browser/gotoError.js';
import './vscode/src/vs/editor/contrib/gpu/browser/gpuActions.js';
import './vscode/src/vs/editor/contrib/hover/browser/hoverContribution.js';
import './vscode/src/vs/editor/contrib/indentation/browser/indentation.js';
import './vscode/src/vs/editor/contrib/inlayHints/browser/inlayHintsContribution.js';
import './vscode/src/vs/editor/contrib/inPlaceReplace/browser/inPlaceReplace.js';
import './vscode/src/vs/editor/contrib/insertFinalNewLine/browser/insertFinalNewLine.js';
import './vscode/src/vs/editor/contrib/lineSelection/browser/lineSelection.js';
import './vscode/src/vs/editor/contrib/linesOperations/browser/linesOperations.js';
import './vscode/src/vs/editor/contrib/linkedEditing/browser/linkedEditing.js';
import './vscode/src/vs/editor/contrib/links/browser/links.js';
import './vscode/src/vs/editor/contrib/longLinesHelper/browser/longLinesHelper.js';
import './vscode/src/vs/editor/contrib/middleScroll/browser/middleScroll.contribution.js';
import './vscode/src/vs/editor/contrib/multicursor/browser/multicursor.js';
import './vscode/src/vs/editor/contrib/parameterHints/browser/parameterHints.js';
import './vscode/src/vs/editor/contrib/placeholderText/browser/placeholderText.contribution.js';
import './vscode/src/vs/editor/contrib/rename/browser/rename.js';
import './vscode/src/vs/editor/contrib/sectionHeaders/browser/sectionHeaders.js';
import './vscode/src/vs/editor/contrib/semanticTokens/browser/documentSemanticTokens.js';
import './vscode/src/vs/editor/contrib/semanticTokens/browser/viewportSemanticTokens.js';
import './vscode/src/vs/editor/contrib/smartSelect/browser/smartSelect.js';
import './vscode/src/vs/editor/contrib/snippet/browser/snippetController2.js';
import './vscode/src/vs/editor/contrib/stickyScroll/browser/stickyScrollContribution.js';
import './vscode/src/vs/editor/contrib/suggest/browser/suggestController.js';
import './vscode/src/vs/editor/contrib/suggest/browser/suggestInlineCompletions.js';
import './vscode/src/vs/editor/contrib/tokenization/browser/tokenization.js';
import './vscode/src/vs/editor/contrib/toggleTabFocusMode/browser/toggleTabFocusMode.js';
import './vscode/src/vs/editor/contrib/unicodeHighlighter/browser/unicodeHighlighter.js';
import './vscode/src/vs/editor/contrib/unusualLineTerminators/browser/unusualLineTerminators.js';
import './vscode/src/vs/editor/contrib/wordHighlighter/browser/wordHighlighter.js';
import './vscode/src/vs/editor/contrib/wordOperations/browser/wordOperations.js';
import './vscode/src/vs/editor/contrib/wordPartOperations/browser/wordPartOperations.js';
import './vscode/src/vs/editor/contrib/readOnlyMessage/browser/contribution.js';
import './vscode/src/vs/editor/contrib/diffEditorBreadcrumbs/browser/contribution.js';
import './vscode/src/vs/editor/contrib/floatingMenu/browser/floatingMenu.contribution.js';
export { InstantiationType, registerSingleton } from './vscode/src/vs/platform/instantiation/common/extensions.js';
export { IEditorWorkerService } from './vscode/src/vs/editor/common/services/editorWorker.service.js';
import './vscode/src/vs/editor/browser/services/editorWorkerService.js';
import './vscode/src/vs/editor/common/standaloneStrings.js';
import './vscode/src/vs/base/browser/ui/codicons/codiconStyles.js';
import './vscode/src/vs/editor/standalone/browser/iPadShowKeyboard/iPadShowKeyboard.js';
export { default as Severity } from './vscode/src/vs/base/common/severity.js';
export { StorageScope, StorageTarget } from './vscode/src/vs/platform/storage/common/storage.js';
import { StandaloneServices } from './vscode/src/vs/editor/standalone/browser/standaloneServices.js';
import { IInstantiationService } from './vscode/src/vs/platform/instantiation/common/instantiation.js';
import { DisposableStore } from './vscode/src/vs/base/common/lifecycle.js';
import { registerWorkbenchContribution2 } from './vscode/src/vs/workbench/common/contributions.js';
export { WorkbenchPhase } from './vscode/src/vs/workbench/common/contributions.js';
import { IProductService } from './vscode/src/vs/platform/product/common/productService.service.js';
import { initialize as initialize$1 } from './workbench.js';
import { checkServicesNotInitialized, startup, waitServicesReady, checkServicesReady, serviceInitializedBarrier, serviceInitializedEmitter } from './lifecycle.js';
import getServiceOverride$2 from '@codingame/monaco-vscode-quickaccess-service-override';
import getServiceOverride$3 from '@codingame/monaco-vscode-files-service-override';
import getServiceOverride$4 from '@codingame/monaco-vscode-extensions-service-override';
import getServiceOverride$5 from '@codingame/monaco-vscode-environment-service-override';
import getServiceOverride$6 from '@codingame/monaco-vscode-layout-service-override';
import getServiceOverride$1 from '@codingame/monaco-vscode-host-service-override';
import getServiceOverride from '@codingame/monaco-vscode-base-service-override';
import { injectCss } from './css.js';
import product from './vscode/src/vs/platform/product/common/product.js';
import { mixin } from './vscode/src/vs/base/common/objects.js';
import { CommandsRegistry } from './vscode/src/vs/platform/commands/common/commands.js';
import { asArray } from './vscode/src/vs/base/common/arrays.js';
import { MenuId, MenuRegistry } from './vscode/src/vs/platform/actions/common/actions.js';
import { Menu } from './vscode/src/vs/workbench/browser/web.api.js';
export { ColorScheme } from './vscode/src/vs/workbench/browser/web.api.js';
export { SyncDescriptor } from './vscode/src/vs/platform/instantiation/common/descriptors.js';
export { ICommandService } from './vscode/src/vs/platform/commands/common/commands.service.js';
export { INotificationService } from './vscode/src/vs/platform/notification/common/notification.service.js';
export { IBulkEditService } from './vscode/src/vs/editor/browser/services/bulkEditService.service.js';
export { ICodeEditorService } from './vscode/src/vs/editor/browser/services/codeEditorService.service.js';
export { ILanguageService } from './vscode/src/vs/editor/common/languages/language.service.js';
export { ILanguageConfigurationService } from './vscode/src/vs/editor/common/languages/languageConfigurationRegistry.service.js';
export { ILanguageFeaturesService } from './vscode/src/vs/editor/common/services/languageFeatures.service.js';
export { IModelService } from './vscode/src/vs/editor/common/services/model.service.js';
export { ITextModelService } from './vscode/src/vs/editor/common/services/resolverService.service.js';
export { IClipboardService } from './vscode/src/vs/platform/clipboard/common/clipboardService.service.js';
export { IDialogService, IFileDialogService } from './vscode/src/vs/platform/dialogs/common/dialogs.service.js';
export { IFileService } from './vscode/src/vs/platform/files/common/files.service.js';
export { IMarkerService } from './vscode/src/vs/platform/markers/common/markers.service.js';
export { IOpenerService } from './vscode/src/vs/platform/opener/common/opener.service.js';
export { IQuickInputService } from './vscode/src/vs/platform/quickinput/common/quickInput.service.js';
export { ICustomEndpointTelemetryService, ITelemetryService } from './vscode/src/vs/platform/telemetry/common/telemetry.service.js';
export { IUriIdentityService } from './vscode/src/vs/platform/uriIdentity/common/uriIdentity.service.js';
export { IBreadcrumbsService } from './vscode/src/vs/workbench/browser/parts/editor/breadcrumbs.service.js';
export { IEditorGroupsService } from './vscode/src/vs/workbench/services/editor/common/editorGroupsService.service.js';
export { GroupDirection, GroupLocation, GroupOrientation, GroupsArrangement, GroupsOrder, MergeGroupMode } from './vscode/src/vs/workbench/services/editor/common/editorGroupsService.js';
export { CloseDirection, EditorsOrder } from './vscode/src/vs/workbench/common/editor.js';
export { EditorInput } from './vscode/src/vs/workbench/common/editor/editorInput.js';
export { IEditorService } from './vscode/src/vs/workbench/services/editor/common/editorService.service.js';
export { IEditorResolverService } from './vscode/src/vs/workbench/services/editor/common/editorResolverService.service.js';
export { ITextEditorService } from './vscode/src/vs/workbench/services/textfile/common/textEditorService.service.js';
export { IWorkbenchEnvironmentService } from './vscode/src/vs/workbench/services/environment/common/environmentService.service.js';
export { IHostService } from './vscode/src/vs/workbench/services/host/browser/host.service.js';
export { ILanguageStatusService } from './vscode/src/vs/workbench/services/languageStatus/common/languageStatusService.service.js';
export { IPaneCompositePartService } from './vscode/src/vs/workbench/services/panecomposite/browser/panecomposite.service.js';
export { IPathService } from './vscode/src/vs/workbench/services/path/common/pathService.service.js';
export { ITextFileService } from './vscode/src/vs/workbench/services/textfile/common/textfiles.service.js';
export { IWorkingCopyFileService } from './vscode/src/vs/workbench/services/workingCopy/common/workingCopyFileService.service.js';
export { IConfigurationService } from './vscode/src/vs/platform/configuration/common/configuration.service.js';
export { IContextKeyService } from './vscode/src/vs/platform/contextkey/common/contextkey.service.js';
export { IThemeService } from './vscode/src/vs/platform/theme/common/themeService.service.js';
export { IWorkbenchThemeService } from './vscode/src/vs/workbench/services/themes/common/workbenchThemeService.service.js';
export { ISnippetsService } from './vscode/src/vs/workbench/contrib/snippets/browser/snippets.service.js';
export { IWorkspaceContextService } from './vscode/src/vs/platform/workspace/common/workspace.service.js';
export { IStorageService } from './vscode/src/vs/platform/storage/common/storage.service.js';
export { IAccessibilitySignalService } from './vscode/src/vs/platform/accessibilitySignal/browser/accessibilitySignalService.service.js';
export { IDebugService } from './vscode/src/vs/workbench/contrib/debug/common/debug.service.js';
export { ILogService, ILoggerService } from './vscode/src/vs/platform/log/common/log.service.js';
export { LogLevel } from './vscode/src/vs/platform/log/common/log.js';
export { IViewDescriptorService } from './vscode/src/vs/workbench/common/views.service.js';
export { IViewsService } from './vscode/src/vs/workbench/services/views/common/viewsService.service.js';
export { ILayoutService } from './vscode/src/vs/platform/layout/browser/layoutService.service.js';
export { IPreferencesService } from './vscode/src/vs/workbench/services/preferences/common/preferences.service.js';
export { IPreferencesSearchService } from './vscode/src/vs/workbench/contrib/preferences/common/preferences.service.js';
export { IKeybindingEditingService } from './vscode/src/vs/workbench/services/keybinding/common/keybindingEditing.service.js';
export { IOutputService } from './vscode/src/vs/workbench/services/output/common/output.service.js';
export { IHistoryService } from './vscode/src/vs/workbench/services/history/common/history.service.js';
export { GoFilter, GoScope } from './vscode/src/vs/workbench/services/history/common/history.js';
export { IRemoteAgentService } from './vscode/src/vs/workbench/services/remote/common/remoteAgentService.service.js';
export { IRemoteAuthorityResolverService } from './vscode/src/vs/platform/remote/common/remoteAuthorityResolver.service.js';
export { IRemoteSocketFactoryService } from './vscode/src/vs/platform/remote/common/remoteSocketFactoryService.service.js';
export { IFilesConfigurationService } from './vscode/src/vs/workbench/services/filesConfiguration/common/filesConfigurationService.service.js';
export { ILabelService } from './vscode/src/vs/platform/label/common/label.service.js';
export { IWorkbenchLayoutService } from './vscode/src/vs/workbench/services/layout/browser/layoutService.service.js';
export { ILanguagePackService } from './vscode/src/vs/platform/languagePacks/common/languagePacks.service.js';
export { IActiveLanguagePackService, ILocaleService } from './vscode/src/vs/workbench/services/localization/common/locale.service.js';
export { NeverShowAgainScope, NoOpNotification, NoOpProgress, NotificationPriority, NotificationsFilter } from './vscode/src/vs/platform/notification/common/notification.js';
export { IActivityService } from './vscode/src/vs/workbench/services/activity/common/activity.service.js';
export { IHoverService } from './vscode/src/vs/platform/hover/browser/hover.service.js';
export { IExplorerService } from './vscode/src/vs/workbench/contrib/files/browser/files.service.js';
export { IStatusbarService } from './vscode/src/vs/workbench/services/statusbar/browser/statusbar.service.js';
export { ITitleService } from './vscode/src/vs/workbench/services/title/browser/titleService.service.js';
export { IBannerService } from './vscode/src/vs/workbench/services/banner/browser/bannerService.service.js';
export { IWorkspaceTrustEnablementService, IWorkspaceTrustManagementService, IWorkspaceTrustRequestService } from './vscode/src/vs/platform/workspace/common/workspaceTrust.service.js';
export { IKeybindingService } from './vscode/src/vs/platform/keybinding/common/keybinding.service.js';
export { ISecretStorageService } from './vscode/src/vs/platform/secrets/common/secrets.service.js';
export { ConfigurationTarget } from './vscode/src/vs/platform/configuration/common/configuration.js';
export { IRemoteExplorerService } from './vscode/src/vs/workbench/services/remote/common/remoteExplorerService.service.js';
export { ITunnelService } from './vscode/src/vs/platform/tunnel/common/tunnel.service.js';
export { ITreeViewsDnDService } from './vscode/src/vs/editor/common/services/treeViewsDndService.service.js';
export { IAccessibleViewService } from './vscode/src/vs/platform/accessibility/browser/accessibleView.service.js';
export { IActionViewItemService } from './vscode/src/vs/platform/actions/browser/actionViewItemService.service.js';
export { IExtensionHostDebugService } from './vscode/src/vs/platform/debug/common/extensionHostDebug.service.js';
export { NullDiagnosticsService } from './vscode/src/vs/platform/diagnostics/common/diagnostics.js';
export { IDiagnosticsService } from './vscode/src/vs/platform/diagnostics/common/diagnostics.service.js';
export { IDownloadService } from './vscode/src/vs/platform/download/common/download.service.js';
export { IEncryptionService } from './vscode/src/vs/platform/encryption/common/encryptionService.service.js';
export { IEnvironmentService } from './vscode/src/vs/platform/environment/common/environment.service.js';
export { IAllowedExtensionsService, IExtensionGalleryService, IExtensionTipsService, IGlobalExtensionEnablementService } from './vscode/src/vs/platform/extensionManagement/common/extensionManagement.service.js';
export { IExtensionStorageService } from './vscode/src/vs/platform/extensionManagement/common/extensionStorage.service.js';
export { IExtensionsProfileScannerService } from './vscode/src/vs/platform/extensionManagement/common/extensionsProfileScannerService.service.js';
export { IExtensionsScannerService } from './vscode/src/vs/platform/extensionManagement/common/extensionsScannerService.service.js';
export { IExtensionRecommendationNotificationService } from './vscode/src/vs/platform/extensionRecommendations/common/extensionRecommendations.service.js';
export { IExtensionResourceLoaderService } from './vscode/src/vs/platform/extensionResourceLoader/common/extensionResourceLoader.service.js';
export { IBuiltinExtensionsScannerService } from './vscode/src/vs/platform/extensions/common/extensions.service.js';
export { IKeyboardLayoutService } from './vscode/src/vs/platform/keyboardLayout/common/keyboardLayout.service.js';
export { NullPolicyService } from './vscode/src/vs/platform/policy/common/policy.js';
export { IPolicyService } from './vscode/src/vs/platform/policy/common/policy.service.js';
export { IRemoteExtensionsScannerService } from './vscode/src/vs/platform/remote/common/remoteExtensionsScanner.service.js';
export { IRequestService } from './vscode/src/vs/platform/request/common/request.service.js';
export { ISignService } from './vscode/src/vs/platform/sign/common/sign.service.js';
export { NullEndpointTelemetryService } from './vscode/src/vs/platform/telemetry/common/telemetryUtils.js';
export { TerminalLocation } from './vscode/src/vs/platform/terminal/common/terminal.js';
export { ITerminalLogService } from './vscode/src/vs/platform/terminal/common/terminal.service.js';
export { State } from './vscode/src/vs/platform/update/common/update.js';
export { IUpdateService } from './vscode/src/vs/platform/update/common/update.service.js';
export { UriIdentityService } from './vscode/src/vs/platform/uriIdentity/common/uriIdentityService.js';
export { IURLService } from './vscode/src/vs/platform/url/common/url.service.js';
export { toUserDataProfile } from './vscode/src/vs/platform/userDataProfile/common/userDataProfile.js';
export { IUserDataProfilesService } from './vscode/src/vs/platform/userDataProfile/common/userDataProfile.service.js';
export { IUserDataProfileStorageService } from './vscode/src/vs/platform/userDataProfile/common/userDataProfileStorageService.service.js';
export { IIgnoredExtensionsManagementService } from './vscode/src/vs/platform/userDataSync/common/ignoredExtensions.service.js';
export { IUserDataAutoSyncService, IUserDataSyncEnablementService, IUserDataSyncLocalStoreService, IUserDataSyncLogService, IUserDataSyncResourceProviderService, IUserDataSyncService, IUserDataSyncStoreManagementService, IUserDataSyncStoreService, IUserDataSyncUtilService } from './vscode/src/vs/platform/userDataSync/common/userDataSync.service.js';
export { IUserDataSyncAccountService } from './vscode/src/vs/platform/userDataSync/common/userDataSyncAccount.service.js';
export { IUserDataSyncMachinesService } from './vscode/src/vs/platform/userDataSync/common/userDataSyncMachines.service.js';
export { ICanonicalUriService } from './vscode/src/vs/platform/workspace/common/canonicalUri.service.js';
export { IEditSessionIdentityService } from './vscode/src/vs/platform/workspace/common/editSessions.service.js';
export { WorkspaceTrustUriResponse } from './vscode/src/vs/platform/workspace/common/workspaceTrust.js';
export { IWorkspacesService } from './vscode/src/vs/platform/workspaces/common/workspaces.service.js';
export { IExtensionStatusBarItemService, StatusBarUpdateKind } from './vscode/src/vs/workbench/api/browser/statusBarService.js';
export { DEFAULT_EDITOR_PART_OPTIONS } from './vscode/src/vs/workbench/browser/parts/editor/editor.js';
export { IChatAccessibilityService, IChatCodeBlockContextProviderService, IChatWidgetService, IQuickChatService } from './vscode/src/vs/workbench/contrib/chat/browser/chat.service.js';
export { IChatAgentNameService, IChatAgentService } from './vscode/src/vs/workbench/contrib/chat/common/participants/chatAgents.service.js';
export { ICodeMapperService } from './vscode/src/vs/workbench/contrib/chat/common/editing/chatCodeMapperService.service.js';
export { IChatEditingService } from './vscode/src/vs/workbench/contrib/chat/common/editing/chatEditingService.service.js';
export { IChatService } from './vscode/src/vs/workbench/contrib/chat/common/chatService/chatService.service.js';
export { IChatSlashCommandService } from './vscode/src/vs/workbench/contrib/chat/common/participants/chatSlashCommands.service.js';
export { IChatVariablesService } from './vscode/src/vs/workbench/contrib/chat/common/attachments/chatVariables.service.js';
export { IChatWidgetHistoryService } from './vscode/src/vs/workbench/contrib/chat/common/widget/chatWidgetHistoryService.service.js';
export { ILanguageModelIgnoredFilesService } from './vscode/src/vs/workbench/contrib/chat/common/ignoredFiles.service.js';
export { ILanguageModelStatsService } from './vscode/src/vs/workbench/contrib/chat/common/languageModelStats.service.js';
export { ILanguageModelToolsService } from './vscode/src/vs/workbench/contrib/chat/common/tools/languageModelToolsService.service.js';
export { ILanguageModelsService } from './vscode/src/vs/workbench/contrib/chat/common/languageModels.service.js';
export { ICommentService } from './vscode/src/vs/workbench/contrib/comments/browser/commentService.service.js';
export { ICustomEditorService } from './vscode/src/vs/workbench/contrib/customEditor/common/customEditor.service.js';
export { IDebugVisualizerService } from './vscode/src/vs/workbench/contrib/debug/common/debugVisualizers.service.js';
export { IEditSessionsLogService, IEditSessionsStorageService } from './vscode/src/vs/workbench/contrib/editSessions/common/editSessions.service.js';
export { IExtensionsWorkbenchService } from './vscode/src/vs/workbench/contrib/extensions/common/extensions.service.js';
export { IExternalUriOpenerService } from './vscode/src/vs/workbench/contrib/externalUriOpener/common/externalUriOpenerService.service.js';
export { IInlineChatSessionService } from './vscode/src/vs/workbench/contrib/inlineChat/browser/inlineChatSessionService.service.js';
export { IInteractiveDocumentService } from './vscode/src/vs/workbench/contrib/interactive/browser/interactiveDocumentService.service.js';
export { IInteractiveHistoryService } from './vscode/src/vs/workbench/contrib/interactive/browser/interactiveHistoryService.service.js';
export { ITroubleshootIssueService } from './vscode/src/vs/workbench/contrib/issue/browser/issueTroubleshoot.service.js';
export { IIssueFormService, IWorkbenchIssueService } from './vscode/src/vs/workbench/contrib/issue/common/issue.service.js';
export { IDefaultLogLevelsService } from './vscode/src/vs/workbench/services/log/common/defaultLogLevels.service.js';
export { IMultiDiffSourceResolverService } from './vscode/src/vs/workbench/contrib/multiDiffEditor/browser/multiDiffSourceResolverService.service.js';
export { INotebookOriginalCellModelFactory } from './vscode/src/vs/workbench/contrib/notebook/browser/diff/inlineDiff/notebookOriginalCellModelFactory.service.js';
export { INotebookOriginalModelReferenceFactory } from './vscode/src/vs/workbench/contrib/notebook/browser/diff/inlineDiff/notebookOriginalModelRefFactory.service.js';
export { INotebookEditorService } from './vscode/src/vs/workbench/contrib/notebook/browser/services/notebookEditorService.service.js';
export { INotebookCellStatusBarService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookCellStatusBarService.service.js';
export { INotebookEditorModelResolverService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookEditorModelResolverService.service.js';
export { INotebookExecutionService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookExecutionService.service.js';
export { INotebookExecutionStateService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookExecutionStateService.service.js';
export { INotebookKernelHistoryService, INotebookKernelService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookKernelService.service.js';
export { INotebookKeymapService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookKeymapService.service.js';
export { INotebookLoggingService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookLoggingService.service.js';
export { INotebookRendererMessagingService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookRendererMessagingService.service.js';
export { INotebookService } from './vscode/src/vs/workbench/contrib/notebook/common/notebookService.service.js';
export { INotebookEditorWorkerService } from './vscode/src/vs/workbench/contrib/notebook/common/services/notebookWorkerService.service.js';
export { IQuickDiffModelService } from './vscode/src/vs/workbench/contrib/scm/browser/quickDiffModel.service.js';
export { IQuickDiffService } from './vscode/src/vs/workbench/contrib/scm/common/quickDiff.service.js';
export { ISCMService, ISCMViewService } from './vscode/src/vs/workbench/contrib/scm/common/scm.service.js';
export { IReplaceService } from './vscode/src/vs/workbench/contrib/search/browser/replace.service.js';
export { ISearchViewModelWorkbenchService } from './vscode/src/vs/workbench/contrib/search/browser/searchTreeModel/searchViewModelWorkbenchService.service.js';
export { INotebookSearchService } from './vscode/src/vs/workbench/contrib/search/common/notebookSearch.service.js';
export { ISearchHistoryService } from './vscode/src/vs/workbench/contrib/search/common/searchHistoryService.service.js';
export { IShareService } from './vscode/src/vs/workbench/contrib/share/common/share.service.js';
export { ISpeechService } from './vscode/src/vs/workbench/contrib/speech/common/speechService.service.js';
export { NoOpWorkspaceTagsService } from './vscode/src/vs/workbench/contrib/tags/browser/workspaceTagsService.js';
export { IWorkspaceTagsService } from './vscode/src/vs/workbench/contrib/tags/common/workspaceTags.service.js';
export { ITaskService } from './vscode/src/vs/workbench/contrib/tasks/common/taskService.service.js';
export { ITerminalConfigurationService, ITerminalEditorService, ITerminalGroupService, ITerminalInstanceService, ITerminalService } from './vscode/src/vs/workbench/contrib/terminal/browser/terminal.service.js';
export { IEnvironmentVariableService } from './vscode/src/vs/workbench/contrib/terminal/common/environmentVariable.service.js';
export { ITerminalProfileResolverService, ITerminalProfileService } from './vscode/src/vs/workbench/contrib/terminal/common/terminal.service.js';
export { ITerminalContributionService } from './vscode/src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.service.js';
export { ITerminalLinkProviderService } from './vscode/src/vs/workbench/contrib/terminalContrib/links/browser/links.service.js';
export { ITerminalQuickFixService } from './vscode/src/vs/workbench/contrib/terminalContrib/quickFix/browser/quickFix.service.js';
export { ITerminalCompletionService } from './vscode/src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.service.js';
export { ITestCoverageService } from './vscode/src/vs/workbench/contrib/testing/common/testCoverageService.service.js';
export { ITestExplorerFilterState } from './vscode/src/vs/workbench/contrib/testing/common/testExplorerFilterState.service.js';
export { ITestProfileService } from './vscode/src/vs/workbench/contrib/testing/common/testProfileService.service.js';
export { ITestResultService } from './vscode/src/vs/workbench/contrib/testing/common/testResultService.service.js';
export { ITestResultStorage } from './vscode/src/vs/workbench/contrib/testing/common/testResultStorage.service.js';
export { ITestService } from './vscode/src/vs/workbench/contrib/testing/common/testService.service.js';
export { ITestingContinuousRunService } from './vscode/src/vs/workbench/contrib/testing/common/testingContinuousRunService.service.js';
export { ITestingDecorationsService } from './vscode/src/vs/workbench/contrib/testing/common/testingDecorations.service.js';
export { ITestingPeekOpener } from './vscode/src/vs/workbench/contrib/testing/common/testingPeekOpener.service.js';
export { ITimelineService } from './vscode/src/vs/workbench/contrib/timeline/common/timeline.service.js';
export { ITrustedDomainService } from './vscode/src/vs/workbench/contrib/url/browser/trustedDomainService.service.js';
export { IWebviewService } from './vscode/src/vs/workbench/contrib/webview/browser/webview.service.js';
export { IWebviewWorkbenchService } from './vscode/src/vs/workbench/contrib/webviewPanel/browser/webviewWorkbenchService.service.js';
export { IWebviewViewService } from './vscode/src/vs/workbench/contrib/webviewView/browser/webviewViewService.service.js';
export { IWalkthroughsService } from './vscode/src/vs/workbench/contrib/welcomeGettingStarted/browser/gettingStartedService.service.js';
export { IAccessibleViewInformationService } from './vscode/src/vs/workbench/services/accessibility/common/accessibleViewInformationService.service.js';
export { IAiEmbeddingVectorService } from './vscode/src/vs/workbench/services/aiEmbeddingVector/common/aiEmbeddingVectorService.service.js';
export { IAiRelatedInformationService } from './vscode/src/vs/workbench/services/aiRelatedInformation/common/aiRelatedInformation.service.js';
export { IWorkbenchAssignmentService } from './vscode/src/vs/workbench/services/assignment/common/assignmentService.service.js';
export { IAuthenticationAccessService } from './vscode/src/vs/workbench/services/authentication/browser/authenticationAccessService.service.js';
export { IAuthenticationUsageService } from './vscode/src/vs/workbench/services/authentication/browser/authenticationUsageService.service.js';
export { IAuthenticationExtensionsService, IAuthenticationService } from './vscode/src/vs/workbench/services/authentication/common/authentication.service.js';
export { IAuxiliaryWindowService } from './vscode/src/vs/workbench/services/auxiliaryWindow/browser/auxiliaryWindowService.service.js';
export { IJSONEditingService } from './vscode/src/vs/workbench/services/configuration/common/jsonEditing.service.js';
export { IConfigurationResolverService } from './vscode/src/vs/workbench/services/configurationResolver/common/configurationResolver.service.js';
export { IDecorationsService } from './vscode/src/vs/workbench/services/decorations/common/decorations.service.js';
export { ICustomEditorLabelService } from './vscode/src/vs/workbench/services/editor/common/customEditorLabelService.service.js';
export { IEditorPaneService } from './vscode/src/vs/workbench/services/editor/common/editorPaneService.service.js';
export { IExtensionBisectService } from './vscode/src/vs/workbench/services/extensionManagement/browser/extensionBisect.service.js';
export { IExtensionFeaturesManagementService } from './vscode/src/vs/workbench/services/extensionManagement/common/extensionFeatures.service.js';
export { IExtensionManagementServerService, IWebExtensionsScannerService, IWorkbenchExtensionEnablementService, IWorkbenchExtensionManagementService } from './vscode/src/vs/workbench/services/extensionManagement/common/extensionManagement.service.js';
export { IExtensionIgnoredRecommendationsService, IExtensionRecommendationsService } from './vscode/src/vs/workbench/services/extensionRecommendations/common/extensionRecommendations.service.js';
export { IWorkspaceExtensionsConfigService } from './vscode/src/vs/workbench/services/extensionRecommendations/common/workspaceExtensionsConfig.service.js';
export { IExtensionUrlHandler } from './vscode/src/vs/workbench/services/extensions/browser/extensionUrlHandler.service.js';
export { IExtensionManifestPropertiesService } from './vscode/src/vs/workbench/services/extensions/common/extensionManifestPropertiesService.service.js';
export { NullExtensionService } from './vscode/src/vs/workbench/services/extensions/common/extensions.js';
export { IExtensionService } from './vscode/src/vs/workbench/services/extensions/common/extensions.service.js';
export { IElevatedFileService } from './vscode/src/vs/workbench/services/files/common/elevatedFileService.service.js';
export { IIntegrityService } from './vscode/src/vs/workbench/services/integrity/common/integrity.service.js';
export { FallbackKeyboardMapper } from './vscode/src/vs/workbench/services/keybinding/common/fallbackKeyboardMapper.js';
export { ILanguageDetectionService } from './vscode/src/vs/workbench/services/languageDetection/common/languageDetectionWorkerService.service.js';
export { ILifecycleService } from './vscode/src/vs/workbench/services/lifecycle/common/lifecycle.service.js';
export { INotebookDocumentService } from './vscode/src/vs/workbench/services/notebook/common/notebookDocumentService.service.js';
export { IOutlineService } from './vscode/src/vs/workbench/services/outline/browser/outline.service.js';
export { ISearchService } from './vscode/src/vs/workbench/services/search/common/search.service.js';
export { IEmbedderTerminalService } from './vscode/src/vs/workbench/services/terminal/common/embedderTerminalService.service.js';
export { ITextMateTokenizationService } from './vscode/src/vs/workbench/services/textMate/browser/textMateTokenizationFeature.service.js';
export { IHostColorSchemeService } from './vscode/src/vs/workbench/services/themes/common/hostColorSchemeService.service.js';
export { ITimerService } from './vscode/src/vs/workbench/services/timer/browser/timerService.service.js';
export { IUntitledTextEditorService } from './vscode/src/vs/workbench/services/untitled/common/untitledTextEditorService.service.js';
export { IUserActivityService } from './vscode/src/vs/workbench/services/userActivity/common/userActivityService.service.js';
export { IUserDataInitializationService } from './vscode/src/vs/workbench/services/userData/browser/userDataInit.service.js';
export { IRemoteUserDataProfilesService } from './vscode/src/vs/workbench/services/userDataProfile/common/remoteUserDataProfiles.service.js';
export { IUserDataProfileImportExportService, IUserDataProfileManagementService, IUserDataProfileService } from './vscode/src/vs/workbench/services/userDataProfile/common/userDataProfile.service.js';
export { UserDataProfileService } from './vscode/src/vs/workbench/services/userDataProfile/common/userDataProfileService.js';
export { AccountStatus } from './vscode/src/vs/workbench/services/userDataSync/common/userDataSync.js';
export { IUserDataSyncWorkbenchService } from './vscode/src/vs/workbench/services/userDataSync/common/userDataSync.service.js';
export { IWorkingCopyBackupService } from './vscode/src/vs/workbench/services/workingCopy/common/workingCopyBackup.service.js';
export { IWorkingCopyEditorService } from './vscode/src/vs/workbench/services/workingCopy/common/workingCopyEditorService.service.js';
export { IWorkingCopyHistoryService } from './vscode/src/vs/workbench/services/workingCopy/common/workingCopyHistory.service.js';
export { IWorkingCopyService } from './vscode/src/vs/workbench/services/workingCopy/common/workingCopyService.service.js';
export { IWorkspaceEditingService } from './vscode/src/vs/workbench/services/workspaces/common/workspaceEditing.service.js';
export { IWorkspaceIdentityService } from './vscode/src/vs/workbench/services/workspaces/common/workspaceIdentityService.service.js';
export { IChatMarkdownAnchorService } from './vscode/src/vs/workbench/contrib/chat/browser/widget/chatContentParts/chatMarkdownAnchorService.service.js';
export { getBuiltInExtensionTranslationsUris, getExtensionIdProvidingCurrentLocale } from './l10n.js';
export { unsupported } from './tools.js';
export { IChatEntitlementService } from './vscode/src/vs/workbench/services/chat/common/chatEntitlementService.service.js';
export { IPromptsService } from './vscode/src/vs/workbench/contrib/chat/common/promptSyntax/service/promptsService.service.js';
export { ISuggestMemoryService } from './vscode/src/vs/editor/contrib/suggest/browser/suggestMemory.service.js';
export { LanguageConfigurationService } from './vscode/src/vs/editor/common/languages/languageConfigurationRegistry.js';
export { ISemanticTokensStylingService } from './vscode/src/vs/editor/common/services/semanticTokensStyling.service.js';
export { ILanguageFeatureDebounceService } from './vscode/src/vs/editor/common/services/languageFeatureDebounce.service.js';
export { IDiffProviderFactoryService } from './vscode/src/vs/editor/browser/widget/diffEditor/diffProviderFactoryService.service.js';
export { IOutlineModelService } from './vscode/src/vs/editor/contrib/documentSymbols/browser/outlineModel.service.js';
export { IMarkerNavigationService } from './vscode/src/vs/editor/contrib/gotoError/browser/markerNavigationService.service.js';
export { ICodeLensCache } from './vscode/src/vs/editor/contrib/codelens/browser/codeLensCache.service.js';
export { IInlayHintsCache } from './vscode/src/vs/editor/contrib/inlayHints/browser/inlayHintsController.service.js';
export { ISymbolNavigationService } from './vscode/src/vs/editor/contrib/gotoSymbol/browser/symbolNavigation.service.js';
export { IEditorCancellationTokens } from './vscode/src/vs/editor/contrib/editorState/browser/keybindingCancellation.service.js';
export { IPeekViewService } from './vscode/src/vs/editor/contrib/peekView/browser/peekView.service.js';
export { IUndoRedoService } from './vscode/src/vs/platform/undoRedo/common/undoRedo.service.js';
export { IActionWidgetService } from './vscode/src/vs/platform/actionWidget/browser/actionWidget.service.js';
export { IMcpElicitationService, IMcpSamplingService, IMcpService, IMcpWorkbenchService } from './vscode/src/vs/workbench/contrib/mcp/common/mcpTypes.service.js';
export { IMcpResourceScannerService } from './vscode/src/vs/platform/mcp/common/mcpResourceScannerService.service.js';
export { IMcpRegistry } from './vscode/src/vs/workbench/contrib/mcp/common/mcpRegistryTypes.service.js';
export { IExtensionGalleryManifestService } from './vscode/src/vs/platform/extensionManagement/common/extensionGalleryManifest.service.js';
export { ISharedWebContentExtractorService, IWebContentExtractorService } from './vscode/src/vs/platform/webContentExtractor/common/webContentExtractor.service.js';
export { IDefaultAccountService } from './vscode/src/vs/platform/defaultAccount/common/defaultAccount.service.js';
export { IChatTransferService } from './vscode/src/vs/workbench/contrib/chat/common/model/chatTransferService.service.js';
export { IChatStatusItemService } from './vscode/src/vs/workbench/contrib/chat/browser/chatStatus/chatStatusItemService.service.js';
export { IAiSettingsSearchService } from './vscode/src/vs/workbench/services/aiSettingsSearch/common/aiSettingsSearch.service.js';
export { IDynamicAuthenticationProviderStorageService } from './vscode/src/vs/workbench/services/authentication/common/dynamicAuthenticationProviderStorage.service.js';
export { IAuthenticationMcpService } from './vscode/src/vs/workbench/services/authentication/browser/authenticationMcpService.service.js';
export { IAuthenticationMcpAccessService } from './vscode/src/vs/workbench/services/authentication/browser/authenticationMcpAccessService.service.js';
export { IAuthenticationMcpUsageService } from './vscode/src/vs/workbench/services/authentication/browser/authenticationMcpUsageService.service.js';
export { IBrowserElementsService } from './vscode/src/vs/workbench/services/browserElements/browser/browserElementsService.service.js';
export { IChatContextPickService } from './vscode/src/vs/workbench/contrib/chat/browser/attachments/chatContextPickService.service.js';
export { IMcpGalleryService, IMcpManagementService } from './vscode/src/vs/platform/mcp/common/mcpManagement.service.js';
export { ITreeSitterThemeService } from './vscode/src/vs/editor/common/services/treeSitter/treeSitterThemeService.service.js';
export { ITreeSitterLibraryService } from './vscode/src/vs/editor/common/services/treeSitter/treeSitterLibraryService.service.js';
export { IChatAttachmentResolveService } from './vscode/src/vs/workbench/contrib/chat/browser/attachments/chatAttachmentResolveService.service.js';
export { IRemoteCodingAgentsService } from './vscode/src/vs/workbench/contrib/remoteCodingAgents/common/remoteCodingAgentsService.service.js';
export { IAuthenticationQueryService } from './vscode/src/vs/workbench/services/authentication/common/authenticationQuery.service.js';
export { IWorkbenchMcpManagementService } from './vscode/src/vs/workbench/services/mcp/common/mcpWorkbenchManagementService.service.js';
export { IChatLayoutService } from './vscode/src/vs/workbench/contrib/chat/common/widget/chatLayoutService.service.js';
export { IAiEditTelemetryService } from './vscode/src/vs/workbench/contrib/editTelemetry/browser/telemetry/aiEditTelemetry/aiEditTelemetryService.service.js';
export { IInlineCompletionsUnificationService } from './vscode/src/vs/workbench/services/inlineCompletions/common/inlineCompletionsUnification.service.js';
export { IMcpGalleryManifestService } from './vscode/src/vs/platform/mcp/common/mcpGalleryManifest.service.js';
export { IDataChannelService } from './vscode/src/vs/platform/dataChannel/common/dataChannel.service.js';
export { IProgressService } from './vscode/src/vs/platform/progress/common/progress.service.js';
export { IMarkdownRendererService } from './vscode/src/vs/platform/markdown/browser/markdownRenderer.service.js';
export { IChatContextService } from './vscode/src/vs/workbench/contrib/chat/browser/contextContrib/chatContextService.service.js';
export { ILanguageModelToolsConfirmationService } from './vscode/src/vs/workbench/contrib/chat/common/tools/languageModelToolsConfirmationService.service.js';
export { IRandomService } from './vscode/src/vs/workbench/contrib/editTelemetry/browser/randomService.service.js';
export { ISCMRepositorySelectionMode } from './vscode/src/vs/workbench/contrib/scm/common/scm.js';
export { IAgentSessionsService } from './vscode/src/vs/workbench/contrib/chat/browser/agentSessions/agentSessionsService.service.js';
export { IUserAttentionService } from './vscode/src/vs/workbench/services/userAttention/common/userAttentionService.service.js';
export { ICodeCompareModelService } from './vscode/src/vs/workbench/contrib/chat/browser/widget/chatContentParts/chatTextEditContentPart.service.js';

if (window.monacoVscodeApiBuildId != null && window.monacoVscodeApiBuildId !== "1.108.2-21930f76-a5a8-4dff-86b6-358ec12025e6") {
    throw new Error(`Another version of monaco-vscode-api has already been loaded. Trying to load ${"1.108.2-21930f76-a5a8-4dff-86b6-358ec12025e6"}, ${window.monacoVscodeApiBuildId} is already loaded`);
}
window.monacoVscodeApiBuildId = "1.108.2-21930f76-a5a8-4dff-86b6-358ec12025e6";
function registerCommands(options) {
    function asMenuId(menu) {
        switch (menu) {
            case Menu.CommandPalette:
                return MenuId.CommandPalette;
            case Menu.StatusBarWindowIndicatorMenu:
                return MenuId.StatusBarWindowIndicatorMenu;
        }
    }
    if (Array.isArray(options.commands)) {
        for (const command of options.commands) {
            CommandsRegistry.registerCommand(command.id, (accessor, ...args) => {
                return command.handler(...args);
            });
            if (command.label) {
                for (const menu of asArray(command.menu ?? Menu.CommandPalette)) {
                    MenuRegistry.appendMenuItem(asMenuId(menu), {
                        command: { id: command.id, title: command.label }
                    });
                }
            }
        }
    }
}
async function initialize(overrides, container = document.body, configuration = {}, env) {
    checkServicesNotInitialized();
    injectCss(container);
    registerCommands(configuration);
    initialize$1(container, configuration, env);
    const productService = mixin({ _serviceBrand: undefined, ...product }, configuration.productConfiguration);
    const instantiationService = StandaloneServices.initialize({
        [IProductService.toString()]: productService,
        ...getServiceOverride$6(),
        ...getServiceOverride$5(),
        ...getServiceOverride$4(),
        ...getServiceOverride$3(),
        ...getServiceOverride$2(),
        ...getServiceOverride$1(),
        ...getServiceOverride(),
        ...overrides
    });
    await startup(instantiationService);
}
async function getService(identifier) {
    await waitServicesReady();
    return StandaloneServices.get(identifier);
}
async function createInstance(ctor, ...args) {
    await waitServicesReady();
    return StandaloneServices.get(IInstantiationService).createInstance(ctor, ...args);
}
function createInstanceSync(ctor, ...args) {
    checkServicesReady();
    return StandaloneServices.get(IInstantiationService).createInstance(ctor, ...args);
}
function withReadyServices(callback) {
    if (serviceInitializedBarrier.isOpen()) {
        return StandaloneServices.get(IInstantiationService).invokeFunction(callback);
    }
    const disposable = new DisposableStore();
    const listener = disposable.add(serviceInitializedEmitter.event(() => {
        listener.dispose();
        disposable.add(StandaloneServices.get(IInstantiationService).invokeFunction(callback));
    }));
    return disposable;
}
const registerWorkbenchContribution = (id, contribution, instantiation) => {
    let Contribution = class Contribution {
        constructor(instantiationService) {
            instantiationService.invokeFunction(contribution);
        }
    };
    Contribution = __decorate([
        __param(0, IInstantiationService)
    ], Contribution);
    registerWorkbenchContribution2(id, Contribution, instantiation);
};

export { IInstantiationService, IProductService, Menu, StandaloneServices, createInstance, createInstanceSync, getService, initialize, registerWorkbenchContribution, withReadyServices };
