// @codingame/monaco-vscode-view-common-service-override/vscode/src/vs/workbench/contrib/mergeEditor/browser/mergeEditorAccessibilityHelp.js

import { ICodeEditorService } from '@codingame/monaco-vscode-api/vscode/vs/editor/browser/services/codeEditorService.service';
import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { AccessibleViewType, AccessibleContentProvider, AccessibleViewProviderId } from '@codingame/monaco-vscode-api/vscode/vs/platform/accessibility/browser/accessibleView';
import { ContextKeyEqualsExpr } from '@codingame/monaco-vscode-api/vscode/vs/platform/contextkey/common/contextkey';
import { AccessibilityVerbositySettingId } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/accessibility/browser/accessibilityConfiguration';

class MergeEditorAccessibilityHelpProvider {
    constructor() {
        this.name = 'mergeEditor';
        this.type = AccessibleViewType.Help;
        this.priority = 125;
        this.when = ContextKeyEqualsExpr.create('isMergeEditor', true);
    }
    getProvider(accessor) {
        const codeEditorService = accessor.get(ICodeEditorService);
        const codeEditor = codeEditorService.getActiveCodeEditor() || codeEditorService.getFocusedCodeEditor();
        if (!codeEditor) {
            return;
        }
        const content = [
            ( localize(9108, "You are in a merge editor.")),
            ( localize(
            9109,
            "Navigate between merge conflicts using the commands Go to Next Unhandled Conflict{0} and Go to Previous Unhandled Conflict{1}.",
            '<keybinding:merge.goToNextUnhandledConflict>',
            '<keybinding:merge.goToPreviousUnhandledConflict>'
        )),
            ( localize(
            9110,
            "Run the command Merge Editor: Accept All Incoming Changes from the Left{0} and Merge Editor: Accept All Current Changes from the Right{1}",
            '<keybinding:merge.acceptAllInput1>',
            '<keybinding:merge.acceptAllInput2>'
        )),
            ( localize(9111, "Complete the Merge{0}.", '<keybinding:mergeEditor.acceptMerge>')),
            ( localize(
            9112,
            "Toggle between merge editor inputs, incoming and current changes {0}.",
            '<keybinding:mergeEditor.toggleBetweenInputs>'
        )),
        ];
        return ( new AccessibleContentProvider(
            AccessibleViewProviderId.MergeEditor,
            { type: AccessibleViewType.Help },
            () => content.join('\n'),
            () => codeEditor.focus(),
            AccessibilityVerbositySettingId.MergeEditor
        ));
    }
}

export { MergeEditorAccessibilityHelpProvider };
