// @codingame/monaco-vscode-view-common-service-override/vscode/src/vs/workbench/contrib/mergeEditor/browser/mergeEditorSerializer.js

import { onUnexpectedError } from '@codingame/monaco-vscode-api/vscode/vs/base/common/errors';
import { parse } from '@codingame/monaco-vscode-api/vscode/vs/base/common/marshalling';
import { MergeEditorInput, MergeEditorInputData } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/mergeEditor/browser/mergeEditorInput';

class MergeEditorSerializer {
    canSerialize() {
        return true;
    }
    serialize(editor) {
        return JSON.stringify(this.toJSON(editor));
    }
    toJSON(editor) {
        return {
            base: editor.base,
            input1: editor.input1,
            input2: editor.input2,
            result: editor.result,
        };
    }
    deserialize(instantiationService, raw) {
        try {
            const data = parse(raw);
            return instantiationService.createInstance(MergeEditorInput, data.base, ( new MergeEditorInputData(
                data.input1.uri,
                data.input1.title,
                data.input1.detail,
                data.input1.description
            )), ( new MergeEditorInputData(
                data.input2.uri,
                data.input2.title,
                data.input2.detail,
                data.input2.description
            )), data.result);
        }
        catch (err) {
            onUnexpectedError(err);
            return undefined;
        }
    }
}

export { MergeEditorSerializer };
