// @codingame/monaco-vscode-view-common-service-override/vscode/src/vs/workbench/api/browser/viewsExtensionPoint.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { MarkdownString } from '@codingame/monaco-vscode-api/vscode/vs/base/common/htmlContent';
import { Disposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import { joinPath } from '@codingame/monaco-vscode-api/vscode/vs/base/common/resources';
import { isFalsyOrWhitespace } from '@codingame/monaco-vscode-api/vscode/vs/base/common/strings';
import { ThemeIcon } from '@codingame/monaco-vscode-api/vscode/vs/base/common/themables';
import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { ContextKeyExpr } from '@codingame/monaco-vscode-api/vscode/vs/platform/contextkey/common/contextkey';
import { ExtensionIdentifierSet, ExtensionIdentifier } from '@codingame/monaco-vscode-api/vscode/vs/platform/extensions/common/extensions';
import { SyncDescriptor } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/descriptors';
import { IInstantiationService } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/instantiation';
import { ILogService } from '@codingame/monaco-vscode-api/vscode/vs/platform/log/common/log.service';
import { Registry } from '@codingame/monaco-vscode-api/vscode/vs/platform/registry/common/platform';
import { Extensions as Extensions$1 } from '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/panecomposite';
import { CustomTreeView, TreeViewPane } from '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/parts/views/treeView';
import { ViewPaneContainer } from '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/parts/views/viewPaneContainer';
import { registerWorkbenchContribution2, WorkbenchPhase } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/contributions';
import { Extensions, ViewContainerLocation } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/views';
import { VIEWLET_ID as VIEWLET_ID$3 } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/debug/common/debug';
import { VIEWLET_ID as VIEWLET_ID$1 } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/files/common/files';
import { VIEWLET_ID } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/remote/browser/remoteExplorer';
import { VIEWLET_ID as VIEWLET_ID$2 } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/scm/common/scm';
import { WebviewViewPane } from '../../contrib/webviewView/browser/webviewViewPane.js';
import { Extensions as Extensions$2 } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/extensionManagement/common/extensionFeatures';
import { isProposedApiEnabled } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/extensions/common/extensions';
import { ExtensionsRegistry } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/extensions/common/extensionsRegistry';

const viewsContainerSchema = {
    type: 'object',
    properties: {
        id: {
            description: ( localize(
                2576,
                "Unique id used to identify the container in which views can be contributed using 'views' contribution point"
            )),
            type: 'string',
            pattern: '^[a-zA-Z0-9_-]+$'
        },
        title: {
            description: ( localize(2577, 'Human readable string used to render the container')),
            type: 'string'
        },
        icon: {
            description: ( localize(
                2578,
                "Path to the container icon. Icons are 24x24 centered on a 50x40 block and have a fill color of 'rgb(215, 218, 224)' or '#d7dae0'. It is recommended that icons be in SVG, though any image file type is accepted."
            )),
            type: 'string'
        }
    },
    required: ['id', 'title', 'icon']
};
const viewsContainersContribution = {
    description: ( localize(2579, 'Contributes views containers to the editor')),
    type: 'object',
    properties: {
        'activitybar': {
            description: ( localize(2580, "Contribute views containers to Activity Bar")),
            type: 'array',
            items: viewsContainerSchema
        },
        'panel': {
            description: ( localize(2581, "Contribute views containers to Panel")),
            type: 'array',
            items: viewsContainerSchema
        },
        'secondarySidebar': {
            description: ( localize(2582, "Contribute views containers to Secondary Side Bar")),
            type: 'array',
            items: viewsContainerSchema
        }
    },
    additionalProperties: false
};
var ViewType;
(function (ViewType) {
    ViewType["Tree"] = "tree";
    ViewType["Webview"] = "webview";
})(ViewType || (ViewType = {}));
var InitialVisibility;
(function (InitialVisibility) {
    InitialVisibility["Visible"] = "visible";
    InitialVisibility["Hidden"] = "hidden";
    InitialVisibility["Collapsed"] = "collapsed";
})(InitialVisibility || (InitialVisibility = {}));
const viewDescriptor = {
    type: 'object',
    required: ['id', 'name', 'icon'],
    defaultSnippets: [{ body: { id: '${1:id}', name: '${2:name}', icon: '${3:icon}' } }],
    properties: {
        type: {
            markdownDescription: ( localize(
                2583,
                "Type of the view. This can either be `tree` for a tree view based view or `webview` for a webview based view. The default is `tree`."
            )),
            type: 'string',
            enum: [
                'tree',
                'webview',
            ],
            markdownEnumDescriptions: [
                ( localize(2584, "The view is backed by a `TreeView` created by `createTreeView`.")),
                ( localize(
                2585,
                "The view is backed by a `WebviewView` registered by `registerWebviewViewProvider`."
            )),
            ]
        },
        id: {
            markdownDescription: ( localize(
                2586,
                'Identifier of the view. This should be unique across all views. It is recommended to include your extension id as part of the view id. Use this to register a data provider through `vscode.window.registerTreeDataProviderForView` API. Also to trigger activating your extension by registering `onView:${id}` event to `activationEvents`.'
            )),
            type: 'string'
        },
        name: {
            description: ( localize(2587, 'The human-readable name of the view. Will be shown')),
            type: 'string'
        },
        when: {
            description: ( localize(2588, 'Condition which must be true to show this view')),
            type: 'string'
        },
        icon: {
            description: ( localize(
                2589,
                "Path to the view icon. View icons are displayed when the name of the view cannot be shown. It is recommended that icons be in SVG, though any image file type is accepted."
            )),
            type: 'string'
        },
        contextualTitle: {
            description: ( localize(
                2590,
                "Human-readable context for when the view is moved out of its original location. By default, the view's container name will be used."
            )),
            type: 'string'
        },
        visibility: {
            description: ( localize(
                2591,
                "Initial state of the view when the extension is first installed. Once the user has changed the view state by collapsing, moving, or hiding the view, the initial state will not be used again."
            )),
            type: 'string',
            enum: [
                'visible',
                'hidden',
                'collapsed'
            ],
            default: 'visible',
            enumDescriptions: [
                ( localize(
                2592,
                "The default initial state for the view. In most containers the view will be expanded, however; some built-in containers (explorer, scm, and debug) show all contributed views collapsed regardless of the `visibility`."
            )),
                ( localize(
                2593,
                "The view will not be shown in the view container, but will be discoverable through the views menu and other view entry points and can be un-hidden by the user."
            )),
                ( localize(2594, "The view will show in the view container, but will be collapsed."))
            ]
        },
        initialSize: {
            type: 'number',
            description: ( localize(
                2595,
                "The initial size of the view. The size will behave like the css 'flex' property, and will set the initial size when the view is first shown. In the side bar, this is the height of the view. This value is only respected when the same extension owns both the view and the view container."
            )),
        },
        accessibilityHelpContent: {
            type: 'string',
            markdownDescription: ( localize(
                2596,
                "When the accessibility help dialog is invoked in this view, this content will be presented to the user as a markdown string. Keybindings will be resolved when provided in the format of <keybinding:commandId>. If there is no keybinding, that will be indicated and this command will be included in a quickpick for easy configuration."
            ))
        }
    }
};
const remoteViewDescriptor = {
    type: 'object',
    required: ['id', 'name'],
    properties: {
        id: {
            description: ( localize(
                2586,
                'Identifier of the view. This should be unique across all views. It is recommended to include your extension id as part of the view id. Use this to register a data provider through `vscode.window.registerTreeDataProviderForView` API. Also to trigger activating your extension by registering `onView:${id}` event to `activationEvents`.'
            )),
            type: 'string'
        },
        name: {
            description: ( localize(2587, 'The human-readable name of the view. Will be shown')),
            type: 'string'
        },
        when: {
            description: ( localize(2588, 'Condition which must be true to show this view')),
            type: 'string'
        },
        group: {
            description: ( localize(2597, 'Nested group in the viewlet')),
            type: 'string'
        },
        remoteName: {
            description: ( localize(2598, 'The name of the remote type associated with this view')),
            type: ['string', 'array'],
            items: {
                type: 'string'
            }
        }
    }
};
const viewsContribution = {
    description: ( localize(2599, "Contributes views to the editor")),
    type: 'object',
    properties: {
        'explorer': {
            description: ( localize(2600, "Contributes views to Explorer container in the Activity bar")),
            type: 'array',
            items: viewDescriptor,
            default: []
        },
        'debug': {
            description: ( localize(2601, "Contributes views to Debug container in the Activity bar")),
            type: 'array',
            items: viewDescriptor,
            default: []
        },
        'scm': {
            description: ( localize(2602, "Contributes views to SCM container in the Activity bar")),
            type: 'array',
            items: viewDescriptor,
            default: []
        },
        'test': {
            description: ( localize(2603, "Contributes views to Test container in the Activity bar")),
            type: 'array',
            items: viewDescriptor,
            default: []
        },
        'remote': {
            description: ( localize(
                2604,
                "Contributes views to Remote container in the Activity bar. To contribute to this container, the 'contribViewsRemote' API proposal must be enabled."
            )),
            type: 'array',
            items: remoteViewDescriptor,
            default: []
        },
    },
    additionalProperties: {
        description: ( localize(2605, "Contributes views to contributed views container")),
        type: 'array',
        items: viewDescriptor,
        default: []
    }
};
const viewsContainersExtensionPoint = ExtensionsRegistry.registerExtensionPoint({
    extensionPoint: 'viewsContainers',
    jsonSchema: viewsContainersContribution
});
const viewsExtensionPoint = ExtensionsRegistry.registerExtensionPoint({
    extensionPoint: 'views',
    deps: [viewsContainersExtensionPoint],
    jsonSchema: viewsContribution,
    activationEventsGenerator: function* (viewExtensionPointTypeArray) {
        for (const viewExtensionPointType of viewExtensionPointTypeArray) {
            for (const viewDescriptors of ( Object.values(viewExtensionPointType))) {
                for (const viewDescriptor of viewDescriptors) {
                    if (viewDescriptor.id) {
                        yield `onView:${viewDescriptor.id}`;
                    }
                }
            }
        }
    }
});
const CUSTOM_VIEWS_START_ORDER = 7;
let ViewsExtensionHandler = class ViewsExtensionHandler {
    static { this.ID = 'workbench.contrib.viewsExtensionHandler'; }
    constructor(instantiationService, logService) {
        this.instantiationService = instantiationService;
        this.logService = logService;
        this.viewContainersRegistry = ( Registry.as(Extensions.ViewContainersRegistry));
        this.viewsRegistry = ( Registry.as(Extensions.ViewsRegistry));
        this.handleAndRegisterCustomViewContainers();
        this.handleAndRegisterCustomViews();
    }
    handleAndRegisterCustomViewContainers() {
        viewsContainersExtensionPoint.setHandler((extensions, { added, removed }) => {
            if (removed.length) {
                this.removeCustomViewContainers(removed);
            }
            if (added.length) {
                this.addCustomViewContainers(added, this.viewContainersRegistry.all);
            }
        });
    }
    addCustomViewContainers(extensionPoints, existingViewContainers) {
        const viewContainersRegistry = ( Registry.as(Extensions.ViewContainersRegistry));
        let activityBarOrder = CUSTOM_VIEWS_START_ORDER + viewContainersRegistry.all.filter(v => !!v.extensionId && viewContainersRegistry.getViewContainerLocation(v) === ViewContainerLocation.Sidebar).length;
        let panelOrder = 5 + viewContainersRegistry.all.filter(v => !!v.extensionId && viewContainersRegistry.getViewContainerLocation(v) === ViewContainerLocation.Panel).length + 1;
        let auxiliaryBarOrder = 100 + viewContainersRegistry.all.filter(v => !!v.extensionId && viewContainersRegistry.getViewContainerLocation(v) === ViewContainerLocation.AuxiliaryBar).length + 1;
        for (const { value, collector, description } of extensionPoints) {
            Object.entries(value).forEach(([key, value]) => {
                if (!this.isValidViewsContainer(value, collector)) {
                    return;
                }
                switch (key) {
                    case 'activitybar':
                        activityBarOrder = this.registerCustomViewContainers(value, description, activityBarOrder, existingViewContainers, ViewContainerLocation.Sidebar);
                        break;
                    case 'panel':
                        panelOrder = this.registerCustomViewContainers(value, description, panelOrder, existingViewContainers, ViewContainerLocation.Panel);
                        break;
                    case 'secondarySidebar':
                        auxiliaryBarOrder = this.registerCustomViewContainers(value, description, auxiliaryBarOrder, existingViewContainers, ViewContainerLocation.AuxiliaryBar);
                        break;
                }
            });
        }
    }
    removeCustomViewContainers(extensionPoints) {
        const viewContainersRegistry = ( Registry.as(Extensions.ViewContainersRegistry));
        const removedExtensions = extensionPoints.reduce((result, e) => { result.add(e.description.identifier); return result; }, ( new ExtensionIdentifierSet()));
        for (const viewContainer of viewContainersRegistry.all) {
            if (viewContainer.extensionId && ( removedExtensions.has(viewContainer.extensionId))) {
                const views = this.viewsRegistry.getViews(viewContainer);
                if (views.length) {
                    this.viewsRegistry.moveViews(views, this.getDefaultViewContainer());
                }
                this.deregisterCustomViewContainer(viewContainer);
            }
        }
    }
    isValidViewsContainer(viewsContainersDescriptors, collector) {
        if (!Array.isArray(viewsContainersDescriptors)) {
            collector.error(( localize(2606, "views containers must be an array")));
            return false;
        }
        for (const descriptor of viewsContainersDescriptors) {
            if (typeof descriptor.id !== 'string' && isFalsyOrWhitespace(descriptor.id)) {
                collector.error(( localize(
                    2607,
                    "property `{0}` is mandatory and must be of type `string` with non-empty value. Only alphanumeric characters, '_', and '-' are allowed.",
                    'id'
                )));
                return false;
            }
            if (!(/^[a-z0-9_-]+$/i.test(descriptor.id))) {
                collector.error(( localize(
                    2607,
                    "property `{0}` is mandatory and must be of type `string` with non-empty value. Only alphanumeric characters, '_', and '-' are allowed.",
                    'id'
                )));
                return false;
            }
            if (typeof descriptor.title !== 'string') {
                collector.error(( localize(2608, "property `{0}` is mandatory and must be of type `string`", 'title')));
                return false;
            }
            if (typeof descriptor.icon !== 'string') {
                collector.error(( localize(2608, "property `{0}` is mandatory and must be of type `string`", 'icon')));
                return false;
            }
            if (isFalsyOrWhitespace(descriptor.title)) {
                collector.warn(( localize(
                    2609,
                    "property `{0}` is mandatory and must be of type `string` with non-empty value",
                    'title'
                )));
                return true;
            }
        }
        return true;
    }
    registerCustomViewContainers(containers, extension, order, existingViewContainers, location) {
        containers.forEach(descriptor => {
            const themeIcon = ThemeIcon.fromString(descriptor.icon);
            const icon = themeIcon || joinPath(extension.extensionLocation, descriptor.icon);
            const id = `workbench.view.extension.${descriptor.id}`;
            const title = descriptor.title || id;
            const viewContainer = this.registerCustomViewContainer(id, title, icon, order++, extension.identifier, location);
            if (existingViewContainers.length) {
                const viewsToMove = [];
                for (const existingViewContainer of existingViewContainers) {
                    if (viewContainer !== existingViewContainer) {
                        viewsToMove.push(...this.viewsRegistry.getViews(existingViewContainer).filter(view => view.originalContainerId === descriptor.id));
                    }
                }
                if (viewsToMove.length) {
                    this.viewsRegistry.moveViews(viewsToMove, viewContainer);
                }
            }
        });
        return order;
    }
    registerCustomViewContainer(id, title, icon, order, extensionId, location) {
        let viewContainer = this.viewContainersRegistry.get(id);
        if (!viewContainer) {
            viewContainer = this.viewContainersRegistry.registerViewContainer({
                id,
                title: { value: title, original: title },
                extensionId,
                ctorDescriptor: ( new SyncDescriptor(ViewPaneContainer, [id, { mergeViewWithContainerWhenSingleView: true }])),
                hideIfEmpty: true,
                order,
                icon,
            }, location);
        }
        return viewContainer;
    }
    deregisterCustomViewContainer(viewContainer) {
        this.viewContainersRegistry.deregisterViewContainer(viewContainer);
        ( Registry.as(Extensions$1.Viewlets)).deregisterPaneComposite(viewContainer.id);
    }
    handleAndRegisterCustomViews() {
        viewsExtensionPoint.setHandler((extensions, { added, removed }) => {
            if (removed.length) {
                this.removeViews(removed);
            }
            if (added.length) {
                this.addViews(added);
            }
        });
    }
    addViews(extensions) {
        const viewIds = ( new Set());
        const allViewDescriptors = [];
        for (const extension of extensions) {
            const { value, collector } = extension;
            Object.entries(value).forEach(([key, value]) => {
                if (!this.isValidViewDescriptors(value, collector)) {
                    return;
                }
                if (key === 'remote' && !isProposedApiEnabled(extension.description, 'contribViewsRemote')) {
                    collector.warn(( localize(
                        2610,
                        "View container '{0}' requires 'enabledApiProposals: [\"contribViewsRemote\"]' to be added to 'Remote'.",
                        key
                    )));
                    return;
                }
                if (key === 'agentSessions' && !isProposedApiEnabled(extension.description, 'chatSessionsProvider')) {
                    collector.warn(( localize(
                        2611,
                        "View container '{0}' requires 'enabledApiProposals: [\"chatSessionsProvider\"]'.",
                        key
                    )));
                    return;
                }
                const viewContainer = this.getViewContainer(key);
                if (!viewContainer) {
                    collector.warn(( localize(
                        2612,
                        "View container '{0}' does not exist and all views registered to it will be added to 'Explorer'.",
                        key
                    )));
                }
                const container = viewContainer || this.getDefaultViewContainer();
                const viewDescriptors = [];
                for (let index = 0; index < value.length; index++) {
                    const item = value[index];
                    if (( viewIds.has(item.id))) {
                        collector.error(( localize(2613, "Cannot register multiple views with same id `{0}`", item.id)));
                        continue;
                    }
                    if (this.viewsRegistry.getView(item.id) !== null) {
                        collector.error(( localize(2614, "A view with id `{0}` is already registered.", item.id)));
                        continue;
                    }
                    const order = ExtensionIdentifier.equals(extension.description.identifier, container.extensionId)
                        ? index + 1
                        : container.viewOrderDelegate
                            ? container.viewOrderDelegate.getOrder(item.group)
                            : undefined;
                    let icon;
                    if (typeof item.icon === 'string') {
                        icon = ThemeIcon.fromString(item.icon) || joinPath(extension.description.extensionLocation, item.icon);
                    }
                    const initialVisibility = this.convertInitialVisibility(item.visibility);
                    const type = this.getViewType(item.type);
                    if (!type) {
                        collector.error(( localize(2615, "Unknown view type `{0}`.", item.type)));
                        continue;
                    }
                    let weight = undefined;
                    if (typeof item.initialSize === 'number') {
                        if (container.extensionId?.value === extension.description.identifier.value) {
                            weight = item.initialSize;
                        }
                        else {
                            this.logService.warn(`${extension.description.identifier.value} tried to set the view size of ${item.id} but it was ignored because the view container does not belong to it.`);
                        }
                    }
                    let accessibilityHelpContent;
                    if (isProposedApiEnabled(extension.description, 'contribAccessibilityHelpContent') && item.accessibilityHelpContent) {
                        accessibilityHelpContent = ( new MarkdownString(item.accessibilityHelpContent));
                    }
                    const viewDescriptor = {
                        type: type,
                        ctorDescriptor: type === ViewType.Tree ? ( new SyncDescriptor(TreeViewPane)) : ( new SyncDescriptor(WebviewViewPane)),
                        id: item.id,
                        name: { value: item.name, original: item.name },
                        when: ContextKeyExpr.deserialize(item.when),
                        containerIcon: icon || viewContainer?.icon,
                        containerTitle: item.contextualTitle || (viewContainer && (typeof viewContainer.title === 'string' ? viewContainer.title : viewContainer.title.value)),
                        canToggleVisibility: true,
                        canMoveView: viewContainer?.id !== VIEWLET_ID,
                        treeView: type === ViewType.Tree ? this.instantiationService.createInstance(CustomTreeView, item.id, item.name, extension.description.identifier.value) : undefined,
                        collapsed: this.showCollapsed(container) || initialVisibility === InitialVisibility.Collapsed,
                        order: order,
                        extensionId: extension.description.identifier,
                        originalContainerId: key,
                        group: item.group,
                        remoteAuthority: item.remoteName || item.remoteAuthority,
                        virtualWorkspace: item.virtualWorkspace,
                        hideByDefault: initialVisibility === InitialVisibility.Hidden,
                        workspace: viewContainer?.id === VIEWLET_ID ? true : undefined,
                        weight,
                        accessibilityHelpContent
                    };
                    viewIds.add(viewDescriptor.id);
                    viewDescriptors.push(viewDescriptor);
                }
                allViewDescriptors.push({ viewContainer: container, views: viewDescriptors });
            });
        }
        this.viewsRegistry.registerViews2(allViewDescriptors);
    }
    getViewType(type) {
        if (type === ViewType.Webview) {
            return ViewType.Webview;
        }
        if (!type || type === ViewType.Tree) {
            return ViewType.Tree;
        }
        return undefined;
    }
    getDefaultViewContainer() {
        return this.viewContainersRegistry.get(VIEWLET_ID$1);
    }
    removeViews(extensions) {
        const removedExtensions = extensions.reduce((result, e) => { result.add(e.description.identifier); return result; }, ( new ExtensionIdentifierSet()));
        for (const viewContainer of this.viewContainersRegistry.all) {
            const removedViews = this.viewsRegistry.getViews(viewContainer).filter(v => v.extensionId && ( removedExtensions.has(v.extensionId)));
            if (removedViews.length) {
                this.viewsRegistry.deregisterViews(removedViews, viewContainer);
                for (const view of removedViews) {
                    const anyView = view;
                    if (anyView.treeView) {
                        anyView.treeView.dispose();
                    }
                }
            }
        }
    }
    convertInitialVisibility(value) {
        if (( Object.values(InitialVisibility)).includes(value)) {
            return value;
        }
        return undefined;
    }
    isValidViewDescriptors(viewDescriptors, collector) {
        if (!Array.isArray(viewDescriptors)) {
            collector.error(( localize(2616, "views must be an array")));
            return false;
        }
        for (const descriptor of viewDescriptors) {
            if (typeof descriptor.id !== 'string') {
                collector.error(( localize(2608, "property `{0}` is mandatory and must be of type `string`", 'id')));
                return false;
            }
            if (typeof descriptor.name !== 'string') {
                collector.error(( localize(2608, "property `{0}` is mandatory and must be of type `string`", 'name')));
                return false;
            }
            if (descriptor.when && typeof descriptor.when !== 'string') {
                collector.error(( localize(2617, "property `{0}` can be omitted or must be of type `string`", 'when')));
                return false;
            }
            if (descriptor.icon && typeof descriptor.icon !== 'string') {
                collector.error(( localize(2617, "property `{0}` can be omitted or must be of type `string`", 'icon')));
                return false;
            }
            if (descriptor.contextualTitle && typeof descriptor.contextualTitle !== 'string') {
                collector.error(( localize(
                    2617,
                    "property `{0}` can be omitted or must be of type `string`",
                    'contextualTitle'
                )));
                return false;
            }
            if (descriptor.visibility && !this.convertInitialVisibility(descriptor.visibility)) {
                collector.error(( localize(
                    2618,
                    "property `{0}` can be omitted or must be one of {1}",
                    'visibility',
                    ( Object.values(InitialVisibility)).join(', ')
                )));
                return false;
            }
        }
        return true;
    }
    getViewContainer(value) {
        switch (value) {
            case 'explorer': return this.viewContainersRegistry.get(VIEWLET_ID$1);
            case 'debug': return this.viewContainersRegistry.get(VIEWLET_ID$3);
            case 'scm': return this.viewContainersRegistry.get(VIEWLET_ID$2);
            case 'remote': return this.viewContainersRegistry.get(VIEWLET_ID);
            default: return this.viewContainersRegistry.get(`workbench.view.extension.${value}`);
        }
    }
    showCollapsed(container) {
        switch (container.id) {
            case VIEWLET_ID$1:
            case VIEWLET_ID$2:
            case VIEWLET_ID$3:
                return true;
        }
        return false;
    }
};
ViewsExtensionHandler = ( __decorate([
    ( __param(0, IInstantiationService)),
    ( __param(1, ILogService))
], ViewsExtensionHandler));
class ViewContainersDataRenderer extends Disposable {
    constructor() {
        super(...arguments);
        this.type = 'table';
    }
    shouldRender(manifest) {
        return !!manifest.contributes?.viewsContainers;
    }
    render(manifest) {
        const contrib = manifest.contributes?.viewsContainers || {};
        const viewContainers = ( Object.keys(contrib)).reduce((result, location) => {
            const viewContainersForLocation = contrib[location];
            result.push(...( viewContainersForLocation.map(viewContainer => ({ ...viewContainer, location }))));
            return result;
        }, []);
        if (!viewContainers.length) {
            return { data: { headers: [], rows: [] }, dispose: () => { } };
        }
        const headers = [
            ( localize(2619, "ID")),
            ( localize(2620, "Title")),
            ( localize(2621, "Where")),
        ];
        const rows = ( viewContainers
            .sort((a, b) => a.id.localeCompare(b.id))
            .map(viewContainer => {
            return [
                viewContainer.id,
                viewContainer.title,
                viewContainer.location
            ];
        }));
        return {
            data: {
                headers,
                rows
            },
            dispose: () => { }
        };
    }
}
class ViewsDataRenderer extends Disposable {
    constructor() {
        super(...arguments);
        this.type = 'table';
    }
    shouldRender(manifest) {
        return !!manifest.contributes?.views;
    }
    render(manifest) {
        const contrib = manifest.contributes?.views || {};
        const views = ( Object.keys(contrib)).reduce((result, location) => {
            const viewsForLocation = contrib[location];
            result.push(...( viewsForLocation.map(view => ({ ...view, location }))));
            return result;
        }, []);
        if (!views.length) {
            return { data: { headers: [], rows: [] }, dispose: () => { } };
        }
        const headers = [
            ( localize(2622, "ID")),
            ( localize(2623, "Name")),
            ( localize(2621, "Where")),
        ];
        const rows = ( views
            .sort((a, b) => a.id.localeCompare(b.id))
            .map(view => {
            return [
                view.id,
                view.name,
                view.location
            ];
        }));
        return {
            data: {
                headers,
                rows
            },
            dispose: () => { }
        };
    }
}
( Registry.as(Extensions$2.ExtensionFeaturesRegistry)).registerExtensionFeature({
    id: 'viewsContainers',
    label: ( localize(2624, "View Containers")),
    access: {
        canToggle: false
    },
    renderer: ( new SyncDescriptor(ViewContainersDataRenderer)),
});
( Registry.as(Extensions$2.ExtensionFeaturesRegistry)).registerExtensionFeature({
    id: 'views',
    label: ( localize(2625, "Views")),
    access: {
        canToggle: false
    },
    renderer: ( new SyncDescriptor(ViewsDataRenderer)),
});
registerWorkbenchContribution2(ViewsExtensionHandler.ID, ViewsExtensionHandler, WorkbenchPhase.BlockStartup);

export { viewsContainersContribution };
