// @codingame/monaco-vscode-api/vscode/src/vs/workbench/browser/parts/views/viewMenuActions.js

import { __decorate, __param } from '../../../../../../../external/tslib/tslib.es6.js';
import { Emitter, Event } from '../../../../base/common/event.js';
import { Disposable } from '../../../../base/common/lifecycle.js';
import { getActionBarActions } from '../../../../platform/actions/browser/menuEntryActionViewItem.js';
import { MenuId } from '../../../../platform/actions/common/actions.js';
import { IMenuService } from '../../../../platform/actions/common/actions.service.js';
import { IContextKeyService } from '../../../../platform/contextkey/common/contextkey.service.js';
import { ViewContainerLocationToString } from '../../../common/views.js';
import { IViewDescriptorService } from '../../../common/views.service.js';

let ViewMenuActions = class ViewMenuActions extends Disposable {
    constructor(menuId, contextMenuId, options, contextKeyService, menuService) {
        super();
        this.menuId = menuId;
        this.contextMenuId = contextMenuId;
        this.options = options;
        this.contextKeyService = contextKeyService;
        this.menuService = menuService;
        this._onDidChange = this._register(( new Emitter()));
        this.onDidChange = this._onDidChange.event;
        this.menu = this._register(menuService.createMenu(menuId, contextKeyService, { emitEventsForSubmenuChanges: true }));
        this._register(this.menu.onDidChange(() => {
            this.actions = undefined;
            this._onDidChange.fire();
        }));
    }
    getActions() {
        if (!this.actions) {
            this.actions = getActionBarActions(this.menu.getActions(this.options));
        }
        return this.actions;
    }
    getPrimaryActions() {
        return this.getActions().primary;
    }
    getSecondaryActions() {
        return this.getActions().secondary;
    }
    getContextMenuActions() {
        if (this.contextMenuId) {
            const menu = this.menuService.getMenuActions(this.contextMenuId, this.contextKeyService, this.options);
            return getActionBarActions(menu).secondary;
        }
        return [];
    }
};
ViewMenuActions = ( __decorate([
    ( __param(3, IContextKeyService)),
    ( __param(4, IMenuService))
], ViewMenuActions));
let ViewContainerMenuActions = class ViewContainerMenuActions extends ViewMenuActions {
    constructor(element, viewContainer, viewDescriptorService, contextKeyService, menuService) {
        const scopedContextKeyService = contextKeyService.createScoped(element);
        scopedContextKeyService.createKey('viewContainer', viewContainer.id);
        const viewContainerLocationKey = scopedContextKeyService.createKey('viewContainerLocation', ViewContainerLocationToString(viewDescriptorService.getViewContainerLocation(viewContainer)));
        super(MenuId.ViewContainerTitle, MenuId.ViewContainerTitleContext, { shouldForwardArgs: true, renderShortTitle: true }, scopedContextKeyService, menuService);
        this._register(scopedContextKeyService);
        this._register(Event.filter(viewDescriptorService.onDidChangeContainerLocation, e => e.viewContainer === viewContainer)(() => viewContainerLocationKey.set(ViewContainerLocationToString(viewDescriptorService.getViewContainerLocation(viewContainer)))));
    }
};
ViewContainerMenuActions = ( __decorate([
    ( __param(2, IViewDescriptorService)),
    ( __param(3, IContextKeyService)),
    ( __param(4, IMenuService))
], ViewContainerMenuActions));

export { ViewContainerMenuActions, ViewMenuActions };
