// @codingame/monaco-vscode-keybindings-service-override/vscode/src/vs/platform/keyboardLayout/common/keyboardConfig.js

import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { OS, OperatingSystem } from '@codingame/monaco-vscode-api/vscode/vs/base/common/platform';
import { Extensions, ConfigurationScope } from '@codingame/monaco-vscode-api/vscode/vs/platform/configuration/common/configurationRegistry';
import { Registry } from '@codingame/monaco-vscode-api/vscode/vs/platform/registry/common/platform';

var DispatchConfig;
(function (DispatchConfig) {
    DispatchConfig[DispatchConfig["Code"] = 0] = "Code";
    DispatchConfig[DispatchConfig["KeyCode"] = 1] = "KeyCode";
})(DispatchConfig || (DispatchConfig = {}));
function readKeyboardConfig(configurationService) {
    const keyboard = configurationService.getValue('keyboard');
    const dispatch = (keyboard?.dispatch === 'keyCode' ? DispatchConfig.KeyCode : DispatchConfig.Code);
    const mapAltGrToCtrlAlt = Boolean(keyboard?.mapAltGrToCtrlAlt);
    return { dispatch, mapAltGrToCtrlAlt };
}
const configurationRegistry = ( Registry.as(Extensions.Configuration));
const keyboardConfiguration = {
    'id': 'keyboard',
    'order': 15,
    'type': 'object',
    'title': ( localize(1952, "Keyboard")),
    'properties': {
        'keyboard.dispatch': {
            scope: ConfigurationScope.APPLICATION,
            type: 'string',
            enum: ['code', 'keyCode'],
            default: 'code',
            markdownDescription: ( localize(
                1953,
                "Controls the dispatching logic for key presses to use either `code` (recommended) or `keyCode`."
            )),
            included: OS === OperatingSystem.Macintosh || OS === OperatingSystem.Linux
        },
        'keyboard.mapAltGrToCtrlAlt': {
            scope: ConfigurationScope.APPLICATION,
            type: 'boolean',
            default: false,
            markdownDescription: ( localize(1954, "Controls if the AltGraph+ modifier should be treated as Ctrl+Alt+.")),
            included: OS === OperatingSystem.Windows
        }
    }
};
configurationRegistry.registerConfiguration(keyboardConfiguration);

export { DispatchConfig, readKeyboardConfig };
