// @codingame/monaco-vscode-view-banner-service-override/vscode/src/vs/workbench/browser/parts/banner/bannerPart.js

import { registerCss } from '@codingame/monaco-vscode-api/css';
import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import * as bannerpart from './media/bannerpart.css';
import { localize, localize2 } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { addDisposableListener, EventType, clearNode, isHTMLElement, $, append } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/dom';
import { asCSSUrl } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/cssValue';
import { ActionBar } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/ui/actionbar/actionbar';
import '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/extensions';
import { IInstantiationService } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/instantiation';
import { IStorageService } from '@codingame/monaco-vscode-api/vscode/vs/platform/storage/common/storage.service';
import { IThemeService } from '@codingame/monaco-vscode-api/vscode/vs/platform/theme/common/themeService.service';
import { ThemeIcon } from '@codingame/monaco-vscode-api/vscode/vs/base/common/themables';
import { Part } from '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/part';
import { Parts } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/layout/browser/layoutService';
import { IWorkbenchLayoutService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/layout/browser/layoutService.service';
import { Action } from '@codingame/monaco-vscode-api/vscode/vs/base/common/actions';
import { Link } from '@codingame/monaco-vscode-api/vscode/vs/platform/opener/browser/link';
import { Emitter } from '@codingame/monaco-vscode-api/vscode/vs/base/common/event';
import { IBannerService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/banner/browser/bannerService.service';
import { IMarkdownRendererService } from '@codingame/monaco-vscode-api/vscode/vs/platform/markdown/browser/markdownRenderer.service';
import { Action2, registerAction2 } from '@codingame/monaco-vscode-api/vscode/vs/platform/actions/common/actions';
import { Categories } from '@codingame/monaco-vscode-api/vscode/vs/platform/action/common/actionCommonCategories';
import { KeybindingsRegistry, KeybindingWeight } from '@codingame/monaco-vscode-api/vscode/vs/platform/keybinding/common/keybindingsRegistry';
import { KeyCode } from '@codingame/monaco-vscode-api/vscode/vs/base/common/keyCodes';
import { IContextKeyService } from '@codingame/monaco-vscode-api/vscode/vs/platform/contextkey/common/contextkey.service';
import { URI } from '@codingame/monaco-vscode-api/vscode/vs/base/common/uri';
import { widgetClose } from '@codingame/monaco-vscode-api/vscode/vs/platform/theme/common/iconRegistry';
import { BannerFocused } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/contextkeys';

registerCss(bannerpart);
let BannerPart = class BannerPart extends Part {
    get minimumHeight() {
        return this.visible ? this.height : 0;
    }
    get maximumHeight() {
        return this.visible ? this.height : 0;
    }
    get onDidChange() { return this._onDidChangeSize.event; }
    constructor(themeService, layoutService, storageService, contextKeyService, instantiationService, markdownRendererService) {
        super(Parts.BANNER_PART, { hasTitle: false }, themeService, storageService, layoutService);
        this.contextKeyService = contextKeyService;
        this.instantiationService = instantiationService;
        this.markdownRendererService = markdownRendererService;
        this.height = 26;
        this.minimumWidth = 0;
        this.maximumWidth = Number.POSITIVE_INFINITY;
        this._onDidChangeSize = this._register(( new Emitter()));
        this.visible = false;
        this.focusedActionIndex = -1;
    }
    createContentArea(parent) {
        this.element = parent;
        this.element.tabIndex = 0;
        this._register(addDisposableListener(this.element, EventType.FOCUS, () => {
            if (this.focusedActionIndex !== -1) {
                this.focusActionLink();
            }
        }));
        const scopedContextKeyService = this._register(this.contextKeyService.createScoped(this.element));
        BannerFocused.bindTo(scopedContextKeyService).set(true);
        return this.element;
    }
    close(item) {
        this.setVisibility(false);
        clearNode(this.element);
        if (typeof item.onClose === 'function') {
            item.onClose();
        }
        this.item = undefined;
    }
    focusActionLink() {
        const length = this.item?.actions?.length ?? 0;
        if (this.focusedActionIndex < length) {
            const actionLink = this.messageActionsContainer?.children[this.focusedActionIndex];
            if (isHTMLElement(actionLink)) {
                this.actionBar?.setFocusable(false);
                actionLink.focus();
            }
        }
        else {
            this.actionBar?.focus(0);
        }
    }
    getAriaLabel(item) {
        if (item.ariaLabel) {
            return item.ariaLabel;
        }
        if (typeof item.message === 'string') {
            return item.message;
        }
        return undefined;
    }
    getBannerMessage(message) {
        if (typeof message === 'string') {
            const element = $('span');
            element.textContent = message;
            return element;
        }
        return this.markdownRendererService.render(message).element;
    }
    setVisibility(visible) {
        if (visible !== this.visible) {
            this.visible = visible;
            this.focusedActionIndex = -1;
            this.layoutService.setPartHidden(!visible, Parts.BANNER_PART);
            this._onDidChangeSize.fire(undefined);
        }
    }
    focus() {
        this.focusedActionIndex = -1;
        this.element.focus();
    }
    focusNextAction() {
        const length = this.item?.actions?.length ?? 0;
        this.focusedActionIndex = this.focusedActionIndex < length ? this.focusedActionIndex + 1 : 0;
        this.focusActionLink();
    }
    focusPreviousAction() {
        const length = this.item?.actions?.length ?? 0;
        this.focusedActionIndex = this.focusedActionIndex > 0 ? this.focusedActionIndex - 1 : length;
        this.focusActionLink();
    }
    hide(id) {
        if (this.item?.id !== id) {
            return;
        }
        this.setVisibility(false);
    }
    show(item) {
        if (item.id === this.item?.id) {
            this.setVisibility(true);
            return;
        }
        clearNode(this.element);
        const ariaLabel = this.getAriaLabel(item);
        if (ariaLabel) {
            this.element.setAttribute('aria-label', ariaLabel);
        }
        const iconContainer = append(this.element, $('div.icon-container'));
        iconContainer.setAttribute('aria-hidden', 'true');
        if (ThemeIcon.isThemeIcon(item.icon)) {
            iconContainer.appendChild($(`div${ThemeIcon.asCSSSelector(item.icon)}`));
        }
        else {
            iconContainer.classList.add('custom-icon');
            if (URI.isUri(item.icon)) {
                iconContainer.style.backgroundImage = asCSSUrl(item.icon);
            }
        }
        const messageContainer = append(this.element, $('div.message-container'));
        messageContainer.setAttribute('aria-hidden', 'true');
        messageContainer.appendChild(this.getBannerMessage(item.message));
        this.messageActionsContainer = append(this.element, $('div.message-actions-container'));
        if (item.actions) {
            for (const action of item.actions) {
                this._register(this.instantiationService.createInstance(Link, this.messageActionsContainer, { ...action, tabIndex: -1 }, {}));
            }
        }
        const actionBarContainer = append(this.element, $('div.action-container'));
        this.actionBar = this._register(( new ActionBar(actionBarContainer)));
        const label = item.closeLabel ?? ( localize(3032, "Close Banner"));
        const closeAction = this._register(( new Action(
            'banner.close',
            label,
            ThemeIcon.asClassName(widgetClose),
            true,
            () => this.close(item)
        )));
        this.actionBar.push(closeAction, { icon: true, label: false });
        this.actionBar.setFocusable(false);
        this.setVisibility(true);
        this.item = item;
    }
    toJSON() {
        return {
            type: Parts.BANNER_PART
        };
    }
};
BannerPart = ( __decorate([
    ( __param(0, IThemeService)),
    ( __param(1, IWorkbenchLayoutService)),
    ( __param(2, IStorageService)),
    ( __param(3, IContextKeyService)),
    ( __param(4, IInstantiationService)),
    ( __param(5, IMarkdownRendererService))
], BannerPart));
KeybindingsRegistry.registerCommandAndKeybindingRule({
    id: 'workbench.banner.focusBanner',
    weight: KeybindingWeight.WorkbenchContrib,
    primary: KeyCode.Escape,
    when: BannerFocused,
    handler: (accessor) => {
        const bannerService = accessor.get(IBannerService);
        bannerService.focus();
    }
});
KeybindingsRegistry.registerCommandAndKeybindingRule({
    id: 'workbench.banner.focusNextAction',
    weight: KeybindingWeight.WorkbenchContrib,
    primary: KeyCode.RightArrow,
    secondary: [KeyCode.DownArrow],
    when: BannerFocused,
    handler: (accessor) => {
        const bannerService = accessor.get(IBannerService);
        bannerService.focusNextAction();
    }
});
KeybindingsRegistry.registerCommandAndKeybindingRule({
    id: 'workbench.banner.focusPreviousAction',
    weight: KeybindingWeight.WorkbenchContrib,
    primary: KeyCode.LeftArrow,
    secondary: [KeyCode.UpArrow],
    when: BannerFocused,
    handler: (accessor) => {
        const bannerService = accessor.get(IBannerService);
        bannerService.focusPreviousAction();
    }
});
class FocusBannerAction extends Action2 {
    static { this.ID = 'workbench.action.focusBanner'; }
    static { this.LABEL = ( localize2(3033, "Focus Banner")); }
    constructor() {
        super({
            id: FocusBannerAction.ID,
            title: FocusBannerAction.LABEL,
            category: Categories.View,
            f1: true
        });
    }
    async run(accessor) {
        const layoutService = accessor.get(IWorkbenchLayoutService);
        layoutService.focusPart(Parts.BANNER_PART);
    }
}
registerAction2(FocusBannerAction);

export { BannerPart };
