// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminal/common/terminalStrings.js

import { localize2, localize } from '../../../../nls.js';

const terminalStrings = {
    terminal: ( localize(11913, "Terminal")),
    new: ( localize(11914, "New Terminal")),
    currentSessionCategory: ( localize(11915, 'current session')),
    previousSessionCategory: ( localize(11916, 'previous session')),
    typeTask: ( localize(11917, "Task")),
    typeLocal: ( localize(11918, "Local")),
    actionCategory: ( localize2(11919, "Terminal")),
    focus: ( localize2(11920, "Focus Terminal")),
    kill: {
        ...( localize2(11921, "Kill Terminal")),
        short: ( localize(11922, "Kill")),
    },
    moveToEditor: ( localize2(11923, "Move Terminal into Editor Area")),
    moveIntoNewWindow: ( localize2(11924, "Move Terminal into New Window")),
    newInNewWindow: ( localize2(11925, "New Terminal Window")),
    moveToTerminalPanel: ( localize2(11926, "Move Terminal into Panel")),
    changeIcon: ( localize2(11927, "Change Icon...")),
    changeColor: ( localize2(11928, "Change Color...")),
    split: {
        ...( localize2(11929, "Split Terminal")),
        short: ( localize(11930, "Split")),
    },
    unsplit: ( localize2(11931, "Unsplit Terminal")),
    rename: ( localize2(11932, "Rename...")),
    toggleSizeToContentWidth: ( localize2(11933, "Toggle Size to Content Width")),
    focusHover: ( localize2(11934, "Focus Hover")),
    newWithCwd: ( localize2(11935, "Create New Terminal Starting in a Custom Working Directory")),
    renameWithArgs: ( localize2(11936, "Rename the Currently Active Terminal")),
    scrollToPreviousCommand: ( localize2(11937, "Scroll to Previous Command")),
    scrollToNextCommand: ( localize2(11938, "Scroll to Next Command"))};

export { terminalStrings };
