// @codingame/monaco-vscode-keybindings-service-override/vscode/src/vs/workbench/services/actions/common/menusExtensionPoint.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { isFalsyOrWhitespace } from '@codingame/monaco-vscode-api/vscode/vs/base/common/strings';
import { joinPath } from '@codingame/monaco-vscode-api/vscode/vs/base/common/resources';
import { ExtensionsRegistry } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/extensions/common/extensionsRegistry';
import { ContextKeyExpr } from '@codingame/monaco-vscode-api/vscode/vs/platform/contextkey/common/contextkey';
import { MenuId, MenuRegistry } from '@codingame/monaco-vscode-api/vscode/vs/platform/actions/common/actions';
import { DisposableStore, Disposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import { ThemeIcon } from '@codingame/monaco-vscode-api/vscode/vs/base/common/themables';
import { index } from '@codingame/monaco-vscode-api/vscode/vs/base/common/arrays';
import { isProposedApiEnabled } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/extensions/common/extensions';
import { Extensions } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/extensionManagement/common/extensionFeatures';
import { Registry } from '@codingame/monaco-vscode-api/vscode/vs/platform/registry/common/platform';
import { SyncDescriptor } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/descriptors';
import { platform } from '@codingame/monaco-vscode-api/vscode/vs/base/common/process';
import { MarkdownString } from '@codingame/monaco-vscode-api/vscode/vs/base/common/htmlContent';
import { IKeybindingService } from '@codingame/monaco-vscode-api/vscode/vs/platform/keybinding/common/keybinding.service';

const apiMenus = [
    {
        key: 'commandPalette',
        id: MenuId.CommandPalette,
        description: ( localize(13586, "The Command Palette")),
        supportsSubmenus: false
    },
    {
        key: 'touchBar',
        id: MenuId.TouchBarContext,
        description: ( localize(13587, "The touch bar (macOS only)")),
        supportsSubmenus: false
    },
    {
        key: 'editor/title',
        id: MenuId.EditorTitle,
        description: ( localize(13588, "The editor title menu"))
    },
    {
        key: 'editor/title/run',
        id: MenuId.EditorTitleRun,
        description: ( localize(13589, "Run submenu inside the editor title menu"))
    },
    {
        key: 'editor/context',
        id: MenuId.EditorContext,
        description: ( localize(13590, "The editor context menu"))
    },
    {
        key: 'editor/context/copy',
        id: MenuId.EditorContextCopy,
        description: ( localize(13591, "'Copy as' submenu in the editor context menu"))
    },
    {
        key: 'editor/context/share',
        id: MenuId.EditorContextShare,
        description: ( localize(13592, "'Share' submenu in the editor context menu")),
        proposed: 'contribShareMenu'
    },
    {
        key: 'explorer/context',
        id: MenuId.ExplorerContext,
        description: ( localize(13593, "The file explorer context menu"))
    },
    {
        key: 'explorer/context/share',
        id: MenuId.ExplorerContextShare,
        description: ( localize(13594, "'Share' submenu in the file explorer context menu")),
        proposed: 'contribShareMenu'
    },
    {
        key: 'editor/title/context',
        id: MenuId.EditorTitleContext,
        description: ( localize(13595, "The editor tabs context menu"))
    },
    {
        key: 'editor/title/context/share',
        id: MenuId.EditorTitleContextShare,
        description: ( localize(13596, "'Share' submenu inside the editor title context menu")),
        proposed: 'contribShareMenu'
    },
    {
        key: 'debug/callstack/context',
        id: MenuId.DebugCallStackContext,
        description: ( localize(13597, "The debug callstack view context menu"))
    },
    {
        key: 'debug/variables/context',
        id: MenuId.DebugVariablesContext,
        description: ( localize(13598, "The debug variables view context menu"))
    },
    {
        key: 'debug/watch/context',
        id: MenuId.DebugWatchContext,
        description: ( localize(13599, "The debug watch view context menu"))
    },
    {
        key: 'debug/toolBar',
        id: MenuId.DebugToolBar,
        description: ( localize(13600, "The debug toolbar menu"))
    },
    {
        key: 'debug/createConfiguration',
        id: MenuId.DebugCreateConfiguration,
        proposed: 'contribDebugCreateConfiguration',
        description: ( localize(13601, "The debug create configuration menu"))
    },
    {
        key: 'notebook/variables/context',
        id: MenuId.NotebookVariablesContext,
        description: ( localize(13602, "The notebook variables view context menu"))
    },
    {
        key: 'menuBar/home',
        id: MenuId.MenubarHomeMenu,
        description: ( localize(13603, "The home indicator context menu (web only)")),
        proposed: 'contribMenuBarHome',
        supportsSubmenus: false
    },
    {
        key: 'menuBar/edit/copy',
        id: MenuId.MenubarCopy,
        description: ( localize(13604, "'Copy as' submenu in the top level Edit menu"))
    },
    {
        key: 'scm/title',
        id: MenuId.SCMTitle,
        description: ( localize(13605, "The Source Control title menu"))
    },
    {
        key: 'scm/sourceControl',
        id: MenuId.SCMSourceControl,
        description: ( localize(13606, "The Source Control menu"))
    },
    {
        key: 'scm/repositories/title',
        id: MenuId.SCMSourceControlTitle,
        description: ( localize(13607, "The Source Control Repositories title menu")),
        proposed: 'contribSourceControlTitleMenu'
    },
    {
        key: 'scm/repository',
        id: MenuId.SCMSourceControlInline,
        description: ( localize(13608, "The Source Control repository menu")),
    },
    {
        key: 'scm/resourceState/context',
        id: MenuId.SCMResourceContext,
        description: ( localize(13609, "The Source Control resource state context menu"))
    },
    {
        key: 'scm/resourceFolder/context',
        id: MenuId.SCMResourceFolderContext,
        description: ( localize(13610, "The Source Control resource folder context menu"))
    },
    {
        key: 'scm/resourceGroup/context',
        id: MenuId.SCMResourceGroupContext,
        description: ( localize(13611, "The Source Control resource group context menu"))
    },
    {
        key: 'scm/change/title',
        id: MenuId.SCMChangeContext,
        description: ( localize(13612, "The Source Control inline change menu"))
    },
    {
        key: 'scm/inputBox',
        id: MenuId.SCMInputBox,
        description: ( localize(13613, "The Source Control input box menu")),
        proposed: 'contribSourceControlInputBoxMenu'
    },
    {
        key: 'scm/history/title',
        id: MenuId.SCMHistoryTitle,
        description: ( localize(13614, "The Source Control History title menu")),
        proposed: 'contribSourceControlHistoryTitleMenu'
    },
    {
        key: 'scm/historyItem/context',
        id: MenuId.SCMHistoryItemContext,
        description: ( localize(13615, "The Source Control history item context menu")),
        proposed: 'contribSourceControlHistoryItemMenu'
    },
    {
        key: 'scm/historyItemRef/context',
        id: MenuId.SCMHistoryItemRefContext,
        description: ( localize(13616, "The Source Control history item reference context menu")),
        proposed: 'contribSourceControlHistoryItemMenu'
    },
    {
        key: 'scm/artifactGroup/context',
        id: MenuId.SCMArtifactGroupContext,
        description: ( localize(13617, "The Source Control artifact group context menu")),
        proposed: 'contribSourceControlArtifactGroupMenu'
    },
    {
        key: 'scm/artifact/context',
        id: MenuId.SCMArtifactContext,
        description: ( localize(13618, "The Source Control artifact context menu")),
        proposed: 'contribSourceControlArtifactMenu'
    },
    {
        key: 'statusBar/remoteIndicator',
        id: MenuId.StatusBarRemoteIndicatorMenu,
        description: ( localize(13619, "The remote indicator menu in the status bar")),
        supportsSubmenus: false
    },
    {
        key: 'terminal/context',
        id: MenuId.TerminalInstanceContext,
        description: ( localize(13620, "The terminal context menu"))
    },
    {
        key: 'terminal/title/context',
        id: MenuId.TerminalTabContext,
        description: ( localize(13621, "The terminal tabs context menu"))
    },
    {
        key: 'view/title',
        id: MenuId.ViewTitle,
        description: ( localize(13622, "The contributed view title menu"))
    },
    {
        key: 'viewContainer/title',
        id: MenuId.ViewContainerTitle,
        description: ( localize(13623, "The contributed view container title menu")),
        proposed: 'contribViewContainerTitle'
    },
    {
        key: 'view/item/context',
        id: MenuId.ViewItemContext,
        description: ( localize(13624, "The contributed view item context menu"))
    },
    {
        key: 'comments/comment/editorActions',
        id: MenuId.CommentEditorActions,
        description: ( localize(13625, "The contributed comment editor actions")),
        proposed: 'contribCommentEditorActionsMenu'
    },
    {
        key: 'comments/commentThread/title',
        id: MenuId.CommentThreadTitle,
        description: ( localize(13626, "The contributed comment thread title menu"))
    },
    {
        key: 'comments/commentThread/context',
        id: MenuId.CommentThreadActions,
        description: ( localize(
            13627,
            "The contributed comment thread context menu, rendered as buttons below the comment editor"
        )),
        supportsSubmenus: false
    },
    {
        key: 'comments/commentThread/additionalActions',
        id: MenuId.CommentThreadAdditionalActions,
        description: ( localize(
            13627,
            "The contributed comment thread context menu, rendered as buttons below the comment editor"
        )),
        supportsSubmenus: true,
        proposed: 'contribCommentThreadAdditionalMenu'
    },
    {
        key: 'comments/commentThread/title/context',
        id: MenuId.CommentThreadTitleContext,
        description: ( localize(
            13628,
            "The contributed comment thread title's peek context menu, rendered as a right click menu on the comment thread's peek title."
        )),
        proposed: 'contribCommentPeekContext'
    },
    {
        key: 'comments/comment/title',
        id: MenuId.CommentTitle,
        description: ( localize(13629, "The contributed comment title menu"))
    },
    {
        key: 'comments/comment/context',
        id: MenuId.CommentActions,
        description: ( localize(
            13630,
            "The contributed comment context menu, rendered as buttons below the comment editor"
        )),
        supportsSubmenus: false
    },
    {
        key: 'comments/commentThread/comment/context',
        id: MenuId.CommentThreadCommentContext,
        description: ( localize(
            13631,
            "The contributed comment context menu, rendered as a right click menu on the an individual comment in the comment thread's peek view."
        )),
        proposed: 'contribCommentPeekContext'
    },
    {
        key: 'commentsView/commentThread/context',
        id: MenuId.CommentsViewThreadActions,
        description: ( localize(13632, "The contributed comment thread context menu in the comments view")),
        proposed: 'contribCommentsViewThreadMenus'
    },
    {
        key: 'notebook/toolbar',
        id: MenuId.NotebookToolbar,
        description: ( localize(13633, "The contributed notebook toolbar menu"))
    },
    {
        key: 'notebook/kernelSource',
        id: MenuId.NotebookKernelSource,
        description: ( localize(13634, "The contributed notebook kernel sources menu")),
        proposed: 'notebookKernelSource'
    },
    {
        key: 'notebook/cell/title',
        id: MenuId.NotebookCellTitle,
        description: ( localize(13635, "The contributed notebook cell title menu"))
    },
    {
        key: 'notebook/cell/execute',
        id: MenuId.NotebookCellExecute,
        description: ( localize(13636, "The contributed notebook cell execution menu"))
    },
    {
        key: 'interactive/toolbar',
        id: MenuId.InteractiveToolbar,
        description: ( localize(13637, "The contributed interactive toolbar menu")),
    },
    {
        key: 'interactive/cell/title',
        id: MenuId.InteractiveCellTitle,
        description: ( localize(13638, "The contributed interactive cell title menu")),
    },
    {
        key: 'issue/reporter',
        id: MenuId.IssueReporter,
        description: ( localize(13639, "The contributed issue reporter menu"))
    },
    {
        key: 'testing/item/context',
        id: MenuId.TestItem,
        description: ( localize(13640, "The contributed test item menu")),
    },
    {
        key: 'testing/item/gutter',
        id: MenuId.TestItemGutter,
        description: ( localize(13641, "The menu for a gutter decoration for a test item")),
    },
    {
        key: 'testing/profiles/context',
        id: MenuId.TestProfilesContext,
        description: ( localize(13642, "The menu for configuring testing profiles.")),
    },
    {
        key: 'testing/item/result',
        id: MenuId.TestPeekElement,
        description: ( localize(13643, "The menu for an item in the Test Results view or peek.")),
    },
    {
        key: 'testing/message/context',
        id: MenuId.TestMessageContext,
        description: ( localize(
            13644,
            "A prominent button overlaying editor content where the message is displayed"
        )),
    },
    {
        key: 'testing/message/content',
        id: MenuId.TestMessageContent,
        description: ( localize(13645, "Context menu for the message in the results tree")),
    },
    {
        key: 'extension/context',
        id: MenuId.ExtensionContext,
        description: ( localize(13646, "The extension context menu"))
    },
    {
        key: 'timeline/title',
        id: MenuId.TimelineTitle,
        description: ( localize(13647, "The Timeline view title menu"))
    },
    {
        key: 'timeline/item/context',
        id: MenuId.TimelineItemContext,
        description: ( localize(13648, "The Timeline view item context menu"))
    },
    {
        key: 'ports/item/context',
        id: MenuId.TunnelContext,
        description: ( localize(13649, "The Ports view item context menu"))
    },
    {
        key: 'ports/item/origin/inline',
        id: MenuId.TunnelOriginInline,
        description: ( localize(13650, "The Ports view item origin inline menu"))
    },
    {
        key: 'ports/item/port/inline',
        id: MenuId.TunnelPortInline,
        description: ( localize(13651, "The Ports view item port inline menu"))
    },
    {
        key: 'file/newFile',
        id: MenuId.NewFile,
        description: ( localize(
            13652,
            "The 'New File...' quick pick, shown on welcome page and File menu."
        )),
        supportsSubmenus: false,
    },
    {
        key: 'webview/context',
        id: MenuId.WebviewContext,
        description: ( localize(13653, "The webview context menu"))
    },
    {
        key: 'file/share',
        id: MenuId.MenubarShare,
        description: ( localize(13654, "Share submenu shown in the top level File menu.")),
        proposed: 'contribShareMenu'
    },
    {
        key: 'editor/inlineCompletions/actions',
        id: MenuId.InlineCompletionsActions,
        description: ( localize(13655, "The actions shown when hovering on an inline completion")),
        supportsSubmenus: false,
        proposed: 'inlineCompletionsAdditions'
    },
    {
        key: 'editor/content',
        id: MenuId.EditorContent,
        description: ( localize(13656, "The prominent button in an editor, overlays its content")),
        proposed: 'contribEditorContentMenu'
    },
    {
        key: 'editor/lineNumber/context',
        id: MenuId.EditorLineNumberContext,
        description: ( localize(13657, "The contributed editor line number context menu"))
    },
    {
        key: 'mergeEditor/result/title',
        id: MenuId.MergeInputResultToolbar,
        description: ( localize(13658, "The result toolbar of the merge editor")),
        proposed: 'contribMergeEditorMenus'
    },
    {
        key: 'multiDiffEditor/content',
        id: MenuId.MultiDiffEditorContent,
        description: ( localize(13659, "A prominent button overlaying the multi diff editor")),
        proposed: 'contribEditorContentMenu'
    },
    {
        key: 'multiDiffEditor/resource/title',
        id: MenuId.MultiDiffEditorFileToolbar,
        description: ( localize(13660, "The resource toolbar in the multi diff editor")),
        proposed: 'contribMultiDiffEditorMenus'
    },
    {
        key: 'diffEditor/gutter/hunk',
        id: MenuId.DiffEditorHunkToolbar,
        description: ( localize(13661, "The gutter toolbar in the diff editor")),
        proposed: 'contribDiffEditorGutterToolBarMenus'
    },
    {
        key: 'diffEditor/gutter/selection',
        id: MenuId.DiffEditorSelectionToolbar,
        description: ( localize(13661, "The gutter toolbar in the diff editor")),
        proposed: 'contribDiffEditorGutterToolBarMenus'
    },
    {
        key: 'searchPanel/aiResults/commands',
        id: MenuId.SearchActionMenu,
        description: ( localize(
            13662,
            "The commands that will contribute to the menu rendered as buttons next to the AI search title"
        )),
    },
    {
        key: 'editor/context/chat',
        id: MenuId.ChatTextEditorMenu,
        description: ( localize(13663, "The Chat submenu in the text editor context menu.")),
        supportsSubmenus: false,
        proposed: 'chatParticipantPrivate'
    },
    {
        key: 'chat/input/editing/sessionToolbar',
        id: MenuId.ChatEditingSessionChangesToolbar,
        description: ( localize(13664, "The Chat Editing widget toolbar menu for session changes.")),
        proposed: 'chatSessionsProvider'
    },
    {
        key: 'chat/chatSessions',
        id: MenuId.AgentSessionsContext,
        description: ( localize(13665, "The Chat Sessions menu.")),
        supportsSubmenus: false,
        proposed: 'chatSessionsProvider'
    },
    {
        key: 'chatSessions/newSession',
        id: MenuId.AgentSessionsCreateSubMenu,
        description: ( localize(13666, "Menu for new chat sessions.")),
        supportsSubmenus: false,
        proposed: 'chatSessionsProvider'
    },
    {
        key: 'chat/multiDiff/context',
        id: MenuId.ChatMultiDiffContext,
        description: ( localize(13667, "The Chat Multi-Diff context menu.")),
        supportsSubmenus: false,
        proposed: 'chatSessionsProvider',
    },
];
var schema;
(function (schema) {
    function isMenuItem(item) {
        return typeof item.command === 'string';
    }
    schema.isMenuItem = isMenuItem;
    function isValidMenuItem(item, collector) {
        if (typeof item.command !== 'string') {
            collector.error(( localize(
                13668,
                "property `{0}` is mandatory and must be of type `string`",
                'command'
            )));
            return false;
        }
        if (item.alt && typeof item.alt !== 'string') {
            collector.error(( localize(13669, "property `{0}` can be omitted or must be of type `string`", 'alt')));
            return false;
        }
        if (item.when && typeof item.when !== 'string') {
            collector.error(( localize(13669, "property `{0}` can be omitted or must be of type `string`", 'when')));
            return false;
        }
        if (item.group && typeof item.group !== 'string') {
            collector.error(( localize(
                13669,
                "property `{0}` can be omitted or must be of type `string`",
                'group'
            )));
            return false;
        }
        return true;
    }
    schema.isValidMenuItem = isValidMenuItem;
    function isValidSubmenuItem(item, collector) {
        if (typeof item.submenu !== 'string') {
            collector.error(( localize(
                13668,
                "property `{0}` is mandatory and must be of type `string`",
                'submenu'
            )));
            return false;
        }
        if (item.when && typeof item.when !== 'string') {
            collector.error(( localize(13669, "property `{0}` can be omitted or must be of type `string`", 'when')));
            return false;
        }
        if (item.group && typeof item.group !== 'string') {
            collector.error(( localize(
                13669,
                "property `{0}` can be omitted or must be of type `string`",
                'group'
            )));
            return false;
        }
        return true;
    }
    schema.isValidSubmenuItem = isValidSubmenuItem;
    function isValidItems(items, collector) {
        if (!Array.isArray(items)) {
            collector.error(( localize(13670, "submenu items must be an array")));
            return false;
        }
        for (const item of items) {
            if (isMenuItem(item)) {
                if (!isValidMenuItem(item, collector)) {
                    return false;
                }
            }
            else {
                if (!isValidSubmenuItem(item, collector)) {
                    return false;
                }
            }
        }
        return true;
    }
    schema.isValidItems = isValidItems;
    function isValidSubmenu(submenu, collector) {
        if (typeof submenu !== 'object') {
            collector.error(( localize(13671, "submenu items must be an object")));
            return false;
        }
        if (typeof submenu.id !== 'string') {
            collector.error(( localize(13668, "property `{0}` is mandatory and must be of type `string`", 'id')));
            return false;
        }
        if (typeof submenu.label !== 'string') {
            collector.error(( localize(13668, "property `{0}` is mandatory and must be of type `string`", 'label')));
            return false;
        }
        return true;
    }
    schema.isValidSubmenu = isValidSubmenu;
    const menuItem = {
        type: 'object',
        required: ['command'],
        properties: {
            command: {
                description: ( localize(
                    13672,
                    'Identifier of the command to execute. The command must be declared in the \'commands\'-section'
                )),
                type: 'string'
            },
            alt: {
                description: ( localize(
                    13673,
                    'Identifier of an alternative command to execute. The command must be declared in the \'commands\'-section'
                )),
                type: 'string'
            },
            when: {
                description: ( localize(13674, 'Condition which must be true to show this item')),
                type: 'string'
            },
            group: {
                description: ( localize(13675, 'Group into which this item belongs')),
                type: 'string'
            }
        }
    };
    const submenuItem = {
        type: 'object',
        required: ['submenu'],
        properties: {
            submenu: {
                description: ( localize(13676, 'Identifier of the submenu to display in this item.')),
                type: 'string'
            },
            when: {
                description: ( localize(13674, 'Condition which must be true to show this item')),
                type: 'string'
            },
            group: {
                description: ( localize(13675, 'Group into which this item belongs')),
                type: 'string'
            }
        }
    };
    const submenu = {
        type: 'object',
        required: ['id', 'label'],
        properties: {
            id: {
                description: ( localize(13677, 'Identifier of the menu to display as a submenu.')),
                type: 'string'
            },
            label: {
                description: ( localize(13678, 'The label of the menu item which leads to this submenu.')),
                type: 'string'
            },
            icon: {
                description: ( localize(
                    13679,
                    '(Optional) Icon which is used to represent the submenu in the UI. Either a file path, an object with file paths for dark and light themes, or a theme icon references, like "\\$(zap)"'
                )),
                anyOf: [{
                        type: 'string'
                    },
                    {
                        type: 'object',
                        properties: {
                            light: {
                                description: ( localize(13680, 'Icon path when a light theme is used')),
                                type: 'string'
                            },
                            dark: {
                                description: ( localize(13681, 'Icon path when a dark theme is used')),
                                type: 'string'
                            }
                        }
                    }]
            }
        }
    };
    schema.menusContribution = {
        description: ( localize(13682, "Contributes menu items to the editor")),
        type: 'object',
        properties: index(apiMenus, menu => menu.key, menu => ({
            markdownDescription: menu.proposed ? ( localize(
                13683,
                "Proposed API, requires `enabledApiProposal: [\"{0}\"]` - {1}",
                menu.proposed,
                menu.description
            )) : menu.description,
            type: 'array',
            items: menu.supportsSubmenus === false ? menuItem : { oneOf: [menuItem, submenuItem] }
        })),
        additionalProperties: {
            description: 'Submenu',
            type: 'array',
            items: { oneOf: [menuItem, submenuItem] }
        }
    };
    schema.submenusContribution = {
        description: ( localize(13684, "Contributes submenu items to the editor")),
        type: 'array',
        items: submenu
    };
    function isValidCommand(command, collector) {
        if (!command) {
            collector.error(( localize(13685, "expected non-empty value.")));
            return false;
        }
        if (isFalsyOrWhitespace(command.command)) {
            collector.error(( localize(
                13668,
                "property `{0}` is mandatory and must be of type `string`",
                'command'
            )));
            return false;
        }
        if (!isValidLocalizedString(command.title, collector, 'title')) {
            return false;
        }
        if (command.shortTitle && !isValidLocalizedString(command.shortTitle, collector, 'shortTitle')) {
            return false;
        }
        if (command.enablement && typeof command.enablement !== 'string') {
            collector.error(( localize(
                13669,
                "property `{0}` can be omitted or must be of type `string`",
                'precondition'
            )));
            return false;
        }
        if (command.category && !isValidLocalizedString(command.category, collector, 'category')) {
            return false;
        }
        if (!isValidIcon(command.icon, collector)) {
            return false;
        }
        return true;
    }
    schema.isValidCommand = isValidCommand;
    function isValidIcon(icon, collector) {
        if (typeof icon === 'undefined') {
            return true;
        }
        if (typeof icon === 'string') {
            return true;
        }
        else if (typeof icon.dark === 'string' && typeof icon.light === 'string') {
            return true;
        }
        collector.error(( localize(
            13686,
            "property `icon` can be omitted or must be either a string or a literal like `{dark, light}`"
        )));
        return false;
    }
    function isValidLocalizedString(localized, collector, propertyName) {
        if (typeof localized === 'undefined') {
            collector.error(( localize(
                13687,
                "property `{0}` is mandatory and must be of type `string` or `object`",
                propertyName
            )));
            return false;
        }
        else if (typeof localized === 'string' && isFalsyOrWhitespace(localized)) {
            collector.error(( localize(
                13668,
                "property `{0}` is mandatory and must be of type `string`",
                propertyName
            )));
            return false;
        }
        else if (typeof localized !== 'string' && (isFalsyOrWhitespace(localized.original) || isFalsyOrWhitespace(localized.value))) {
            collector.error(( localize(
                13688,
                "properties `{0}` and `{1}` are mandatory and must be of type `string`",
                `${propertyName}.value`,
                `${propertyName}.original`
            )));
            return false;
        }
        return true;
    }
    const commandType = {
        type: 'object',
        required: ['command', 'title'],
        properties: {
            command: {
                description: ( localize(13689, 'Identifier of the command to execute')),
                type: 'string'
            },
            title: {
                description: ( localize(13690, 'Title by which the command is represented in the UI')),
                type: 'string'
            },
            shortTitle: {
                markdownDescription: ( localize(
                    13691,
                    '(Optional) Short title by which the command is represented in the UI. Menus pick either `title` or `shortTitle` depending on the context in which they show commands.'
                )),
                type: 'string'
            },
            category: {
                description: ( localize(
                    13692,
                    '(Optional) Category string by which the command is grouped in the UI'
                )),
                type: 'string'
            },
            enablement: {
                description: ( localize(
                    13693,
                    '(Optional) Condition which must be true to enable the command in the UI (menu and keybindings). Does not prevent executing the command by other means, like the `executeCommand`-api.'
                )),
                type: 'string'
            },
            icon: {
                description: ( localize(
                    13694,
                    '(Optional) Icon which is used to represent the command in the UI. Either a file path, an object with file paths for dark and light themes, or a theme icon references, like "\\$(zap)"'
                )),
                anyOf: [{
                        type: 'string'
                    },
                    {
                        type: 'object',
                        properties: {
                            light: {
                                description: ( localize(13695, 'Icon path when a light theme is used')),
                                type: 'string'
                            },
                            dark: {
                                description: ( localize(13696, 'Icon path when a dark theme is used')),
                                type: 'string'
                            }
                        }
                    }]
            }
        }
    };
    schema.commandsContribution = {
        description: ( localize(13697, "Contributes commands to the command palette.")),
        oneOf: [
            commandType,
            {
                type: 'array',
                items: commandType
            }
        ]
    };
})(schema || (schema = {}));
const _commandRegistrations = ( new DisposableStore());
const commandsExtensionPoint = ExtensionsRegistry.registerExtensionPoint({
    extensionPoint: 'commands',
    jsonSchema: schema.commandsContribution,
    activationEventsGenerator: function* (contribs) {
        for (const contrib of contribs) {
            if (contrib.command) {
                yield `onCommand:${contrib.command}`;
            }
        }
    }
});
commandsExtensionPoint.setHandler(extensions => {
    function handleCommand(userFriendlyCommand, extension) {
        if (!schema.isValidCommand(userFriendlyCommand, extension.collector)) {
            return;
        }
        const { icon, enablement, category, title, shortTitle, command } = userFriendlyCommand;
        let absoluteIcon;
        if (icon) {
            if (typeof icon === 'string') {
                absoluteIcon = ThemeIcon.fromString(icon) ?? { dark: joinPath(extension.description.extensionLocation, icon), light: joinPath(extension.description.extensionLocation, icon) };
            }
            else {
                absoluteIcon = {
                    dark: joinPath(extension.description.extensionLocation, icon.dark),
                    light: joinPath(extension.description.extensionLocation, icon.light)
                };
            }
        }
        const existingCmd = MenuRegistry.getCommand(command);
        if (existingCmd) {
            if (existingCmd.source) {
                extension.collector.info(( localize(
                    13698,
                    "Command `{0}` already registered by {1} ({2})",
                    userFriendlyCommand.command,
                    existingCmd.source.title,
                    existingCmd.source.id
                )));
            }
            else {
                extension.collector.info(( localize(13699, "Command `{0}` already registered", userFriendlyCommand.command)));
            }
        }
        _commandRegistrations.add(MenuRegistry.addCommand({
            id: command,
            title,
            source: { id: extension.description.identifier.value, title: extension.description.displayName ?? extension.description.name },
            shortTitle,
            tooltip: title,
            category,
            precondition: ContextKeyExpr.deserialize(enablement),
            icon: absoluteIcon
        }));
    }
    _commandRegistrations.clear();
    for (const extension of extensions) {
        const { value } = extension;
        if (Array.isArray(value)) {
            for (const command of value) {
                handleCommand(command, extension);
            }
        }
        else {
            handleCommand(value, extension);
        }
    }
});
const _submenus = ( new Map());
const submenusExtensionPoint = ExtensionsRegistry.registerExtensionPoint({
    extensionPoint: 'submenus',
    jsonSchema: schema.submenusContribution
});
submenusExtensionPoint.setHandler(extensions => {
    _submenus.clear();
    for (const extension of extensions) {
        const { value, collector } = extension;
        for (const [, submenuInfo] of Object.entries(value)) {
            if (!schema.isValidSubmenu(submenuInfo, collector)) {
                continue;
            }
            if (!submenuInfo.id) {
                collector.warn(( localize(13700, "`{0}` is not a valid submenu identifier", submenuInfo.id)));
                continue;
            }
            if (( _submenus.has(submenuInfo.id))) {
                collector.info(( localize(
                    13701,
                    "The `{0}` submenu was already previously registered.",
                    submenuInfo.id
                )));
                continue;
            }
            if (!submenuInfo.label) {
                collector.warn(( localize(13702, "`{0}` is not a valid submenu label", submenuInfo.label)));
                continue;
            }
            let absoluteIcon;
            if (submenuInfo.icon) {
                if (typeof submenuInfo.icon === 'string') {
                    absoluteIcon = ThemeIcon.fromString(submenuInfo.icon) || { dark: joinPath(extension.description.extensionLocation, submenuInfo.icon) };
                }
                else {
                    absoluteIcon = {
                        dark: joinPath(extension.description.extensionLocation, submenuInfo.icon.dark),
                        light: joinPath(extension.description.extensionLocation, submenuInfo.icon.light)
                    };
                }
            }
            const item = {
                id: MenuId.for(`api:${submenuInfo.id}`),
                label: submenuInfo.label,
                icon: absoluteIcon
            };
            _submenus.set(submenuInfo.id, item);
        }
    }
});
const _apiMenusByKey = ( new Map(( apiMenus.map(menu => ([menu.key, menu])))));
const _menuRegistrations = ( new DisposableStore());
const _submenuMenuItems = ( new Map());
const menusExtensionPoint = ExtensionsRegistry.registerExtensionPoint({
    extensionPoint: 'menus',
    jsonSchema: schema.menusContribution,
    deps: [submenusExtensionPoint]
});
menusExtensionPoint.setHandler(extensions => {
    _menuRegistrations.clear();
    _submenuMenuItems.clear();
    for (const extension of extensions) {
        const { value, collector } = extension;
        for (const entry of Object.entries(value)) {
            if (!schema.isValidItems(entry[1], collector)) {
                continue;
            }
            let menu = _apiMenusByKey.get(entry[0]);
            if (!menu) {
                const submenu = _submenus.get(entry[0]);
                if (submenu) {
                    menu = {
                        key: entry[0],
                        id: submenu.id,
                        description: ''
                    };
                }
            }
            if (!menu) {
                continue;
            }
            if (menu.proposed && !isProposedApiEnabled(extension.description, menu.proposed)) {
                collector.error(( localize(
                    13703,
                    "{0} is a proposed menu identifier. It requires 'package.json#enabledApiProposals: [\"{1}\"]' and is only available when running out of dev or with the following command line switch: --enable-proposed-api {2}",
                    entry[0],
                    menu.proposed,
                    extension.description.identifier.value
                )));
                continue;
            }
            for (const menuItem of entry[1]) {
                let item;
                if (schema.isMenuItem(menuItem)) {
                    const command = MenuRegistry.getCommand(menuItem.command);
                    const alt = menuItem.alt && MenuRegistry.getCommand(menuItem.alt) || undefined;
                    if (!command) {
                        collector.error(( localize(
                            13704,
                            "Menu item references a command `{0}` which is not defined in the 'commands' section.",
                            menuItem.command
                        )));
                        continue;
                    }
                    if (menuItem.alt && !alt) {
                        collector.warn(( localize(
                            13705,
                            "Menu item references an alt-command `{0}` which is not defined in the 'commands' section.",
                            menuItem.alt
                        )));
                    }
                    if (menuItem.command === menuItem.alt) {
                        collector.info(( localize(13706, "Menu item references the same command as default and alt-command")));
                    }
                    item = { command, alt, group: undefined, order: undefined, when: undefined };
                }
                else {
                    if (menu.supportsSubmenus === false) {
                        collector.error(( localize(
                            13707,
                            "Menu item references a submenu for a menu which doesn't have submenu support."
                        )));
                        continue;
                    }
                    const submenu = _submenus.get(menuItem.submenu);
                    if (!submenu) {
                        collector.error(( localize(
                            13708,
                            "Menu item references a submenu `{0}` which is not defined in the 'submenus' section.",
                            menuItem.submenu
                        )));
                        continue;
                    }
                    let submenuRegistrations = _submenuMenuItems.get(menu.id.id);
                    if (!submenuRegistrations) {
                        submenuRegistrations = ( new Set());
                        _submenuMenuItems.set(menu.id.id, submenuRegistrations);
                    }
                    if (( submenuRegistrations.has(submenu.id.id))) {
                        collector.warn(( localize(
                            13709,
                            "The `{0}` submenu was already contributed to the `{1}` menu.",
                            menuItem.submenu,
                            entry[0]
                        )));
                        continue;
                    }
                    submenuRegistrations.add(submenu.id.id);
                    item = { submenu: submenu.id, icon: submenu.icon, title: submenu.label, group: undefined, order: undefined, when: undefined };
                }
                if (menuItem.group) {
                    const idx = menuItem.group.lastIndexOf('@');
                    if (idx > 0) {
                        item.group = menuItem.group.substr(0, idx);
                        item.order = Number(menuItem.group.substr(idx + 1)) || undefined;
                    }
                    else {
                        item.group = menuItem.group;
                    }
                }
                if (menu.id === MenuId.ViewContainerTitle && !menuItem.when?.includes('viewContainer == workbench.view.debug')) {
                    collector.error(( localize(
                        13710,
                        "The {0} menu contribution must check {1} in its {2} clause.",
                        '`viewContainer/title`',
                        '`viewContainer == workbench.view.debug`',
                        '"when"'
                    )));
                    continue;
                }
                item.when = ContextKeyExpr.deserialize(menuItem.when);
                _menuRegistrations.add(MenuRegistry.appendMenuItem(menu.id, item));
            }
        }
    }
});
let CommandsTableRenderer = class CommandsTableRenderer extends Disposable {
    constructor(_keybindingService) {
        super();
        this._keybindingService = _keybindingService;
        this.type = 'table';
    }
    shouldRender(manifest) {
        return !!manifest.contributes?.commands;
    }
    render(manifest) {
        const rawCommands = manifest.contributes?.commands || [];
        const commands = ( rawCommands.map(c => ({
            id: c.command,
            title: c.title,
            keybindings: [],
            menus: []
        })));
        const byId = index(commands, c => c.id);
        const menus = manifest.contributes?.menus || {};
        const implicitlyOnCommandPalette = index(commands, c => c.id);
        if (menus['commandPalette']) {
            for (const command of menus['commandPalette']) {
                delete implicitlyOnCommandPalette[command.command];
            }
        }
        if (( Object.keys(implicitlyOnCommandPalette)).length) {
            if (!menus['commandPalette']) {
                menus['commandPalette'] = [];
            }
            for (const command in implicitlyOnCommandPalette) {
                menus['commandPalette'].push({ command });
            }
        }
        for (const context in menus) {
            for (const menu of menus[context]) {
                if (menu.when === 'false') {
                    continue;
                }
                if (menu.command) {
                    let command = byId[menu.command];
                    if (command) {
                        if (!command.menus.includes(context)) {
                            command.menus.push(context);
                        }
                    }
                    else {
                        command = { id: menu.command, title: '', keybindings: [], menus: [context] };
                        byId[command.id] = command;
                        commands.push(command);
                    }
                }
            }
        }
        const rawKeybindings = manifest.contributes?.keybindings ? (Array.isArray(manifest.contributes.keybindings) ? manifest.contributes.keybindings : [manifest.contributes.keybindings]) : [];
        rawKeybindings.forEach(rawKeybinding => {
            const keybinding = this.resolveKeybinding(rawKeybinding);
            if (!keybinding) {
                return;
            }
            let command = byId[rawKeybinding.command];
            if (command) {
                command.keybindings.push(keybinding);
            }
            else {
                command = { id: rawKeybinding.command, title: '', keybindings: [keybinding], menus: [] };
                byId[command.id] = command;
                commands.push(command);
            }
        });
        if (!commands.length) {
            return { data: { headers: [], rows: [] }, dispose: () => { } };
        }
        const headers = [
            ( localize(13711, "ID")),
            ( localize(13712, "Title")),
            ( localize(13713, "Keyboard Shortcuts")),
            ( localize(13714, "Menu Contexts"))
        ];
        const rows = ( commands.sort((a, b) => a.id.localeCompare(b.id))
            .map(command => {
            return [
                ( new MarkdownString()).appendMarkdown(`\`${command.id}\``),
                typeof command.title === 'string' ? command.title : command.title.value,
                command.keybindings,
                ( new MarkdownString()).appendMarkdown(`${( command.menus.sort((a, b) => a.localeCompare(b)).map(menu => `\`${menu}\``)).join('&nbsp;')}`),
            ];
        }));
        return {
            data: {
                headers,
                rows
            },
            dispose: () => { }
        };
    }
    resolveKeybinding(rawKeyBinding) {
        let key;
        switch (platform) {
            case 'win32':
                key = rawKeyBinding.win;
                break;
            case 'linux':
                key = rawKeyBinding.linux;
                break;
            case 'darwin':
                key = rawKeyBinding.mac;
                break;
        }
        return this._keybindingService.resolveUserBinding(key ?? rawKeyBinding.key)[0];
    }
};
CommandsTableRenderer = ( __decorate([
    ( __param(0, IKeybindingService))
], CommandsTableRenderer));
( Registry.as(Extensions.ExtensionFeaturesRegistry)).registerExtensionFeature({
    id: 'commands',
    label: ( localize(13715, "Commands")),
    access: {
        canToggle: false,
    },
    renderer: ( new SyncDescriptor(CommandsTableRenderer)),
});

export { commandsExtensionPoint };
