// @codingame/monaco-vscode-log-service-override/vscode/src/vs/workbench/services/log/common/defaultLogLevels.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { getLogLevel, LogLevelToString, parseLogLevel } from '@codingame/monaco-vscode-api/vscode/vs/platform/log/common/log';
import { ILogService, ILoggerService } from '@codingame/monaco-vscode-api/vscode/vs/platform/log/common/log.service';
import { IWorkbenchEnvironmentService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/environment/common/environmentService.service';
import { toFileOperationResult, FileOperationResult } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/common/files';
import { IFileService } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/common/files.service';
import { IJSONEditingService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/configuration/common/jsonEditing.service';
import { isUndefined, isString } from '@codingame/monaco-vscode-api/vscode/vs/base/common/types';
import { EXTENSION_IDENTIFIER_WITH_LOG_REGEX } from '@codingame/monaco-vscode-api/vscode/vs/platform/environment/common/environmentService';
import '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/extensions';
import { parse } from '@codingame/monaco-vscode-api/vscode/vs/base/common/json';
import { Disposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import { Emitter } from '@codingame/monaco-vscode-api/vscode/vs/base/common/event';
import { equals } from '@codingame/monaco-vscode-api/vscode/vs/base/common/objects';
import '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/instantiation';

let DefaultLogLevelsService = class DefaultLogLevelsService extends Disposable {
    constructor(environmentService, fileService, jsonEditingService, logService, loggerService) {
        super();
        this.environmentService = environmentService;
        this.fileService = fileService;
        this.jsonEditingService = jsonEditingService;
        this.logService = logService;
        this.loggerService = loggerService;
        this._onDidChangeDefaultLogLevels = this._register(( new Emitter()));
        this.onDidChangeDefaultLogLevels = this._onDidChangeDefaultLogLevels.event;
        this._defaultLogLevels = {
            default: this._getDefaultLogLevelFromEnv(),
            extensions: this._getExtensionsDefaultLogLevelsFromEnv()
        };
        this._register(this.fileService.onDidFilesChange(e => {
            if (e.contains(this.environmentService.argvResource)) {
                this.onDidChangeArgv();
            }
        }));
    }
    async onDidChangeArgv() {
        const defaultLogLevelsFromArgv = await this._parseLogLevelsFromArgv();
        this.updateDefaultLogLevels(defaultLogLevelsFromArgv);
    }
    get defaultLogLevels() {
        return this._defaultLogLevels;
    }
    updateDefaultLogLevels(defaultLogLevelsFromArgv) {
        const defaultLogLevels = {
            default: defaultLogLevelsFromArgv?.default ?? this._getDefaultLogLevelFromEnv(),
            extensions: defaultLogLevelsFromArgv?.extensions ?? this._getExtensionsDefaultLogLevelsFromEnv()
        };
        if (!equals(this._defaultLogLevels, defaultLogLevels)) {
            this._defaultLogLevels = defaultLogLevels;
            this._onDidChangeDefaultLogLevels.fire(this._defaultLogLevels);
        }
    }
    getDefaultLogLevel(extensionId) {
        if (extensionId) {
            extensionId = extensionId.toLowerCase();
            return this._getDefaultLogLevel(this._defaultLogLevels, extensionId);
        }
        else {
            return this._getDefaultLogLevel(this._defaultLogLevels);
        }
    }
    async setDefaultLogLevel(defaultLogLevel, extensionId) {
        const defaultLogLevelsFromArgv = (await this._parseLogLevelsFromArgv()) ?? {};
        if (extensionId) {
            extensionId = extensionId.toLowerCase();
            const currentDefaultLogLevel = this._getDefaultLogLevel(defaultLogLevelsFromArgv, extensionId);
            defaultLogLevelsFromArgv.extensions = defaultLogLevelsFromArgv.extensions ?? [];
            const extension = defaultLogLevelsFromArgv.extensions.find(([extension]) => extension === extensionId);
            if (extension) {
                extension[1] = defaultLogLevel;
            }
            else {
                defaultLogLevelsFromArgv.extensions.push([extensionId, defaultLogLevel]);
            }
            await this._writeLogLevelsToArgv(defaultLogLevelsFromArgv);
            const extensionLoggers = [...this.loggerService.getRegisteredLoggers()].filter(logger => logger.extensionId && logger.extensionId.toLowerCase() === extensionId);
            for (const { resource } of extensionLoggers) {
                if (this.loggerService.getLogLevel(resource) === currentDefaultLogLevel) {
                    this.loggerService.setLogLevel(resource, defaultLogLevel);
                }
            }
        }
        else {
            const currentLogLevel = this._getDefaultLogLevel(defaultLogLevelsFromArgv);
            defaultLogLevelsFromArgv.default = defaultLogLevel;
            await this._writeLogLevelsToArgv(defaultLogLevelsFromArgv);
            if (this.loggerService.getLogLevel() === currentLogLevel) {
                this.loggerService.setLogLevel(defaultLogLevel);
            }
        }
        this.updateDefaultLogLevels(defaultLogLevelsFromArgv);
    }
    _getDefaultLogLevel(argvLogLevels, extension) {
        if (extension) {
            const extensionLogLevel = argvLogLevels.extensions?.find(([extensionId]) => extensionId === extension);
            if (extensionLogLevel) {
                return extensionLogLevel[1];
            }
        }
        return argvLogLevels.default ?? getLogLevel(this.environmentService);
    }
    async _writeLogLevelsToArgv(logLevels) {
        const logLevelsValue = [];
        if (!isUndefined(logLevels.default)) {
            logLevelsValue.push(LogLevelToString(logLevels.default));
        }
        for (const [extension, logLevel] of logLevels.extensions ?? []) {
            logLevelsValue.push(`${extension}=${LogLevelToString(logLevel)}`);
        }
        await this.jsonEditingService.write(this.environmentService.argvResource, [{ path: ['log-level'], value: logLevelsValue.length ? logLevelsValue : undefined }], true);
    }
    async _parseLogLevelsFromArgv() {
        const result = { extensions: [] };
        const logLevels = await this._readLogLevelsFromArgv();
        for (const extensionLogLevel of logLevels) {
            const matches = EXTENSION_IDENTIFIER_WITH_LOG_REGEX.exec(extensionLogLevel);
            if (matches && matches[1] && matches[2]) {
                const logLevel = parseLogLevel(matches[2]);
                if (!isUndefined(logLevel)) {
                    result.extensions?.push([matches[1].toLowerCase(), logLevel]);
                }
            }
            else {
                const logLevel = parseLogLevel(extensionLogLevel);
                if (!isUndefined(logLevel)) {
                    result.default = logLevel;
                }
            }
        }
        return !isUndefined(result.default) || result.extensions?.length ? result : undefined;
    }
    async _readLogLevelsFromArgv() {
        try {
            const content = await this.fileService.readFile(this.environmentService.argvResource);
            const argv = parse(( content.value.toString()));
            return isString(argv['log-level']) ? [argv['log-level']] : Array.isArray(argv['log-level']) ? argv['log-level'] : [];
        }
        catch (error) {
            if (toFileOperationResult(error) !== FileOperationResult.FILE_NOT_FOUND) {
                this.logService.error(error);
            }
        }
        return [];
    }
    _getDefaultLogLevelFromEnv() {
        return getLogLevel(this.environmentService);
    }
    _getExtensionsDefaultLogLevelsFromEnv() {
        const result = [];
        for (const [extension, logLevelValue] of this.environmentService.extensionLogLevel ?? []) {
            const logLevel = parseLogLevel(logLevelValue);
            if (!isUndefined(logLevel)) {
                result.push([extension, logLevel]);
            }
        }
        return result;
    }
};
DefaultLogLevelsService = ( __decorate([
    ( __param(0, IWorkbenchEnvironmentService)),
    ( __param(1, IFileService)),
    ( __param(2, IJSONEditingService)),
    ( __param(3, ILogService)),
    ( __param(4, ILoggerService))
], DefaultLogLevelsService));

export { DefaultLogLevelsService };
