// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminalContrib/history/common/terminal.history.js

import { localize } from '../../../../../nls.js';

var TerminalHistoryCommandId;
(function (TerminalHistoryCommandId) {
    TerminalHistoryCommandId["ClearPreviousSessionHistory"] = "workbench.action.terminal.clearPreviousSessionHistory";
    TerminalHistoryCommandId["GoToRecentDirectory"] = "workbench.action.terminal.goToRecentDirectory";
    TerminalHistoryCommandId["RunRecentCommand"] = "workbench.action.terminal.runRecentCommand";
})(TerminalHistoryCommandId || (TerminalHistoryCommandId = {}));
const defaultTerminalHistoryCommandsToSkipShell = [
    TerminalHistoryCommandId.GoToRecentDirectory,
    TerminalHistoryCommandId.RunRecentCommand
];
var TerminalHistorySettingId;
(function (TerminalHistorySettingId) {
    TerminalHistorySettingId["ShellIntegrationCommandHistory"] = "terminal.integrated.shellIntegration.history";
})(TerminalHistorySettingId || (TerminalHistorySettingId = {}));
const terminalHistoryConfiguration = {
    [TerminalHistorySettingId.ShellIntegrationCommandHistory]: {
        restricted: true,
        markdownDescription: ( localize(
            12189,
            "Controls the number of recently used commands to keep in the terminal command history. Set to 0 to disable terminal command history."
        )),
        type: 'number',
        default: 100
    },
};

export { TerminalHistoryCommandId, TerminalHistorySettingId, defaultTerminalHistoryCommandsToSkipShell, terminalHistoryConfiguration };
