// @codingame/monaco-vscode-api/vscode/src/vs/editor/contrib/floatingMenu/browser/floatingMenu.js

import { __decorate, __param } from '../../../../../../../external/tslib/tslib.es6.js';
import { h } from '../../../../base/browser/dom.js';
import { Disposable } from '../../../../base/common/lifecycle.js';
import '../../../../base/common/observableInternal/index.js';
import { MenuEntryActionViewItem } from '../../../../platform/actions/browser/menuEntryActionViewItem.js';
import { MenuWorkbenchToolBar, HiddenItemStrategy } from '../../../../platform/actions/browser/toolbar.js';
import { MenuId, MenuItemAction } from '../../../../platform/actions/common/actions.js';
import { IMenuService } from '../../../../platform/actions/common/actions.service.js';
import { IInstantiationService } from '../../../../platform/instantiation/common/instantiation.js';
import { IKeybindingService } from '../../../../platform/keybinding/common/keybinding.service.js';
import { OverlayWidgetPositionPreference } from '../../../browser/editorBrowser.js';
import { observableCodeEditor } from '../../../browser/observableCodeEditor.js';
import { observableFromEvent } from '../../../../base/common/observableInternal/observables/observableFromEvent.js';
import { autorun } from '../../../../base/common/observableInternal/reactions/autorun.js';
import { constObservable } from '../../../../base/common/observableInternal/observables/constObservable.js';

let FloatingEditorToolbar = class FloatingEditorToolbar extends Disposable {
    static { this.ID = 'editor.contrib.floatingToolbar'; }
    constructor(editor, instantiationService, keybindingService, menuService) {
        super();
        const editorObs = this._register(observableCodeEditor(editor));
        const menu = this._register(menuService.createMenu(MenuId.EditorContent, editor.contextKeyService));
        const menuIsEmptyObs = observableFromEvent(this, menu.onDidChange, () => menu.getActions().length === 0);
        this._register(autorun(reader => {
            const menuIsEmpty = menuIsEmptyObs.read(reader);
            if (menuIsEmpty) {
                return;
            }
            const container = h('div.floating-menu-overlay-widget');
            container.root.style.height = '28px';
            const toolbar = instantiationService.createInstance(MenuWorkbenchToolBar, container.root, MenuId.EditorContent, {
                actionViewItemProvider: (action, options) => {
                    if (!(action instanceof MenuItemAction)) {
                        return undefined;
                    }
                    const keybinding = keybindingService.lookupKeybinding(action.id);
                    if (!keybinding) {
                        return undefined;
                    }
                    return instantiationService.createInstance(class extends MenuEntryActionViewItem {
                        updateLabel() {
                            if (this.options.label && this.label) {
                                this.label.textContent = `${this._commandAction.label} (${keybinding.getLabel()})`;
                            }
                        }
                    }, action, { ...options, keybindingNotRenderedWithLabel: true });
                },
                hiddenItemStrategy: HiddenItemStrategy.Ignore,
                menuOptions: {
                    shouldForwardArgs: true
                },
                telemetrySource: 'editor.overlayToolbar',
                toolbarOptions: {
                    primaryGroup: () => true,
                    useSeparatorsInPrimaryActions: true
                },
            });
            reader.store.add(toolbar);
            reader.store.add(autorun(reader => {
                const model = editorObs.model.read(reader);
                toolbar.context = model?.uri;
            }));
            reader.store.add(editorObs.createOverlayWidget({
                allowEditorOverflow: false,
                domNode: container.root,
                minContentWidthInPx: constObservable(0),
                position: constObservable({
                    preference: OverlayWidgetPositionPreference.BOTTOM_RIGHT_CORNER
                })
            }));
        }));
    }
};
FloatingEditorToolbar = ( __decorate([
    ( __param(1, IInstantiationService)),
    ( __param(2, IKeybindingService)),
    ( __param(3, IMenuService))
], FloatingEditorToolbar));

export { FloatingEditorToolbar };
