// @codingame/monaco-vscode-workbench-service-override/index.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { Workbench } from './vscode/src/vs/workbench/browser/workbench.js';
import { ILogService } from '@codingame/monaco-vscode-api/vscode/vs/platform/log/common/log.service';
import { ServiceCollection } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/serviceCollection';
import { IInstantiationService } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/instantiation';
import { IWorkbenchLayoutService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/layout/browser/layoutService.service';
import { SyncDescriptor } from '@codingame/monaco-vscode-api/vscode/vs/platform/instantiation/common/descriptors';
import { IEditorGroupsService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/editor/common/editorGroupsService.service';
import { IEditorService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/editor/common/editorService.service';
import { EditorService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/editor/browser/editorService';
import { EditorParts } from '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/parts/editor/editorParts';
import { setUnexpectedErrorHandler, onUnexpectedError } from '@codingame/monaco-vscode-api/vscode/vs/base/common/errors';
import { BrowserWindow } from '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/window';
import { detectFullscreen } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/dom';
import { mainWindow } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/window';
import { setFullscreen } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/browser';
import getServiceOverride$4 from '@codingame/monaco-vscode-keybindings-service-override';
import getServiceOverride$5 from '@codingame/monaco-vscode-quickaccess-service-override';
import getServiceOverride$3 from '@codingame/monaco-vscode-view-title-bar-service-override';
import getServiceOverride$2 from '@codingame/monaco-vscode-view-status-bar-service-override';
import getServiceOverride$1 from '@codingame/monaco-vscode-view-banner-service-override';
import getServiceOverride$6 from '@codingame/monaco-vscode-view-common-service-override';
import { getWorkbenchContainer } from '@codingame/monaco-vscode-api/workbench';
import { onLayout, onRenderWorkbench } from '@codingame/monaco-vscode-api/lifecycle';
export { EditorPane, SimpleEditorInput, SimpleEditorPane, getPanelPosition, getSideBarPosition, isPartVisibile, onDidChangePanelPosition, onDidChangeSideBarPosition, registerCustomView, registerEditor, registerEditorPane, registerEditorSerializer, setPartVisibility, viewContainerRegistry, viewRegistry } from '@codingame/monaco-vscode-api/service-override/tools/views';
export { AbstractResourceEditorInput } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/editor/resourceEditorInput';
export { AbstractTextResourceEditorInput } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/editor/textResourceEditorInput';
export { ConfirmResult } from '@codingame/monaco-vscode-api/vscode/vs/platform/dialogs/common/dialogs';
export { DomScrollableElement } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/ui/scrollbar/scrollableElement';
export { EditorInput } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/editor/editorInput';
export { EditorInputCapabilities } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/editor';
export { Parts } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/layout/browser/layoutService';
export { RegisteredEditorPriority } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/editor/common/editorResolverService';
export { SplitView } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/ui/splitview/splitview';
export { ViewContainerLocation } from '@codingame/monaco-vscode-api/vscode/vs/workbench/common/views';
export { ViewPaneContainer } from '@codingame/monaco-vscode-api/vscode/vs/workbench/browser/parts/views/viewPaneContainer';

let CustomWorkbench = class CustomWorkbench extends Workbench {
    constructor(options, logService, instantiationService) {
        super(getWorkbenchContainer(), options, new ServiceCollection(), logService);
        this.instantiationService = instantiationService;
        this.mainContainer.classList.add('monaco-workbench-part');
    }
    registerErrorHandler(logService) {
        setUnexpectedErrorHandler((error) => this.handleUnexpectedError(error, logService));
    }
    createNotificationsHandlers() {
    }
    initServices() {
        return this.instantiationService;
    }
    restore(instantiationService) {
        try {
            this.restoreParts(instantiationService);
        }
        catch (error) {
            onUnexpectedError(error);
        }
    }
};
CustomWorkbench = __decorate([
    __param(1, ILogService),
    __param(2, IInstantiationService)
], CustomWorkbench);
onLayout(async (accessor) => {
    accessor.get(IWorkbenchLayoutService).startup();
    const detectedFullScreen = detectFullscreen(mainWindow, getWorkbenchContainer());
    setFullscreen(detectedFullScreen != null && !detectedFullScreen.guess, mainWindow);
});
onRenderWorkbench(async (accessor) => {
    accessor.get(IInstantiationService).createInstance(BrowserWindow);
});
function getServiceOverride(options, _webviewIframeAlternateDomains) {
    return {
        ...getServiceOverride$6(_webviewIframeAlternateDomains),
        ...getServiceOverride$5({
            isKeybindingConfigurationVisible: () => true,
            shouldUseGlobalPicker: () => true
        }),
        ...getServiceOverride$4({
            shouldUseGlobalKeybindings: () => true
        }),
        [IWorkbenchLayoutService.toString()]: new SyncDescriptor(CustomWorkbench, [options], false),
        [IEditorGroupsService.toString()]: new SyncDescriptor(EditorParts, [], false),
        [IEditorService.toString()]: new SyncDescriptor(EditorService, [undefined], false),
        ...getServiceOverride$3(),
        ...getServiceOverride$2(),
        ...getServiceOverride$1()
    };
}

export { getServiceOverride as default };
