// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminalContrib/zoom/common/terminal.zoom.js

import { isMacintosh } from '../../../../../base/common/platform.js';
import { localize } from '../../../../../nls.js';

var TerminalZoomCommandId;
(function (TerminalZoomCommandId) {
    TerminalZoomCommandId["FontZoomIn"] = "workbench.action.terminal.fontZoomIn";
    TerminalZoomCommandId["FontZoomOut"] = "workbench.action.terminal.fontZoomOut";
    TerminalZoomCommandId["FontZoomReset"] = "workbench.action.terminal.fontZoomReset";
})(TerminalZoomCommandId || (TerminalZoomCommandId = {}));
var TerminalZoomSettingId;
(function (TerminalZoomSettingId) {
    TerminalZoomSettingId["MouseWheelZoom"] = "terminal.integrated.mouseWheelZoom";
})(TerminalZoomSettingId || (TerminalZoomSettingId = {}));
const terminalZoomConfiguration = {
    [TerminalZoomSettingId.MouseWheelZoom]: {
        markdownDescription: isMacintosh
            ? ( localize(
            12400,
            "Zoom the font of the terminal when using mouse wheel and holding `Cmd`."
        ))
            : ( localize(
            12401,
            "Zoom the font of the terminal when using mouse wheel and holding `Ctrl`."
        )),
        type: 'boolean',
        default: false
    },
};

export { TerminalZoomCommandId, TerminalZoomSettingId, terminalZoomConfiguration };
