// @codingame/monaco-vscode-api/vscode/src/vs/base/common/observableInternal/index.js

import '../errors.js';
import '../equals.js';
import '../event.js';
import '../lifecycle.js';
import { setLogObservableFn, addLogger } from './logging/logging.js';
import { _setDebugGetObservableGraph } from './observables/baseObservable.js';
export { DebugLocation } from './debugLocation.js';
import './reactions/autorunImpl.js';
export { derived, derivedDisposable, derivedHandleChanges, derivedOpts, derivedWithSetter } from './observables/derived.js';
import '../cancellation.js';
import './observables/derivedImpl.js';
export { debouncedObservable, debouncedObservable2, derivedObservableWithCache, derivedObservableWithWritableCache, isObservable, keepObserved, mapObservableArrayCached, observableFromPromise, recomputeInitiallyAndOnChange, wasEventTriggeredRecently } from './utils/utils.js';
export { observableFromEvent, observableFromEventOpts } from './observables/observableFromEvent.js';
import { logObservableToConsole } from './logging/consoleObservableLogger.js';
import { DevToolsLogger } from './logging/debugger/devToolsLogger.js';
import { env } from '../process.js';
import { debugGetObservableGraph } from './logging/debugGetDependencyGraph.js';

_setDebugGetObservableGraph(debugGetObservableGraph);
setLogObservableFn(logObservableToConsole);
if (env && env['VSCODE_DEV_DEBUG_OBSERVABLES']) {
    addLogger(DevToolsLogger.getInstance());
}
