// @codingame/monaco-vscode-api/vscode/src/vs/workbench/services/extensionManagement/common/extensionManagement.js


var ExtensionInstallLocation;
(function (ExtensionInstallLocation) {
    ExtensionInstallLocation[ExtensionInstallLocation["Local"] = 1] = "Local";
    ExtensionInstallLocation[ExtensionInstallLocation["Remote"] = 2] = "Remote";
    ExtensionInstallLocation[ExtensionInstallLocation["Web"] = 3] = "Web";
})(ExtensionInstallLocation || (ExtensionInstallLocation = {}));
var EnablementState;
(function (EnablementState) {
    EnablementState[EnablementState["DisabledByTrustRequirement"] = 0] = "DisabledByTrustRequirement";
    EnablementState[EnablementState["DisabledByExtensionKind"] = 1] = "DisabledByExtensionKind";
    EnablementState[EnablementState["DisabledByEnvironment"] = 2] = "DisabledByEnvironment";
    EnablementState[EnablementState["EnabledByEnvironment"] = 3] = "EnabledByEnvironment";
    EnablementState[EnablementState["DisabledByMalicious"] = 4] = "DisabledByMalicious";
    EnablementState[EnablementState["DisabledByVirtualWorkspace"] = 5] = "DisabledByVirtualWorkspace";
    EnablementState[EnablementState["DisabledByInvalidExtension"] = 6] = "DisabledByInvalidExtension";
    EnablementState[EnablementState["DisabledByAllowlist"] = 7] = "DisabledByAllowlist";
    EnablementState[EnablementState["DisabledByExtensionDependency"] = 8] = "DisabledByExtensionDependency";
    EnablementState[EnablementState["DisabledByUnification"] = 9] = "DisabledByUnification";
    EnablementState[EnablementState["DisabledGlobally"] = 10] = "DisabledGlobally";
    EnablementState[EnablementState["DisabledWorkspace"] = 11] = "DisabledWorkspace";
    EnablementState[EnablementState["EnabledGlobally"] = 12] = "EnabledGlobally";
    EnablementState[EnablementState["EnabledWorkspace"] = 13] = "EnabledWorkspace";
})(EnablementState || (EnablementState = {}));

export { EnablementState, ExtensionInstallLocation };
