// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/preferences/browser/settingsLayout.js

import { isWindows, isWeb } from '../../../../base/common/platform.js';
import { localize } from '../../../../nls.js';

const defaultCommonlyUsedSettings = [
    'editor.fontSize',
    'editor.formatOnSave',
    'files.autoSave',
    'editor.defaultFormatter',
    'editor.fontFamily',
    'editor.wordWrap',
    'files.exclude',
    'workbench.colorTheme',
    'editor.tabSize',
    'editor.mouseWheelZoom',
    'editor.formatOnPaste'
];
function getCommonlyUsedData(settingGroups, commonlyUsed = defaultCommonlyUsedSettings) {
    const allSettings = ( new Map());
    for (const group of settingGroups) {
        for (const section of group.sections) {
            for (const s of section.settings) {
                allSettings.set(s.key, s);
            }
        }
    }
    const settings = [];
    for (const id of commonlyUsed) {
        const setting = allSettings.get(id);
        if (setting) {
            settings.push(setting);
        }
    }
    return {
        id: 'commonlyUsed',
        label: ( localize(10021, "Commonly Used")),
        settings
    };
}
const tocData = {
    id: 'root',
    label: 'root',
    children: [
        {
            id: 'editor',
            label: ( localize(10022, "Text Editor")),
            settings: ['editor.*'],
            children: [
                {
                    id: 'editor/cursor',
                    label: ( localize(10023, "Cursor")),
                    settings: ['editor.cursor*']
                },
                {
                    id: 'editor/find',
                    label: ( localize(10024, "Find")),
                    settings: ['editor.find.*']
                },
                {
                    id: 'editor/font',
                    label: ( localize(10025, "Font")),
                    settings: ['editor.font*']
                },
                {
                    id: 'editor/format',
                    label: ( localize(10026, "Formatting")),
                    settings: ['editor.format*']
                },
                {
                    id: 'editor/diffEditor',
                    label: ( localize(10027, "Diff Editor")),
                    settings: ['diffEditor.*']
                },
                {
                    id: 'editor/multiDiffEditor',
                    label: ( localize(10028, "Multi-File Diff Editor")),
                    settings: ['multiDiffEditor.*']
                },
                {
                    id: 'editor/minimap',
                    label: ( localize(10029, "Minimap")),
                    settings: ['editor.minimap.*']
                },
                {
                    id: 'editor/suggestions',
                    label: ( localize(10030, "Suggestions")),
                    settings: ['editor.*suggest*']
                },
                {
                    id: 'editor/files',
                    label: ( localize(10031, "Files")),
                    settings: ['files.*']
                }
            ]
        },
        {
            id: 'workbench',
            label: ( localize(10032, "Workbench")),
            settings: ['workbench.*'],
            children: [
                {
                    id: 'workbench/appearance',
                    label: ( localize(10033, "Appearance")),
                    settings: ['workbench.activityBar.*', 'workbench.*color*', 'workbench.fontAliasing', 'workbench.iconTheme', 'workbench.sidebar.location', 'workbench.*.visible', 'workbench.tips.enabled', 'workbench.tree.*', 'workbench.view.*']
                },
                {
                    id: 'workbench/breadcrumbs',
                    label: ( localize(10034, "Breadcrumbs")),
                    settings: ['breadcrumbs.*']
                },
                {
                    id: 'workbench/editor',
                    label: ( localize(10035, "Editor Management")),
                    settings: ['workbench.editor.*']
                },
                {
                    id: 'workbench/settings',
                    label: ( localize(10036, "Settings Editor")),
                    settings: ['workbench.settings.*']
                },
                {
                    id: 'workbench/zenmode',
                    label: ( localize(10037, "Zen Mode")),
                    settings: ['zenmode.*']
                },
                {
                    id: 'workbench/screencastmode',
                    label: ( localize(10038, "Screencast Mode")),
                    settings: ['screencastMode.*']
                }
            ]
        },
        {
            id: 'window',
            label: ( localize(10039, "Window")),
            settings: ['window.*'],
            children: [
                {
                    id: 'window/newWindow',
                    label: ( localize(10040, "New Window")),
                    settings: ['window.*newwindow*']
                }
            ]
        },
        {
            id: 'features',
            label: ( localize(10041, "Features")),
            children: [
                {
                    id: 'features/accessibilitySignals',
                    label: ( localize(10042, 'Accessibility Signals')),
                    settings: ['accessibility.signal*']
                },
                {
                    id: 'features/accessibility',
                    label: ( localize(10043, "Accessibility")),
                    settings: ['accessibility.*']
                },
                {
                    id: 'features/explorer',
                    label: ( localize(10044, "Explorer")),
                    settings: ['explorer.*', 'outline.*']
                },
                {
                    id: 'features/search',
                    label: ( localize(10045, "Search")),
                    settings: ['search.*']
                },
                {
                    id: 'features/debug',
                    label: ( localize(10046, "Debug")),
                    settings: ['debug.*', 'launch']
                },
                {
                    id: 'features/testing',
                    label: ( localize(10047, "Testing")),
                    settings: ['testing.*']
                },
                {
                    id: 'features/scm',
                    label: ( localize(10048, "Source Control")),
                    settings: ['scm.*']
                },
                {
                    id: 'features/extensions',
                    label: ( localize(10049, "Extensions")),
                    settings: ['extensions.*']
                },
                {
                    id: 'features/terminal',
                    label: ( localize(10050, "Terminal")),
                    settings: ['terminal.*']
                },
                {
                    id: 'features/task',
                    label: ( localize(10051, "Task")),
                    settings: ['task.*']
                },
                {
                    id: 'features/problems',
                    label: ( localize(10052, "Problems")),
                    settings: ['problems.*']
                },
                {
                    id: 'features/output',
                    label: ( localize(10053, "Output")),
                    settings: ['output.*']
                },
                {
                    id: 'features/comments',
                    label: ( localize(10054, "Comments")),
                    settings: ['comments.*']
                },
                {
                    id: 'features/remote',
                    label: ( localize(10055, "Remote")),
                    settings: ['remote.*']
                },
                {
                    id: 'features/timeline',
                    label: ( localize(10056, "Timeline")),
                    settings: ['timeline.*']
                },
                {
                    id: 'features/notebook',
                    label: ( localize(10057, 'Notebook')),
                    settings: ['notebook.*', 'interactiveWindow.*']
                },
                {
                    id: 'features/mergeEditor',
                    label: ( localize(10058, 'Merge Editor')),
                    settings: ['mergeEditor.*']
                },
                {
                    id: 'features/chat',
                    label: ( localize(10059, 'Chat')),
                    settings: ['chat.*', 'inlineChat.*', 'mcp']
                },
                {
                    id: 'features/issueReporter',
                    label: ( localize(10060, 'Issue Reporter')),
                    settings: ['issueReporter.*'],
                    hide: !isWeb
                }
            ]
        },
        {
            id: 'application',
            label: ( localize(10061, "Application")),
            children: [
                {
                    id: 'application/http',
                    label: ( localize(10062, "Proxy")),
                    settings: ['http.*']
                },
                {
                    id: 'application/keyboard',
                    label: ( localize(10063, "Keyboard")),
                    settings: ['keyboard.*']
                },
                {
                    id: 'application/update',
                    label: ( localize(10064, "Update")),
                    settings: ['update.*']
                },
                {
                    id: 'application/telemetry',
                    label: ( localize(10065, "Telemetry")),
                    settings: ['telemetry.*']
                },
                {
                    id: 'application/settingsSync',
                    label: ( localize(10066, "Settings Sync")),
                    settings: ['settingsSync.*']
                },
                {
                    id: 'application/experimental',
                    label: ( localize(10067, "Experimental")),
                    settings: ['application.experimental.*']
                },
                {
                    id: 'application/other',
                    label: ( localize(10068, "Other")),
                    settings: ['application.*'],
                    hide: isWindows
                }
            ]
        },
        {
            id: 'security',
            label: ( localize(10069, "Security")),
            settings: ['security.*'],
            children: [
                {
                    id: 'security/workspace',
                    label: ( localize(10070, "Workspace")),
                    settings: ['security.workspace.*']
                }
            ]
        }
    ]
};

export { getCommonlyUsedData, tocData };
