// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminalContrib/accessibility/common/terminalAccessibilityConfiguration.js

import { localize } from '../../../../../nls.js';

var TerminalAccessibilitySettingId;
(function (TerminalAccessibilitySettingId) {
    TerminalAccessibilitySettingId["AccessibleViewPreserveCursorPosition"] = "terminal.integrated.accessibleViewPreserveCursorPosition";
    TerminalAccessibilitySettingId["AccessibleViewFocusOnCommandExecution"] = "terminal.integrated.accessibleViewFocusOnCommandExecution";
})(TerminalAccessibilitySettingId || (TerminalAccessibilitySettingId = {}));
const terminalAccessibilityConfiguration = {
    [TerminalAccessibilitySettingId.AccessibleViewPreserveCursorPosition]: {
        markdownDescription: ( localize(
            11964,
            "Preserve the cursor position on reopen of the terminal's accessible view rather than setting it to the bottom of the buffer."
        )),
        type: 'boolean',
        default: false
    },
    [TerminalAccessibilitySettingId.AccessibleViewFocusOnCommandExecution]: {
        markdownDescription: ( localize(11965, "Focus the terminal accessible view when a command is executed.")),
        type: 'boolean',
        default: false
    },
};

export { TerminalAccessibilitySettingId, terminalAccessibilityConfiguration };
