// @codingame/monaco-vscode-view-common-service-override/vscode/src/vs/workbench/contrib/sash/browser/sash.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { clamp } from '@codingame/monaco-vscode-api/vscode/vs/base/common/numbers';
import { setGlobalSashSize, setGlobalHoverDelay } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/ui/sash/sash';
import { Event } from '@codingame/monaco-vscode-api/vscode/vs/base/common/event';
import { Disposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import { IConfigurationService } from '@codingame/monaco-vscode-api/vscode/vs/platform/configuration/common/configuration.service';
import { createStyleSheet } from '@codingame/monaco-vscode-api/vscode/vs/base/browser/domStylesheets';

let SashSettingsController = class SashSettingsController extends Disposable {
    static { this.ID = 'workbench.contrib.sash'; }
    constructor(configurationService) {
        super();
        this.configurationService = configurationService;
        this.styleSheet = createStyleSheet();
        const onDidChangeSize = Event.filter(configurationService.onDidChangeConfiguration, e => e.affectsConfiguration('workbench.sash.size'));
        onDidChangeSize(this.onDidChangeSize, this, this._store);
        this.onDidChangeSize();
        const onDidChangeHoverDelay = Event.filter(configurationService.onDidChangeConfiguration, e => e.affectsConfiguration('workbench.sash.hoverDelay'));
        onDidChangeHoverDelay(this.onDidChangeHoverDelay, this, this._store);
        this.onDidChangeHoverDelay();
    }
    onDidChangeSize() {
        const configuredSize = this.configurationService.getValue('workbench.sash.size');
        const size = clamp(configuredSize, 4, 20);
        const hoverSize = clamp(configuredSize, 1, 8);
        this.styleSheet.textContent = `
			.monaco-workbench {
				--vscode-sash-size: ${size}px;
				--vscode-sash-hover-size: ${hoverSize}px;
			}
		`;
        setGlobalSashSize(size);
    }
    onDidChangeHoverDelay() {
        setGlobalHoverDelay(this.configurationService.getValue('workbench.sash.hoverDelay'));
    }
};
SashSettingsController = ( __decorate([
    ( __param(0, IConfigurationService))
], SashSettingsController));

export { SashSettingsController };
