// @codingame/monaco-vscode-api/vscode/src/vs/workbench/api/common/extHost.api.impl.js

import { CancellationTokenSource } from '../../../base/common/cancellation.js';
import { onUnexpectedExternalError, ReadonlyError, CancellationError } from '../../../base/common/errors.js';
import { Event, Emitter } from '../../../base/common/event.js';
import { combinedDisposable } from '../../../base/common/lifecycle.js';
import { matchesScheme, Schemas } from '../../../base/common/network.js';
import Severity$1 from '../../../base/common/severity.js';
import { URI } from '../../../base/common/uri.js';
import { TextEditorCursorStyle } from '../../../editor/common/config/editorOptions.js';
import { targetsNotebooks, score } from '../../../editor/common/languageSelector.js';
import { IndentAction } from '../../../editor/common/languages/languageConfiguration.js';
import { OverviewRulerLane } from '../../../editor/common/model.js';
import { ExtensionError, ExtensionIdentifierSet } from '../../../platform/extensions/common/extensions.js';
import { FilePermission, FileType } from '../../../platform/files/common/files.js';
import { LogLevel } from '../../../platform/log/common/log.js';
import { ILoggerService, ILogService } from '../../../platform/log/common/log.service.js';
import { getRemoteName } from '../../../platform/remote/common/remoteHosts.js';
import { TelemetryTrustedValue } from '../../../platform/telemetry/common/telemetryUtils.js';
import { EditSessionIdentityMatch } from '../../../platform/workspace/common/editSessions.js';
import { DebugConfigurationProviderTriggerKind } from '../../contrib/debug/common/debug.js';
import { UIKind } from '../../services/extensions/common/extensionHostProtocol.js';
import { checkProposedApiEnabled, isProposedApiEnabled } from '../../services/extensions/common/extensions.js';
import { TextSearchCompleteMessageType, AISearchKeyword, TextSearchMatch2, TextSearchContext2, ExcludeSettingOptions } from '../../services/search/common/searchExtTypes.js';
import { ExtHostContext, MainContext, CandidatePortSource } from './extHost.protocol.js';
import { ExtHostRelatedInformation } from './extHostAiRelatedInformation.js';
import { ExtHostAiSettingsSearch } from './extHostAiSettingsSearch.js';
import { ExtHostApiCommands } from './extHostApiCommands.js';
import { IExtHostApiDeprecationService } from './extHostApiDeprecationService.js';
import { IExtHostAuthentication } from './extHostAuthentication.js';
import { ExtHostBulkEdits } from './extHostBulkEdits.js';
import { ExtHostChatAgents2 } from './extHostChatAgents2.js';
import { ExtHostChatOutputRenderer } from './extHostChatOutputRenderer.js';
import { ExtHostChatSessions } from './extHostChatSessions.js';
import { ExtHostChatStatus } from './extHostChatStatus.js';
import { ExtHostClipboard } from './extHostClipboard.js';
import { ExtHostEditorInsets } from './extHostCodeInsets.js';
import { ExtHostCodeMapper } from './extHostCodeMapper.js';
import { IExtHostCommands } from './extHostCommands.js';
import { createExtHostComments } from './extHostComments.js';
import { IExtHostConfiguration } from './extHostConfiguration.js';
import { ExtHostCustomEditors } from './extHostCustomEditors.js';
import { IExtHostDataChannels } from './extHostDataChannels.js';
import { IExtHostDebugService } from './extHostDebugService.js';
import { IExtHostDecorations } from './extHostDecorations.js';
import { ExtHostDiagnostics } from './extHostDiagnostics.js';
import { ExtHostDialogs } from './extHostDialogs.js';
import { ExtHostDocumentContentProvider } from './extHostDocumentContentProviders.js';
import { ExtHostDocumentSaveParticipant } from './extHostDocumentSaveParticipant.js';
import { ExtHostDocuments } from './extHostDocuments.js';
import { IExtHostDocumentsAndEditors } from './extHostDocumentsAndEditors.js';
import { IExtHostEditorTabs } from './extHostEditorTabs.js';
import { ExtHostEmbeddings } from './extHostEmbedding.js';
import { ExtHostAiEmbeddingVector } from './extHostEmbeddingVector.js';
import { IExtHostExtensionService, Extension } from './extHostExtensionService.js';
import { ExtHostFileSystem } from './extHostFileSystem.js';
import { IExtHostConsumerFileSystem } from './extHostFileSystemConsumer.js';
import { ExtHostFileSystemEventService } from './extHostFileSystemEventService.js';
import { IExtHostFileSystemInfo } from './extHostFileSystemInfo.js';
import { IExtHostInitDataService } from './extHostInitDataService.js';
import { ExtHostInteractive } from './extHostInteractive.js';
import { ExtHostLabelService } from './extHostLabelService.js';
import { ExtHostLanguageFeatures } from './extHostLanguageFeatures.js';
import { ExtHostLanguageModelTools } from './extHostLanguageModelTools.js';
import { IExtHostLanguageModels } from './extHostLanguageModels.js';
import { ExtHostLanguages } from './extHostLanguages.js';
import { IExtHostLocalizationService } from './extHostLocalizationService.js';
import { IExtHostManagedSockets } from './extHostManagedSockets.js';
import { IExtHostMpcService } from './extHostMcp.js';
import { ExtHostMessageService } from './extHostMessageService.js';
import { ExtHostNotebookController } from './extHostNotebook.js';
import { ExtHostNotebookDocumentSaveParticipant } from './extHostNotebookDocumentSaveParticipant.js';
import { ExtHostNotebookDocuments } from './extHostNotebookDocuments.js';
import { ExtHostNotebookEditors } from './extHostNotebookEditors.js';
import { ExtHostNotebookKernels } from './extHostNotebookKernels.js';
import { ExtHostNotebookRenderers } from './extHostNotebookRenderers.js';
import { IExtHostOutputService } from './extHostOutput.js';
import { ExtHostProfileContentHandlers } from './extHostProfileContentHandler.js';
import { IExtHostProgress } from './extHostProgress.js';
import { ExtHostQuickDiff } from './extHostQuickDiff.js';
import { createExtHostQuickOpen } from './extHostQuickOpen.js';
import { IExtHostRpcService } from './extHostRpcService.js';
import { ExtHostSCM } from './extHostSCM.js';
import { IExtHostSearch } from './extHostSearch.js';
import { IExtHostSecretState } from './extHostSecretState.js';
import { ExtHostShare } from './extHostShare.js';
import { ExtHostSpeech } from './extHostSpeech.js';
import { ExtHostStatusBar } from './extHostStatusBar.js';
import { IExtHostStorage } from './extHostStorage.js';
import { IExtensionStoragePaths } from './extHostStoragePaths.js';
import { IExtHostTask } from './extHostTask.js';
import { IExtHostTelemetry, ExtHostTelemetryLogger, isNewAppInstall } from './extHostTelemetry.js';
import { IExtHostTerminalService } from './extHostTerminalService.js';
import { IExtHostTerminalShellIntegration } from './extHostTerminalShellIntegration.js';
import { IExtHostTesting } from './extHostTesting.js';
import { ExtHostEditors } from './extHostTextEditors.js';
import { ExtHostTheming } from './extHostTheming.js';
import { ExtHostTimeline } from './extHostTimeline.js';
import { ExtHostTreeViews } from './extHostTreeViews.js';
import { IExtHostTunnelService } from './extHostTunnelService.js';
import { LanguageSelector } from './extHostTypeConverters.js';
import { ExtensionKind, ProgressLocation, SettingsSearchResultKind, McpToolAvailability, McpStdioServerDefinition, McpHttpServerDefinition, ChatErrorLevel, NewSymbolNameTriggerKind, NewSymbolNameTag, NewSymbolName, LanguageModelPromptTsxPart, LanguageModelChatToolMode, ExtendedLanguageModelToolResult, LanguageModelToolMCPSource, LanguageModelToolExtensionSource, LanguageModelDataPart, LanguageModelToolResult2, LanguageModelToolResult, LanguageModelError, LanguageModelThinkingPart, LanguageModelToolCallPart, LanguageModelPartAudience, LanguageModelTextPart, LanguageModelToolResultPart, LanguageModelChatMessage2, LanguageModelChatMessage, LanguageModelChatMessageRole, ChatRequestEditedFileEventKind, ChatReferenceBinaryData, ChatRequestNotebookData, ChatRequestEditorData, ChatSessionStatus, ChatLocation, ChatToolInvocationPart, ChatResponseTurn2, ChatResponseTurn, ChatRequestTurn, ChatResponseClearToPreviousToolInvocationReason, ChatResponseReferencePartStatusKind, ChatResponseMultiDiffPart, ChatPrepareToolInvocationPart, ChatResponsePullRequestPart, ChatResponseExternalEditPart, ChatResponseExtensionsPart, ChatResponseMovePart, ChatResponseConfirmationPart, ChatResponseCommandButtonPart, ChatResponseMarkdownWithVulnerabilitiesPart, ChatResponseNotebookEditPart, ChatResponseTextEditPart, ChatResponseWarningPart, ChatResponseCodeblockUriPart, ChatResponseCodeCitationPart, ChatResponseReferencePart, ChatResponseThinkingProgressPart, ChatResponseProgressPart2, ChatResponseProgressPart, ChatResponseAnchorPart, ChatResponseFileTreePart, ChatResponseMarkdownPart, ChatImageMimeType, KeywordRecognitionStatus, InlineCompletionDisplayLocationKind, InlineCompletionEndOfLifeReasonKind, PartialAcceptTriggerKind, TextToSpeechStatus, SpeechToTextStatus, RelatedInformationType, DebugThread, DebugStackFrame, InteractiveEditorResponseFeedbackKind, ChatEditingSessionActionOutcome, ChatSessionChangedFile, ChatCopyKind, InteractiveSessionVoteDirection, TextMultiDiffTabInput, ChatEditorTabInput, InteractiveWindowInput, TerminalEditorTabInput, WebviewEditorTabInput, NotebookDiffEditorTabInput, NotebookEditorTabInput, CustomEditorTabInput, TextMergeTabInput, TextDiffTabInput, TextTabInput, InputBoxValidationSeverity, QuickPickItemKind, LanguageStatusSeverity, WorkspaceTrustState, DeclarationCoverage, BranchCoverage, StatementCoverage, FileCoverage, TestCoverageCount, DataTransferItem, DataTransfer, TestRunProfileKind, TestTag, TestMessageStackFrame, TestMessage, TestRunRequest, TestResultState, LinkedEditingRanges, PortAttributes, NotebookVariablesRequestKind, NotebookKernelSourceAction, NotebookControllerAffinity2, NotebookControllerAffinity, NotebookCellStatusBarItem, CellErrorStackFrame, NotebookEditorRevealType, NotebookCellStatusBarAlignment, NotebookRendererScript, NotebookCellExecutionState, TimelineItem, ExtensionRuntime, SourceControlInputBoxValidationType, ManagedResolvedAuthority, ResolvedAuthority, RemoteAuthorityResolverError, InlayHintKind, InlayHintLabelPart, InlayHint, DocumentPasteEdit, DocumentDropOrPasteEditKind, DocumentDropEdit, DocumentPasteTriggerKind, ViewColumn, TypeHierarchyItem, TreeItemCollapsibleState, TreeItemCheckboxState, TreeItem, ThemeIcon, ThemeColor, TextDocumentChangeReason, SyntaxTokenType, TextEditorSelectionChangeKind, TextEditorRevealType, TextEditorLineNumbersStyle, TextEditorChangeKind, TextDocumentSaveReason, TerminalShellType, TerminalCompletionList, TerminalCompletionItemKind, TerminalCompletionItem, TerminalShellExecutionCommandLineConfidence, TerminalExitReason, TerminalProfile, TerminalLocation, TerminalQuickFixOpener, TerminalQuickFixCommand, TerminalLink, TaskScope, TaskRevealKind, TaskPanelKind, TaskGroup, TaskEventKind, Task, StatusBarAlignment, StandardTokenType, SourceBreakpoint, SignatureInformation, SignatureHelpTriggerKind, SignatureHelp, ShellQuoting, ShellExecution, SemanticTokensLegend, SemanticTokensEdits, SemanticTokensEdit, SemanticTokensBuilder, SemanticTokens, SelectionRange, RelativePattern, QuickInputButtons, QuickInputButtonLocation, ProcessExecution, PortAutoForwardAction, ParameterInformation, HoverVerbosityAction, VerboseHover, Hover, InlineSuggestionList, InlineSuggestion, FunctionBreakpoint, FoldingRangeKind, FoldingRange, FileSystemError, FileDecoration, FileChangeType, ExternalUriOpenerPriority, ExtensionMode, InlineCompletionsDisposeReasonKind, InlineCompletionTriggerKind, InlineValueEvaluatableExpression, InlineValueVariableLookup, InlineValueText, EvaluatableExpression, EnvironmentVariableMutatorType, DocumentSymbol, DocumentLink, MultiDocumentHighlight, DocumentHighlightKind, DocumentHighlight, Disposable, DecorationRangeBehavior, DebugVisualization, DebugConsoleMode, DebugAdapterServer, DebugAdapterNamedPipeServer, DebugAdapterInlineImplementation, DebugAdapterExecutable, CustomExecution, ConfigurationTarget, CompletionTriggerKind, CompletionList, CompletionItemTag, CompletionItemKind, CompletionItem, CommentThreadFocus, CommentThreadApplicability, CommentThreadState, CommentThreadCollapsibleState, CommentState, CommentMode, ColorThemeKind, ColorPresentation, ColorInformation, Color, CodeLens, CodeActionTriggerKind, CodeAction, CallHierarchyOutgoingCall, CallHierarchyItem, CallHierarchyIncomingCall, ChatReferenceDiagnostic, ChatCompletionItem, ChatVariableLevel, ChatResultFeedbackKind, TerminalOutputAnchor, Breakpoint } from './extHostTypes.js';
import { ExtHostUriOpeners } from './extHostUriOpener.js';
import { IURITransformerService } from './extHostUriTransformerService.js';
import { IExtHostUrlsService } from './extHostUrls.js';
import { ExtHostWebviews } from './extHostWebview.js';
import { ExtHostWebviewPanels } from './extHostWebviewPanels.js';
import { ExtHostWebviewViews } from './extHostWebviewView.js';
import { IExtHostWindow } from './extHostWindow.js';
import { IExtHostWorkspace } from './extHostWorkspace.js';
import { ExtHostChatContext } from './extHostChatContext.js';
import { NotebookEdit, NotebookCellOutputItem, NotebookCellOutput, NotebookData, NotebookCellData, NotebookCellKind, NotebookRange } from './extHostTypes/notebooks.js';
import { WorkspaceEdit } from './extHostTypes/workspaceEdit.js';
import { SnippetTextEdit } from './extHostTypes/snippetTextEdit.js';
import { TextEdit, EndOfLine } from './extHostTypes/textEdit.js';
import { SymbolTag, SymbolKind, SymbolInformation } from './extHostTypes/symbolInformation.js';
import { SnippetString } from './extHostTypes/snippetString.js';
import { Selection } from './extHostTypes/selection.js';
import { Range } from './extHostTypes/range.js';
import { Position } from './extHostTypes/position.js';
import { MarkdownString } from './extHostTypes/markdownString.js';
import { Location } from './extHostTypes/location.js';
import { DiagnosticTag, DiagnosticSeverity, DiagnosticRelatedInformation, Diagnostic } from './extHostTypes/diagnostic.js';
import { CodeActionKind } from './extHostTypes/codeActionKind.js';

function createApiFactoryAndRegisterActors(accessor) {
    const initData = accessor.get(IExtHostInitDataService);
    const extHostFileSystemInfo = accessor.get(IExtHostFileSystemInfo);
    const extHostConsumerFileSystem = accessor.get(IExtHostConsumerFileSystem);
    const extensionService = accessor.get(IExtHostExtensionService);
    const extHostWorkspace = accessor.get(IExtHostWorkspace);
    const extHostTelemetry = accessor.get(IExtHostTelemetry);
    const extHostConfiguration = accessor.get(IExtHostConfiguration);
    const uriTransformer = accessor.get(IURITransformerService);
    const rpcProtocol = accessor.get(IExtHostRpcService);
    const extHostStorage = accessor.get(IExtHostStorage);
    const extensionStoragePaths = accessor.get(IExtensionStoragePaths);
    const extHostLoggerService = accessor.get(ILoggerService);
    const extHostLogService = accessor.get(ILogService);
    const extHostTunnelService = accessor.get(IExtHostTunnelService);
    const extHostApiDeprecation = accessor.get(IExtHostApiDeprecationService);
    const extHostWindow = accessor.get(IExtHostWindow);
    const extHostUrls = accessor.get(IExtHostUrlsService);
    const extHostSecretState = accessor.get(IExtHostSecretState);
    const extHostEditorTabs = accessor.get(IExtHostEditorTabs);
    const extHostManagedSockets = accessor.get(IExtHostManagedSockets);
    const extHostProgress = accessor.get(IExtHostProgress);
    const extHostAuthentication = accessor.get(IExtHostAuthentication);
    const extHostLanguageModels = accessor.get(IExtHostLanguageModels);
    const extHostMcp = accessor.get(IExtHostMpcService);
    const extHostDataChannels = accessor.get(IExtHostDataChannels);
    rpcProtocol.set(ExtHostContext.ExtHostFileSystemInfo, extHostFileSystemInfo);
    rpcProtocol.set(ExtHostContext.ExtHostLogLevelServiceShape, extHostLoggerService);
    rpcProtocol.set(ExtHostContext.ExtHostWorkspace, extHostWorkspace);
    rpcProtocol.set(ExtHostContext.ExtHostConfiguration, extHostConfiguration);
    rpcProtocol.set(ExtHostContext.ExtHostExtensionService, extensionService);
    rpcProtocol.set(ExtHostContext.ExtHostStorage, extHostStorage);
    rpcProtocol.set(ExtHostContext.ExtHostTunnelService, extHostTunnelService);
    rpcProtocol.set(ExtHostContext.ExtHostWindow, extHostWindow);
    rpcProtocol.set(ExtHostContext.ExtHostUrls, extHostUrls);
    rpcProtocol.set(ExtHostContext.ExtHostSecretState, extHostSecretState);
    rpcProtocol.set(ExtHostContext.ExtHostTelemetry, extHostTelemetry);
    rpcProtocol.set(ExtHostContext.ExtHostEditorTabs, extHostEditorTabs);
    rpcProtocol.set(ExtHostContext.ExtHostManagedSockets, extHostManagedSockets);
    rpcProtocol.set(ExtHostContext.ExtHostProgress, extHostProgress);
    rpcProtocol.set(ExtHostContext.ExtHostAuthentication, extHostAuthentication);
    rpcProtocol.set(ExtHostContext.ExtHostChatProvider, extHostLanguageModels);
    rpcProtocol.set(ExtHostContext.ExtHostDataChannels, extHostDataChannels);
    const extHostDecorations = rpcProtocol.set(ExtHostContext.ExtHostDecorations, accessor.get(IExtHostDecorations));
    const extHostDocumentsAndEditors = rpcProtocol.set(ExtHostContext.ExtHostDocumentsAndEditors, accessor.get(IExtHostDocumentsAndEditors));
    const extHostCommands = rpcProtocol.set(ExtHostContext.ExtHostCommands, accessor.get(IExtHostCommands));
    const extHostTerminalService = rpcProtocol.set(ExtHostContext.ExtHostTerminalService, accessor.get(IExtHostTerminalService));
    const extHostTerminalShellIntegration = rpcProtocol.set(ExtHostContext.ExtHostTerminalShellIntegration, accessor.get(IExtHostTerminalShellIntegration));
    const extHostDebugService = rpcProtocol.set(ExtHostContext.ExtHostDebugService, accessor.get(IExtHostDebugService));
    const extHostSearch = rpcProtocol.set(ExtHostContext.ExtHostSearch, accessor.get(IExtHostSearch));
    const extHostTask = rpcProtocol.set(ExtHostContext.ExtHostTask, accessor.get(IExtHostTask));
    const extHostOutputService = rpcProtocol.set(ExtHostContext.ExtHostOutputService, accessor.get(IExtHostOutputService));
    const extHostLocalization = rpcProtocol.set(ExtHostContext.ExtHostLocalization, accessor.get(IExtHostLocalizationService));
    const extHostDocuments = rpcProtocol.set(ExtHostContext.ExtHostDocuments, ( new ExtHostDocuments(rpcProtocol, extHostDocumentsAndEditors)));
    const extHostDocumentContentProviders = rpcProtocol.set(ExtHostContext.ExtHostDocumentContentProviders, ( new ExtHostDocumentContentProvider(rpcProtocol, extHostDocumentsAndEditors, extHostLogService)));
    const extHostDocumentSaveParticipant = rpcProtocol.set(ExtHostContext.ExtHostDocumentSaveParticipant, ( new ExtHostDocumentSaveParticipant(extHostLogService, extHostDocuments, ( rpcProtocol.getProxy(MainContext.MainThreadBulkEdits)))));
    const extHostNotebook = rpcProtocol.set(ExtHostContext.ExtHostNotebook, ( new ExtHostNotebookController(
        rpcProtocol,
        extHostCommands,
        extHostDocumentsAndEditors,
        extHostDocuments,
        extHostConsumerFileSystem,
        extHostSearch,
        extHostLogService
    )));
    const extHostNotebookDocuments = rpcProtocol.set(ExtHostContext.ExtHostNotebookDocuments, ( new ExtHostNotebookDocuments(extHostNotebook)));
    const extHostNotebookEditors = rpcProtocol.set(ExtHostContext.ExtHostNotebookEditors, ( new ExtHostNotebookEditors(extHostLogService, extHostNotebook)));
    const extHostNotebookKernels = rpcProtocol.set(ExtHostContext.ExtHostNotebookKernels, ( new ExtHostNotebookKernels(rpcProtocol, initData, extHostNotebook, extHostCommands, extHostLogService)));
    const extHostNotebookRenderers = rpcProtocol.set(ExtHostContext.ExtHostNotebookRenderers, ( new ExtHostNotebookRenderers(rpcProtocol, extHostNotebook)));
    const extHostNotebookDocumentSaveParticipant = rpcProtocol.set(ExtHostContext.ExtHostNotebookDocumentSaveParticipant, ( new ExtHostNotebookDocumentSaveParticipant(extHostLogService, extHostNotebook, ( rpcProtocol.getProxy(MainContext.MainThreadBulkEdits)))));
    const extHostEditors = rpcProtocol.set(ExtHostContext.ExtHostEditors, ( new ExtHostEditors(rpcProtocol, extHostDocumentsAndEditors)));
    const extHostTreeViews = rpcProtocol.set(ExtHostContext.ExtHostTreeViews, ( new ExtHostTreeViews(( rpcProtocol.getProxy(MainContext.MainThreadTreeViews)), extHostCommands, extHostLogService)));
    const extHostEditorInsets = rpcProtocol.set(ExtHostContext.ExtHostEditorInsets, ( new ExtHostEditorInsets(( rpcProtocol.getProxy(MainContext.MainThreadEditorInsets)), extHostEditors, initData.remote)));
    const extHostDiagnostics = rpcProtocol.set(ExtHostContext.ExtHostDiagnostics, ( new ExtHostDiagnostics(
        rpcProtocol,
        extHostLogService,
        extHostFileSystemInfo,
        extHostDocumentsAndEditors
    )));
    const extHostLanguages = rpcProtocol.set(ExtHostContext.ExtHostLanguages, ( new ExtHostLanguages(rpcProtocol, extHostDocuments, extHostCommands.converter, uriTransformer)));
    const extHostLanguageFeatures = rpcProtocol.set(ExtHostContext.ExtHostLanguageFeatures, ( new ExtHostLanguageFeatures(
        rpcProtocol,
        uriTransformer,
        extHostDocuments,
        extHostCommands,
        extHostDiagnostics,
        extHostLogService,
        extHostApiDeprecation,
        extHostTelemetry
    )));
    const extHostCodeMapper = rpcProtocol.set(ExtHostContext.ExtHostCodeMapper, ( new ExtHostCodeMapper(rpcProtocol)));
    const extHostFileSystem = rpcProtocol.set(ExtHostContext.ExtHostFileSystem, ( new ExtHostFileSystem(rpcProtocol, extHostLanguageFeatures)));
    const extHostFileSystemEvent = rpcProtocol.set(ExtHostContext.ExtHostFileSystemEventService, ( new ExtHostFileSystemEventService(rpcProtocol, extHostLogService, extHostDocumentsAndEditors)));
    const extHostQuickOpen = rpcProtocol.set(ExtHostContext.ExtHostQuickOpen, createExtHostQuickOpen(rpcProtocol, extHostWorkspace, extHostCommands));
    const extHostSCM = rpcProtocol.set(ExtHostContext.ExtHostSCM, ( new ExtHostSCM(rpcProtocol, extHostCommands, extHostDocuments, extHostLogService)));
    const extHostQuickDiff = rpcProtocol.set(ExtHostContext.ExtHostQuickDiff, ( new ExtHostQuickDiff(rpcProtocol, uriTransformer)));
    const extHostShare = rpcProtocol.set(ExtHostContext.ExtHostShare, ( new ExtHostShare(rpcProtocol, uriTransformer)));
    const extHostComment = rpcProtocol.set(ExtHostContext.ExtHostComments, createExtHostComments(rpcProtocol, extHostCommands, extHostDocuments));
    const extHostLabelService = rpcProtocol.set(ExtHostContext.ExtHostLabelService, ( new ExtHostLabelService(rpcProtocol)));
    const extHostTheming = rpcProtocol.set(ExtHostContext.ExtHostTheming, ( new ExtHostTheming(rpcProtocol)));
    const extHostTimeline = rpcProtocol.set(ExtHostContext.ExtHostTimeline, ( new ExtHostTimeline(rpcProtocol, extHostCommands)));
    const extHostWebviews = rpcProtocol.set(ExtHostContext.ExtHostWebviews, ( new ExtHostWebviews(
        rpcProtocol,
        initData.remote,
        extHostWorkspace,
        extHostLogService,
        extHostApiDeprecation
    )));
    const extHostWebviewPanels = rpcProtocol.set(ExtHostContext.ExtHostWebviewPanels, ( new ExtHostWebviewPanels(rpcProtocol, extHostWebviews, extHostWorkspace)));
    const extHostCustomEditors = rpcProtocol.set(ExtHostContext.ExtHostCustomEditors, ( new ExtHostCustomEditors(
        rpcProtocol,
        extHostDocuments,
        extensionStoragePaths,
        extHostWebviews,
        extHostWebviewPanels
    )));
    const extHostWebviewViews = rpcProtocol.set(ExtHostContext.ExtHostWebviewViews, ( new ExtHostWebviewViews(rpcProtocol, extHostWebviews)));
    const extHostTesting = rpcProtocol.set(ExtHostContext.ExtHostTesting, accessor.get(IExtHostTesting));
    const extHostUriOpeners = rpcProtocol.set(ExtHostContext.ExtHostUriOpeners, ( new ExtHostUriOpeners(rpcProtocol)));
    const extHostProfileContentHandlers = rpcProtocol.set(ExtHostContext.ExtHostProfileContentHandlers, ( new ExtHostProfileContentHandlers(rpcProtocol)));
    const extHostChatOutputRenderer = rpcProtocol.set(ExtHostContext.ExtHostChatOutputRenderer, ( new ExtHostChatOutputRenderer(rpcProtocol, extHostWebviews)));
    rpcProtocol.set(ExtHostContext.ExtHostInteractive, ( new ExtHostInteractive(
        rpcProtocol,
        extHostNotebook,
        extHostDocumentsAndEditors,
        extHostCommands,
        extHostLogService
    )));
    const extHostLanguageModelTools = rpcProtocol.set(ExtHostContext.ExtHostLanguageModelTools, ( new ExtHostLanguageModelTools(rpcProtocol, extHostLanguageModels)));
    const extHostChatSessions = rpcProtocol.set(ExtHostContext.ExtHostChatSessions, ( new ExtHostChatSessions(extHostCommands, extHostLanguageModels, rpcProtocol, extHostLogService)));
    const extHostChatAgents2 = rpcProtocol.set(ExtHostContext.ExtHostChatAgents2, ( new ExtHostChatAgents2(
        rpcProtocol,
        extHostLogService,
        extHostCommands,
        extHostDocuments,
        extHostDocumentsAndEditors,
        extHostLanguageModels,
        extHostDiagnostics,
        extHostLanguageModelTools
    )));
    const extHostChatContext = rpcProtocol.set(ExtHostContext.ExtHostChatContext, ( new ExtHostChatContext(rpcProtocol)));
    const extHostAiRelatedInformation = rpcProtocol.set(ExtHostContext.ExtHostAiRelatedInformation, ( new ExtHostRelatedInformation(rpcProtocol)));
    const extHostAiEmbeddingVector = rpcProtocol.set(ExtHostContext.ExtHostAiEmbeddingVector, ( new ExtHostAiEmbeddingVector(rpcProtocol)));
    const extHostAiSettingsSearch = rpcProtocol.set(ExtHostContext.ExtHostAiSettingsSearch, ( new ExtHostAiSettingsSearch(rpcProtocol)));
    const extHostStatusBar = rpcProtocol.set(ExtHostContext.ExtHostStatusBar, ( new ExtHostStatusBar(rpcProtocol, extHostCommands.converter)));
    const extHostSpeech = rpcProtocol.set(ExtHostContext.ExtHostSpeech, ( new ExtHostSpeech(rpcProtocol)));
    const extHostEmbeddings = rpcProtocol.set(ExtHostContext.ExtHostEmbeddings, ( new ExtHostEmbeddings(rpcProtocol)));
    rpcProtocol.set(ExtHostContext.ExtHostMcp, accessor.get(IExtHostMpcService));
    const expected = ( Object.values(ExtHostContext));
    rpcProtocol.assertRegistered(expected);
    const extHostBulkEdits = ( new ExtHostBulkEdits(rpcProtocol, extHostDocumentsAndEditors));
    const extHostClipboard = ( new ExtHostClipboard(rpcProtocol));
    const extHostMessageService = ( new ExtHostMessageService(rpcProtocol, extHostLogService));
    const extHostDialogs = ( new ExtHostDialogs(rpcProtocol));
    const extHostChatStatus = ( new ExtHostChatStatus(rpcProtocol));
    ExtHostApiCommands.register(extHostCommands);
    return function (extension, extensionInfo, configProvider) {
        function _asExtensionEvent(actual) {
            return (listener, thisArgs, disposables) => {
                const handle = actual(e => {
                    try {
                        listener.call(thisArgs, e);
                    }
                    catch (err) {
                        onUnexpectedExternalError(( new ExtensionError(extension.identifier, err, 'FAILED to handle event')));
                    }
                });
                disposables?.push(handle);
                return handle;
            };
        }
        const checkSelector = (function () {
            let done = !extension.isUnderDevelopment;
            function informOnce() {
                if (!done) {
                    extHostLogService.info(`Extension '${extension.identifier.value}' uses a document selector without scheme. Learn more about this: https://go.microsoft.com/fwlink/?linkid=872305`);
                    done = true;
                }
            }
            return function perform(selector) {
                if (Array.isArray(selector)) {
                    selector.forEach(perform);
                }
                else if (typeof selector === 'string') {
                    informOnce();
                }
                else {
                    const filter = selector;
                    if (typeof filter.scheme === 'undefined') {
                        informOnce();
                    }
                    if (typeof filter.exclusive === 'boolean') {
                        checkProposedApiEnabled(extension, 'documentFiltersExclusive');
                    }
                }
                return selector;
            };
        })();
        const authentication = {
            getSession(providerId, scopesOrChallenge, options) {
                if ((typeof options?.forceNewSession === 'object' && options.forceNewSession.learnMore) ||
                    (typeof options?.createIfNone === 'object' && options.createIfNone.learnMore)) {
                    checkProposedApiEnabled(extension, 'authLearnMore');
                }
                if (options?.authorizationServer) {
                    checkProposedApiEnabled(extension, 'authIssuers');
                }
                return extHostAuthentication.getSession(extension, providerId, scopesOrChallenge, options);
            },
            getAccounts(providerId) {
                return extHostAuthentication.getAccounts(providerId);
            },
            async hasSession(providerId, scopes) {
                checkProposedApiEnabled(extension, 'authSession');
                return !!(await extHostAuthentication.getSession(extension, providerId, scopes, { silent: true }));
            },
            get onDidChangeSessions() {
                return _asExtensionEvent(extHostAuthentication.getExtensionScopedSessionsEvent(extension.identifier.value));
            },
            registerAuthenticationProvider(id, label, provider, options) {
                if (options?.supportedAuthorizationServers) {
                    checkProposedApiEnabled(extension, 'authIssuers');
                }
                return extHostAuthentication.registerAuthenticationProvider(id, label, provider, options);
            }
        };
        const commands = {
            registerCommand(id, command, thisArgs) {
                return extHostCommands.registerCommand(true, id, command, thisArgs, undefined, extension);
            },
            registerTextEditorCommand(id, callback, thisArg) {
                return extHostCommands.registerCommand(true, id, (...args) => {
                    const activeTextEditor = extHostEditors.getActiveTextEditor();
                    if (!activeTextEditor) {
                        extHostLogService.warn('Cannot execute ' + id + ' because there is no active text editor.');
                        return undefined;
                    }
                    return activeTextEditor.edit((edit) => {
                        callback.apply(thisArg, [activeTextEditor, edit, ...args]);
                    }).then((result) => {
                        if (!result) {
                            extHostLogService.warn('Edits from command ' + id + ' were not applied.');
                        }
                    }, (err) => {
                        extHostLogService.warn('An error occurred while running command ' + id, err);
                    });
                }, undefined, undefined, extension);
            },
            registerDiffInformationCommand: (id, callback, thisArg) => {
                checkProposedApiEnabled(extension, 'diffCommand');
                return extHostCommands.registerCommand(true, id, async (...args) => {
                    const activeTextEditor = extHostDocumentsAndEditors.activeEditor(true);
                    if (!activeTextEditor) {
                        extHostLogService.warn('Cannot execute ' + id + ' because there is no active text editor.');
                        return undefined;
                    }
                    const diff = await extHostEditors.getDiffInformation(activeTextEditor.id);
                    callback.apply(thisArg, [diff, ...args]);
                }, undefined, undefined, extension);
            },
            executeCommand(id, ...args) {
                return extHostCommands.executeCommand(id, ...args);
            },
            getCommands(filterInternal = false) {
                return extHostCommands.getCommands(filterInternal);
            }
        };
        const env = {
            get machineId() { return initData.telemetryInfo.machineId; },
            get devDeviceId() {
                checkProposedApiEnabled(extension, 'devDeviceId');
                return initData.telemetryInfo.devDeviceId ?? initData.telemetryInfo.machineId;
            },
            get sessionId() { return initData.telemetryInfo.sessionId; },
            get language() { return initData.environment.appLanguage; },
            get appName() { return initData.environment.appName; },
            get appRoot() { return initData.environment.appRoot?.fsPath ?? ''; },
            get appHost() { return initData.environment.appHost; },
            get uriScheme() { return initData.environment.appUriScheme; },
            get clipboard() { return extHostClipboard.value; },
            get shell() {
                return extHostTerminalService.getDefaultShell(false);
            },
            get onDidChangeShell() {
                return _asExtensionEvent(extHostTerminalService.onDidChangeShell);
            },
            get isTelemetryEnabled() {
                return extHostTelemetry.getTelemetryConfiguration();
            },
            get onDidChangeTelemetryEnabled() {
                return _asExtensionEvent(extHostTelemetry.onDidChangeTelemetryEnabled);
            },
            get telemetryConfiguration() {
                checkProposedApiEnabled(extension, 'telemetry');
                return extHostTelemetry.getTelemetryDetails();
            },
            get onDidChangeTelemetryConfiguration() {
                checkProposedApiEnabled(extension, 'telemetry');
                return _asExtensionEvent(extHostTelemetry.onDidChangeTelemetryConfiguration);
            },
            get isNewAppInstall() {
                return isNewAppInstall(initData.telemetryInfo.firstSessionDate);
            },
            createTelemetryLogger(sender, options) {
                ExtHostTelemetryLogger.validateSender(sender);
                return extHostTelemetry.instantiateLogger(extension, sender, options);
            },
            async openExternal(uri, options) {
                return extHostWindow.openUri(uri, {
                    allowTunneling: initData.remote.isRemote ?? (initData.remote.authority ? await extHostTunnelService.hasTunnelProvider() : false),
                    allowContributedOpeners: options?.allowContributedOpeners,
                });
            },
            async asExternalUri(uri) {
                if (uri.scheme === initData.environment.appUriScheme) {
                    return extHostUrls.createAppUri(uri);
                }
                try {
                    return await extHostWindow.asExternalUri(uri, { allowTunneling: !!initData.remote.authority });
                }
                catch (err) {
                    if (matchesScheme(uri, Schemas.http) || matchesScheme(uri, Schemas.https)) {
                        return uri;
                    }
                    throw err;
                }
            },
            get remoteName() {
                return getRemoteName(initData.remote.authority);
            },
            get remoteAuthority() {
                checkProposedApiEnabled(extension, 'resolvers');
                return initData.remote.authority;
            },
            get uiKind() {
                return initData.uiKind;
            },
            get logLevel() {
                return extHostLogService.getLevel();
            },
            get onDidChangeLogLevel() {
                return _asExtensionEvent(extHostLogService.onDidChangeLogLevel);
            },
            get appQuality() {
                checkProposedApiEnabled(extension, 'resolvers');
                return initData.quality;
            },
            get appCommit() {
                checkProposedApiEnabled(extension, 'resolvers');
                return initData.commit;
            },
            getDataChannel(channelId) {
                checkProposedApiEnabled(extension, 'dataChannels');
                return extHostDataChannels.createDataChannel(extension, channelId);
            }
        };
        const tests = {
            createTestController(provider, label, refreshHandler) {
                return extHostTesting.createTestController(extension, provider, label, refreshHandler);
            },
            createTestObserver() {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.createTestObserver();
            },
            runTests(provider) {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.runTests(provider);
            },
            registerTestFollowupProvider(provider) {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.registerTestFollowupProvider(provider);
            },
            get onDidChangeTestResults() {
                checkProposedApiEnabled(extension, 'testObserver');
                return _asExtensionEvent(extHostTesting.onResultsChanged);
            },
            get testResults() {
                checkProposedApiEnabled(extension, 'testObserver');
                return extHostTesting.results;
            },
        };
        const extensionKind = initData.remote.isRemote
            ? ExtensionKind.Workspace
            : ExtensionKind.UI;
        const extensions = {
            getExtension(extensionId, includeFromDifferentExtensionHosts) {
                if (!isProposedApiEnabled(extension, 'extensionsAny')) {
                    includeFromDifferentExtensionHosts = false;
                }
                const mine = extensionInfo.mine.getExtensionDescription(extensionId);
                if (mine) {
                    return ( new Extension(extensionService, extension.identifier, mine, extensionKind, false));
                }
                if (includeFromDifferentExtensionHosts) {
                    const foreign = extensionInfo.all.getExtensionDescription(extensionId);
                    if (foreign) {
                        return ( new Extension(
                            extensionService,
                            extension.identifier,
                            foreign,
                            extensionKind ,
                            true
                        ));
                    }
                }
                return undefined;
            },
            get all() {
                const result = [];
                for (const desc of extensionInfo.mine.getAllExtensionDescriptions()) {
                    result.push(( new Extension(extensionService, extension.identifier, desc, extensionKind, false)));
                }
                return result;
            },
            get allAcrossExtensionHosts() {
                checkProposedApiEnabled(extension, 'extensionsAny');
                const local = ( new ExtensionIdentifierSet(( extensionInfo.mine.getAllExtensionDescriptions().map(desc => desc.identifier))));
                const result = [];
                for (const desc of extensionInfo.all.getAllExtensionDescriptions()) {
                    const isFromDifferentExtensionHost = !( local.has(desc.identifier));
                    result.push(( new Extension(
                        extensionService,
                        extension.identifier,
                        desc,
                        extensionKind ,
                        isFromDifferentExtensionHost
                    )));
                }
                return result;
            },
            get onDidChange() {
                if (isProposedApiEnabled(extension, 'extensionsAny')) {
                    return _asExtensionEvent(Event.any(extensionInfo.mine.onDidChange, extensionInfo.all.onDidChange));
                }
                return _asExtensionEvent(extensionInfo.mine.onDidChange);
            }
        };
        const languages = {
            createDiagnosticCollection(name) {
                return extHostDiagnostics.createDiagnosticCollection(extension.identifier, name);
            },
            get onDidChangeDiagnostics() {
                return _asExtensionEvent(extHostDiagnostics.onDidChangeDiagnostics);
            },
            getDiagnostics: (resource) => {
                return extHostDiagnostics.getDiagnostics(resource);
            },
            getLanguages() {
                return extHostLanguages.getLanguages();
            },
            setTextDocumentLanguage(document, languageId) {
                return extHostLanguages.changeLanguage(document.uri, languageId);
            },
            match(selector, document) {
                const interalSelector = LanguageSelector.from(selector);
                let notebook;
                if (targetsNotebooks(interalSelector)) {
                    notebook = extHostNotebook.notebookDocuments.find(value => value.apiNotebook.getCells().find(c => c.document === document))?.apiNotebook;
                }
                return score(interalSelector, document.uri, document.languageId, true, notebook?.uri, notebook?.notebookType);
            },
            registerCodeActionsProvider(selector, provider, metadata) {
                return extHostLanguageFeatures.registerCodeActionProvider(extension, checkSelector(selector), provider, metadata);
            },
            registerDocumentPasteEditProvider(selector, provider, metadata) {
                return extHostLanguageFeatures.registerDocumentPasteEditProvider(extension, checkSelector(selector), provider, metadata);
            },
            registerCodeLensProvider(selector, provider) {
                return extHostLanguageFeatures.registerCodeLensProvider(extension, checkSelector(selector), provider);
            },
            registerDefinitionProvider(selector, provider) {
                return extHostLanguageFeatures.registerDefinitionProvider(extension, checkSelector(selector), provider);
            },
            registerDeclarationProvider(selector, provider) {
                return extHostLanguageFeatures.registerDeclarationProvider(extension, checkSelector(selector), provider);
            },
            registerImplementationProvider(selector, provider) {
                return extHostLanguageFeatures.registerImplementationProvider(extension, checkSelector(selector), provider);
            },
            registerTypeDefinitionProvider(selector, provider) {
                return extHostLanguageFeatures.registerTypeDefinitionProvider(extension, checkSelector(selector), provider);
            },
            registerHoverProvider(selector, provider) {
                return extHostLanguageFeatures.registerHoverProvider(extension, checkSelector(selector), provider, extension.identifier);
            },
            registerEvaluatableExpressionProvider(selector, provider) {
                return extHostLanguageFeatures.registerEvaluatableExpressionProvider(extension, checkSelector(selector), provider, extension.identifier);
            },
            registerInlineValuesProvider(selector, provider) {
                return extHostLanguageFeatures.registerInlineValuesProvider(extension, checkSelector(selector), provider, extension.identifier);
            },
            registerDocumentHighlightProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentHighlightProvider(extension, checkSelector(selector), provider);
            },
            registerMultiDocumentHighlightProvider(selector, provider) {
                return extHostLanguageFeatures.registerMultiDocumentHighlightProvider(extension, checkSelector(selector), provider);
            },
            registerLinkedEditingRangeProvider(selector, provider) {
                return extHostLanguageFeatures.registerLinkedEditingRangeProvider(extension, checkSelector(selector), provider);
            },
            registerReferenceProvider(selector, provider) {
                return extHostLanguageFeatures.registerReferenceProvider(extension, checkSelector(selector), provider);
            },
            registerRenameProvider(selector, provider) {
                return extHostLanguageFeatures.registerRenameProvider(extension, checkSelector(selector), provider);
            },
            registerNewSymbolNamesProvider(selector, provider) {
                checkProposedApiEnabled(extension, 'newSymbolNamesProvider');
                return extHostLanguageFeatures.registerNewSymbolNamesProvider(extension, checkSelector(selector), provider);
            },
            registerDocumentSymbolProvider(selector, provider, metadata) {
                return extHostLanguageFeatures.registerDocumentSymbolProvider(extension, checkSelector(selector), provider, metadata);
            },
            registerWorkspaceSymbolProvider(provider) {
                return extHostLanguageFeatures.registerWorkspaceSymbolProvider(extension, provider);
            },
            registerDocumentFormattingEditProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentFormattingEditProvider(extension, checkSelector(selector), provider);
            },
            registerDocumentRangeFormattingEditProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentRangeFormattingEditProvider(extension, checkSelector(selector), provider);
            },
            registerOnTypeFormattingEditProvider(selector, provider, firstTriggerCharacter, ...moreTriggerCharacters) {
                return extHostLanguageFeatures.registerOnTypeFormattingEditProvider(extension, checkSelector(selector), provider, [firstTriggerCharacter].concat(moreTriggerCharacters));
            },
            registerDocumentSemanticTokensProvider(selector, provider, legend) {
                return extHostLanguageFeatures.registerDocumentSemanticTokensProvider(extension, checkSelector(selector), provider, legend);
            },
            registerDocumentRangeSemanticTokensProvider(selector, provider, legend) {
                return extHostLanguageFeatures.registerDocumentRangeSemanticTokensProvider(extension, checkSelector(selector), provider, legend);
            },
            registerSignatureHelpProvider(selector, provider, firstItem, ...remaining) {
                if (typeof firstItem === 'object') {
                    return extHostLanguageFeatures.registerSignatureHelpProvider(extension, checkSelector(selector), provider, firstItem);
                }
                return extHostLanguageFeatures.registerSignatureHelpProvider(extension, checkSelector(selector), provider, typeof firstItem === 'undefined' ? [] : [firstItem, ...remaining]);
            },
            registerCompletionItemProvider(selector, provider, ...triggerCharacters) {
                return extHostLanguageFeatures.registerCompletionItemProvider(extension, checkSelector(selector), provider, triggerCharacters);
            },
            registerInlineCompletionItemProvider(selector, provider, metadata) {
                if (provider.handleDidShowCompletionItem) {
                    checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                }
                if (provider.handleDidPartiallyAcceptCompletionItem) {
                    checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                }
                if (metadata) {
                    checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                }
                return extHostLanguageFeatures.registerInlineCompletionsProvider(extension, checkSelector(selector), provider, metadata);
            },
            get inlineCompletionsUnificationState() {
                checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                return extHostLanguageFeatures.inlineCompletionsUnificationState;
            },
            onDidChangeCompletionsUnificationState(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'inlineCompletionsAdditions');
                return _asExtensionEvent(extHostLanguageFeatures.onDidChangeInlineCompletionsUnificationState)(listener, thisArg, disposables);
            },
            registerDocumentLinkProvider(selector, provider) {
                return extHostLanguageFeatures.registerDocumentLinkProvider(extension, checkSelector(selector), provider);
            },
            registerColorProvider(selector, provider) {
                return extHostLanguageFeatures.registerColorProvider(extension, checkSelector(selector), provider);
            },
            registerFoldingRangeProvider(selector, provider) {
                return extHostLanguageFeatures.registerFoldingRangeProvider(extension, checkSelector(selector), provider);
            },
            registerSelectionRangeProvider(selector, provider) {
                return extHostLanguageFeatures.registerSelectionRangeProvider(extension, selector, provider);
            },
            registerCallHierarchyProvider(selector, provider) {
                return extHostLanguageFeatures.registerCallHierarchyProvider(extension, selector, provider);
            },
            registerTypeHierarchyProvider(selector, provider) {
                return extHostLanguageFeatures.registerTypeHierarchyProvider(extension, selector, provider);
            },
            setLanguageConfiguration: (language, configuration) => {
                return extHostLanguageFeatures.setLanguageConfiguration(extension, language, configuration);
            },
            getTokenInformationAtPosition(doc, pos) {
                checkProposedApiEnabled(extension, 'tokenInformation');
                return extHostLanguages.tokenAtPosition(doc, pos);
            },
            registerInlayHintsProvider(selector, provider) {
                return extHostLanguageFeatures.registerInlayHintsProvider(extension, selector, provider);
            },
            createLanguageStatusItem(id, selector) {
                return extHostLanguages.createLanguageStatusItem(extension, id, selector);
            },
            registerDocumentDropEditProvider(selector, provider, metadata) {
                return extHostLanguageFeatures.registerDocumentOnDropEditProvider(extension, selector, provider, metadata);
            }
        };
        const window = {
            get activeTextEditor() {
                return extHostEditors.getActiveTextEditor();
            },
            get visibleTextEditors() {
                return extHostEditors.getVisibleTextEditors();
            },
            get activeTerminal() {
                return extHostTerminalService.activeTerminal;
            },
            get terminals() {
                return extHostTerminalService.terminals;
            },
            async showTextDocument(documentOrUri, columnOrOptions, preserveFocus) {
                if (URI.isUri(documentOrUri) && documentOrUri.scheme === Schemas.vscodeRemote && !documentOrUri.authority) {
                    extHostApiDeprecation.report('workspace.showTextDocument', extension, `A URI of 'vscode-remote' scheme requires an authority.`);
                }
                const document = await (URI.isUri(documentOrUri)
                    ? Promise.resolve(workspace.openTextDocument(documentOrUri))
                    : Promise.resolve(documentOrUri));
                return extHostEditors.showTextDocument(document, columnOrOptions, preserveFocus);
            },
            createTextEditorDecorationType(options) {
                return extHostEditors.createTextEditorDecorationType(extension, options);
            },
            onDidChangeActiveTextEditor(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostEditors.onDidChangeActiveTextEditor)(listener, thisArg, disposables);
            },
            onDidChangeVisibleTextEditors(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostEditors.onDidChangeVisibleTextEditors)(listener, thisArg, disposables);
            },
            onDidChangeTextEditorSelection(listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostEditors.onDidChangeTextEditorSelection)(listener, thisArgs, disposables);
            },
            onDidChangeTextEditorOptions(listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostEditors.onDidChangeTextEditorOptions)(listener, thisArgs, disposables);
            },
            onDidChangeTextEditorVisibleRanges(listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostEditors.onDidChangeTextEditorVisibleRanges)(listener, thisArgs, disposables);
            },
            onDidChangeTextEditorViewColumn(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostEditors.onDidChangeTextEditorViewColumn)(listener, thisArg, disposables);
            },
            onDidChangeTextEditorDiffInformation(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'textEditorDiffInformation');
                return _asExtensionEvent(extHostEditors.onDidChangeTextEditorDiffInformation)(listener, thisArg, disposables);
            },
            onDidCloseTerminal(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTerminalService.onDidCloseTerminal)(listener, thisArg, disposables);
            },
            onDidOpenTerminal(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTerminalService.onDidOpenTerminal)(listener, thisArg, disposables);
            },
            onDidChangeActiveTerminal(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTerminalService.onDidChangeActiveTerminal)(listener, thisArg, disposables);
            },
            onDidChangeTerminalDimensions(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'terminalDimensions');
                return _asExtensionEvent(extHostTerminalService.onDidChangeTerminalDimensions)(listener, thisArg, disposables);
            },
            onDidChangeTerminalState(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTerminalService.onDidChangeTerminalState)(listener, thisArg, disposables);
            },
            onDidWriteTerminalData(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'terminalDataWriteEvent');
                return _asExtensionEvent(extHostTerminalService.onDidWriteTerminalData)(listener, thisArg, disposables);
            },
            onDidExecuteTerminalCommand(listener, thisArg, disposables) {
                checkProposedApiEnabled(extension, 'terminalExecuteCommandEvent');
                return _asExtensionEvent(extHostTerminalService.onDidExecuteTerminalCommand)(listener, thisArg, disposables);
            },
            onDidChangeTerminalShellIntegration(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTerminalShellIntegration.onDidChangeTerminalShellIntegration)(listener, thisArg, disposables);
            },
            onDidStartTerminalShellExecution(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTerminalShellIntegration.onDidStartTerminalShellExecution)(listener, thisArg, disposables);
            },
            onDidEndTerminalShellExecution(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTerminalShellIntegration.onDidEndTerminalShellExecution)(listener, thisArg, disposables);
            },
            get state() {
                return extHostWindow.getState();
            },
            onDidChangeWindowState(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostWindow.onDidChangeWindowState)(listener, thisArg, disposables);
            },
            showInformationMessage(message, ...rest) {
                return extHostMessageService.showMessage(extension, Severity$1.Info, message, rest[0], rest.slice(1));
            },
            showWarningMessage(message, ...rest) {
                return extHostMessageService.showMessage(extension, Severity$1.Warning, message, rest[0], rest.slice(1));
            },
            showErrorMessage(message, ...rest) {
                return extHostMessageService.showMessage(extension, Severity$1.Error, message, rest[0], rest.slice(1));
            },
            showQuickPick(items, options, token) {
                return extHostQuickOpen.showQuickPick(extension, items, options, token);
            },
            showWorkspaceFolderPick(options) {
                return extHostQuickOpen.showWorkspaceFolderPick(options);
            },
            showInputBox(options, token) {
                return extHostQuickOpen.showInput(options, token);
            },
            showOpenDialog(options) {
                return extHostDialogs.showOpenDialog(options);
            },
            showSaveDialog(options) {
                return extHostDialogs.showSaveDialog(options);
            },
            createStatusBarItem(alignmentOrId, priorityOrAlignment, priorityArg) {
                let id;
                let alignment;
                let priority;
                if (typeof alignmentOrId === 'string') {
                    id = alignmentOrId;
                    alignment = priorityOrAlignment;
                    priority = priorityArg;
                }
                else {
                    alignment = alignmentOrId;
                    priority = priorityOrAlignment;
                }
                return extHostStatusBar.createStatusBarEntry(extension, id, alignment, priority);
            },
            setStatusBarMessage(text, timeoutOrThenable) {
                return extHostStatusBar.setStatusBarMessage(text, timeoutOrThenable);
            },
            withScmProgress(task) {
                extHostApiDeprecation.report('window.withScmProgress', extension, `Use 'withProgress' instead.`);
                return extHostProgress.withProgress(extension, { location: ProgressLocation.SourceControl }, (progress, token) => task({ report(n) { } }));
            },
            withProgress(options, task) {
                return extHostProgress.withProgress(extension, options, task);
            },
            createOutputChannel(name, options) {
                return extHostOutputService.createOutputChannel(name, options, extension);
            },
            createWebviewPanel(viewType, title, showOptions, options) {
                return extHostWebviewPanels.createWebviewPanel(extension, viewType, title, showOptions, options);
            },
            createWebviewTextEditorInset(editor, line, height, options) {
                checkProposedApiEnabled(extension, 'editorInsets');
                return extHostEditorInsets.createWebviewEditorInset(editor, line, height, options, extension);
            },
            createTerminal(nameOrOptions, shellPath, shellArgs) {
                if (typeof nameOrOptions === 'object') {
                    if ('pty' in nameOrOptions) {
                        return extHostTerminalService.createExtensionTerminal(nameOrOptions);
                    }
                    return extHostTerminalService.createTerminalFromOptions(nameOrOptions);
                }
                return extHostTerminalService.createTerminal(nameOrOptions, shellPath, shellArgs);
            },
            registerTerminalLinkProvider(provider) {
                return extHostTerminalService.registerLinkProvider(provider);
            },
            registerTerminalProfileProvider(id, provider) {
                return extHostTerminalService.registerProfileProvider(extension, id, provider);
            },
            registerTerminalCompletionProvider(provider, ...triggerCharacters) {
                checkProposedApiEnabled(extension, 'terminalCompletionProvider');
                return extHostTerminalService.registerTerminalCompletionProvider(extension, provider, ...triggerCharacters);
            },
            registerTerminalQuickFixProvider(id, provider) {
                checkProposedApiEnabled(extension, 'terminalQuickFixProvider');
                return extHostTerminalService.registerTerminalQuickFixProvider(id, extension.identifier.value, provider);
            },
            registerTreeDataProvider(viewId, treeDataProvider) {
                return extHostTreeViews.registerTreeDataProvider(viewId, treeDataProvider, extension);
            },
            createTreeView(viewId, options) {
                return extHostTreeViews.createTreeView(viewId, options, extension);
            },
            registerWebviewPanelSerializer: (viewType, serializer) => {
                return extHostWebviewPanels.registerWebviewPanelSerializer(extension, viewType, serializer);
            },
            registerCustomEditorProvider: (viewType, provider, options = {}) => {
                return extHostCustomEditors.registerCustomEditorProvider(extension, viewType, provider, options);
            },
            registerFileDecorationProvider(provider) {
                return extHostDecorations.registerFileDecorationProvider(provider, extension);
            },
            registerUriHandler(handler) {
                return extHostUrls.registerUriHandler(extension, handler);
            },
            createQuickPick() {
                return extHostQuickOpen.createQuickPick(extension);
            },
            createInputBox() {
                return extHostQuickOpen.createInputBox(extension);
            },
            get activeColorTheme() {
                return extHostTheming.activeColorTheme;
            },
            onDidChangeActiveColorTheme(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostTheming.onDidChangeActiveColorTheme)(listener, thisArg, disposables);
            },
            registerWebviewViewProvider(viewId, provider, options) {
                return extHostWebviewViews.registerWebviewViewProvider(extension, viewId, provider, options?.webviewOptions);
            },
            get activeNotebookEditor() {
                return extHostNotebook.activeNotebookEditor;
            },
            onDidChangeActiveNotebookEditor(listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostNotebook.onDidChangeActiveNotebookEditor)(listener, thisArgs, disposables);
            },
            get visibleNotebookEditors() {
                return extHostNotebook.visibleNotebookEditors;
            },
            get onDidChangeVisibleNotebookEditors() {
                return _asExtensionEvent(extHostNotebook.onDidChangeVisibleNotebookEditors);
            },
            onDidChangeNotebookEditorSelection(listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostNotebookEditors.onDidChangeNotebookEditorSelection)(listener, thisArgs, disposables);
            },
            onDidChangeNotebookEditorVisibleRanges(listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostNotebookEditors.onDidChangeNotebookEditorVisibleRanges)(listener, thisArgs, disposables);
            },
            showNotebookDocument(document, options) {
                return extHostNotebook.showNotebookDocument(document, options);
            },
            registerExternalUriOpener(id, opener, metadata) {
                checkProposedApiEnabled(extension, 'externalUriOpener');
                return extHostUriOpeners.registerExternalUriOpener(extension.identifier, id, opener, metadata);
            },
            registerProfileContentHandler(id, handler) {
                checkProposedApiEnabled(extension, 'profileContentHandlers');
                return extHostProfileContentHandlers.registerProfileContentHandler(extension, id, handler);
            },
            registerQuickDiffProvider(selector, quickDiffProvider, id, label, rootUri) {
                checkProposedApiEnabled(extension, 'quickDiffProvider');
                return extHostQuickDiff.registerQuickDiffProvider(extension, checkSelector(selector), quickDiffProvider, id, label, rootUri);
            },
            get tabGroups() {
                return extHostEditorTabs.tabGroups;
            },
            registerShareProvider(selector, provider) {
                checkProposedApiEnabled(extension, 'shareProvider');
                return extHostShare.registerShareProvider(checkSelector(selector), provider);
            },
            get nativeHandle() {
                checkProposedApiEnabled(extension, 'nativeWindowHandle');
                return extHostWindow.nativeHandle;
            },
            createChatStatusItem: (id) => {
                checkProposedApiEnabled(extension, 'chatStatusItem');
                return extHostChatStatus.createChatStatusItem(extension, id);
            },
        };
        const workspace = {
            get rootPath() {
                extHostApiDeprecation.report('workspace.rootPath', extension, `Please use 'workspace.workspaceFolders' instead. More details: https://aka.ms/vscode-eliminating-rootpath`);
                return extHostWorkspace.getPath();
            },
            set rootPath(value) {
                throw new ReadonlyError('rootPath');
            },
            getWorkspaceFolder(resource) {
                return extHostWorkspace.getWorkspaceFolder(resource);
            },
            get workspaceFolders() {
                return extHostWorkspace.getWorkspaceFolders();
            },
            get name() {
                return extHostWorkspace.name;
            },
            set name(value) {
                throw new ReadonlyError('name');
            },
            get workspaceFile() {
                return extHostWorkspace.workspaceFile;
            },
            set workspaceFile(value) {
                throw new ReadonlyError('workspaceFile');
            },
            updateWorkspaceFolders: (index, deleteCount, ...workspaceFoldersToAdd) => {
                return extHostWorkspace.updateWorkspaceFolders(extension, index, deleteCount || 0, ...workspaceFoldersToAdd);
            },
            onDidChangeWorkspaceFolders: function (listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostWorkspace.onDidChangeWorkspace)(listener, thisArgs, disposables);
            },
            asRelativePath: (pathOrUri, includeWorkspace) => {
                return extHostWorkspace.getRelativePath(pathOrUri, includeWorkspace);
            },
            findFiles: (include, exclude, maxResults, token) => {
                return extHostWorkspace.findFiles(include, exclude, maxResults, extension.identifier, token);
            },
            findFiles2: (filePattern, options, token) => {
                checkProposedApiEnabled(extension, 'findFiles2');
                return extHostWorkspace.findFiles2(filePattern, options, extension.identifier, token);
            },
            findTextInFiles: (query, optionsOrCallback, callbackOrToken, token) => {
                checkProposedApiEnabled(extension, 'findTextInFiles');
                let options;
                let callback;
                if (typeof optionsOrCallback === 'object') {
                    options = optionsOrCallback;
                    callback = callbackOrToken;
                }
                else {
                    options = {};
                    callback = optionsOrCallback;
                    token = callbackOrToken;
                }
                return extHostWorkspace.findTextInFiles(query, options || {}, callback, extension.identifier, token);
            },
            findTextInFiles2: (query, options, token) => {
                checkProposedApiEnabled(extension, 'findTextInFiles2');
                checkProposedApiEnabled(extension, 'textSearchProvider2');
                return extHostWorkspace.findTextInFiles2(query, options, extension.identifier, token);
            },
            save: (uri) => {
                return extHostWorkspace.save(uri);
            },
            saveAs: (uri) => {
                return extHostWorkspace.saveAs(uri);
            },
            saveAll: (includeUntitled) => {
                return extHostWorkspace.saveAll(includeUntitled);
            },
            applyEdit(edit, metadata) {
                return extHostBulkEdits.applyWorkspaceEdit(edit, extension, metadata);
            },
            createFileSystemWatcher: (pattern, optionsOrIgnoreCreate, ignoreChange, ignoreDelete) => {
                const options = {
                    ignoreCreateEvents: Boolean(optionsOrIgnoreCreate),
                    ignoreChangeEvents: Boolean(ignoreChange),
                    ignoreDeleteEvents: Boolean(ignoreDelete),
                };
                return extHostFileSystemEvent.createFileSystemWatcher(extHostWorkspace, configProvider, extHostFileSystemInfo, extension, pattern, options);
            },
            get textDocuments() {
                return ( extHostDocuments.getAllDocumentData().map(data => data.document));
            },
            set textDocuments(value) {
                throw new ReadonlyError('textDocuments');
            },
            openTextDocument(uriOrFileNameOrOptions, options) {
                let uriPromise;
                options = (options ?? uriOrFileNameOrOptions);
                if (typeof uriOrFileNameOrOptions === 'string') {
                    uriPromise = Promise.resolve(URI.file(uriOrFileNameOrOptions));
                }
                else if (URI.isUri(uriOrFileNameOrOptions)) {
                    uriPromise = Promise.resolve(uriOrFileNameOrOptions);
                }
                else if (!options || typeof options === 'object') {
                    uriPromise = extHostDocuments.createDocumentData(options);
                }
                else {
                    throw ( new Error('illegal argument - uriOrFileNameOrOptions'));
                }
                return uriPromise.then(uri => {
                    extHostLogService.trace(`openTextDocument from ${extension.identifier}`);
                    if (uri.scheme === Schemas.vscodeRemote && !uri.authority) {
                        extHostApiDeprecation.report('workspace.openTextDocument', extension, `A URI of 'vscode-remote' scheme requires an authority.`);
                    }
                    return extHostDocuments.ensureDocumentData(uri, options).then(documentData => {
                        return documentData.document;
                    });
                });
            },
            onDidOpenTextDocument: (listener, thisArgs, disposables) => {
                return _asExtensionEvent(extHostDocuments.onDidAddDocument)(listener, thisArgs, disposables);
            },
            onDidCloseTextDocument: (listener, thisArgs, disposables) => {
                return _asExtensionEvent(extHostDocuments.onDidRemoveDocument)(listener, thisArgs, disposables);
            },
            onDidChangeTextDocument: (listener, thisArgs, disposables) => {
                if (isProposedApiEnabled(extension, 'textDocumentChangeReason')) {
                    return _asExtensionEvent(extHostDocuments.onDidChangeDocumentWithReason)(listener, thisArgs, disposables);
                }
                return _asExtensionEvent(extHostDocuments.onDidChangeDocument)(listener, thisArgs, disposables);
            },
            onDidSaveTextDocument: (listener, thisArgs, disposables) => {
                return _asExtensionEvent(extHostDocuments.onDidSaveDocument)(listener, thisArgs, disposables);
            },
            onWillSaveTextDocument: (listener, thisArgs, disposables) => {
                return _asExtensionEvent(extHostDocumentSaveParticipant.getOnWillSaveTextDocumentEvent(extension))(listener, thisArgs, disposables);
            },
            get notebookDocuments() {
                return ( extHostNotebook.notebookDocuments.map(d => d.apiNotebook));
            },
            async openNotebookDocument(uriOrType, content) {
                let uri;
                if (URI.isUri(uriOrType)) {
                    uri = uriOrType;
                    await extHostNotebook.openNotebookDocument(uriOrType);
                }
                else if (typeof uriOrType === 'string') {
                    uri = URI.revive(await extHostNotebook.createNotebookDocument({ viewType: uriOrType, content }));
                }
                else {
                    throw ( new Error('Invalid arguments'));
                }
                return extHostNotebook.getNotebookDocument(uri).apiNotebook;
            },
            onDidSaveNotebookDocument(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostNotebookDocuments.onDidSaveNotebookDocument)(listener, thisArg, disposables);
            },
            onDidChangeNotebookDocument(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostNotebookDocuments.onDidChangeNotebookDocument)(listener, thisArg, disposables);
            },
            onWillSaveNotebookDocument(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostNotebookDocumentSaveParticipant.getOnWillSaveNotebookDocumentEvent(extension))(listener, thisArg, disposables);
            },
            get onDidOpenNotebookDocument() {
                return _asExtensionEvent(extHostNotebook.onDidOpenNotebookDocument);
            },
            get onDidCloseNotebookDocument() {
                return _asExtensionEvent(extHostNotebook.onDidCloseNotebookDocument);
            },
            registerNotebookSerializer(viewType, serializer, options, registration) {
                return extHostNotebook.registerNotebookSerializer(extension, viewType, serializer, options, isProposedApiEnabled(extension, 'notebookLiveShare') ? registration : undefined);
            },
            onDidChangeConfiguration: (listener, thisArgs, disposables) => {
                return _asExtensionEvent(configProvider.onDidChangeConfiguration)(listener, thisArgs, disposables);
            },
            getConfiguration(section, scope) {
                scope = arguments.length === 1 ? undefined : scope;
                return configProvider.getConfiguration(section, scope, extension);
            },
            registerTextDocumentContentProvider(scheme, provider) {
                return extHostDocumentContentProviders.registerTextDocumentContentProvider(scheme, provider);
            },
            registerTaskProvider: (type, provider) => {
                extHostApiDeprecation.report('window.registerTaskProvider', extension, `Use the corresponding function on the 'tasks' namespace instead`);
                return extHostTask.registerTaskProvider(extension, type, provider);
            },
            registerFileSystemProvider(scheme, provider, options) {
                return combinedDisposable(extHostFileSystem.registerFileSystemProvider(extension, scheme, provider, options), extHostConsumerFileSystem.addFileSystemProvider(scheme, provider, options));
            },
            get fs() {
                return extHostConsumerFileSystem.value;
            },
            registerFileSearchProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'fileSearchProvider');
                return extHostSearch.registerFileSearchProviderOld(scheme, provider);
            },
            registerTextSearchProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'textSearchProvider');
                return extHostSearch.registerTextSearchProviderOld(scheme, provider);
            },
            registerAITextSearchProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'aiTextSearchProvider');
                checkProposedApiEnabled(extension, 'textSearchProvider2');
                return extHostSearch.registerAITextSearchProvider(scheme, provider);
            },
            registerFileSearchProvider2: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'fileSearchProvider2');
                return extHostSearch.registerFileSearchProvider(scheme, provider);
            },
            registerTextSearchProvider2: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'textSearchProvider2');
                return extHostSearch.registerTextSearchProvider(scheme, provider);
            },
            registerRemoteAuthorityResolver: (authorityPrefix, resolver) => {
                checkProposedApiEnabled(extension, 'resolvers');
                return extensionService.registerRemoteAuthorityResolver(authorityPrefix, resolver);
            },
            registerResourceLabelFormatter: (formatter) => {
                checkProposedApiEnabled(extension, 'resolvers');
                return extHostLabelService.$registerResourceLabelFormatter(formatter);
            },
            getRemoteExecServer: (authority) => {
                checkProposedApiEnabled(extension, 'resolvers');
                return extensionService.getRemoteExecServer(authority);
            },
            onDidCreateFiles: (listener, thisArg, disposables) => {
                return _asExtensionEvent(extHostFileSystemEvent.onDidCreateFile)(listener, thisArg, disposables);
            },
            onDidDeleteFiles: (listener, thisArg, disposables) => {
                return _asExtensionEvent(extHostFileSystemEvent.onDidDeleteFile)(listener, thisArg, disposables);
            },
            onDidRenameFiles: (listener, thisArg, disposables) => {
                return _asExtensionEvent(extHostFileSystemEvent.onDidRenameFile)(listener, thisArg, disposables);
            },
            onWillCreateFiles: (listener, thisArg, disposables) => {
                return _asExtensionEvent(extHostFileSystemEvent.getOnWillCreateFileEvent(extension))(listener, thisArg, disposables);
            },
            onWillDeleteFiles: (listener, thisArg, disposables) => {
                return _asExtensionEvent(extHostFileSystemEvent.getOnWillDeleteFileEvent(extension))(listener, thisArg, disposables);
            },
            onWillRenameFiles: (listener, thisArg, disposables) => {
                return _asExtensionEvent(extHostFileSystemEvent.getOnWillRenameFileEvent(extension))(listener, thisArg, disposables);
            },
            openTunnel: (forward) => {
                checkProposedApiEnabled(extension, 'tunnels');
                return extHostTunnelService.openTunnel(extension, forward).then(value => {
                    if (!value) {
                        throw ( new Error('cannot open tunnel'));
                    }
                    return value;
                });
            },
            get tunnels() {
                checkProposedApiEnabled(extension, 'tunnels');
                return extHostTunnelService.getTunnels();
            },
            onDidChangeTunnels: (listener, thisArg, disposables) => {
                checkProposedApiEnabled(extension, 'tunnels');
                return _asExtensionEvent(extHostTunnelService.onDidChangeTunnels)(listener, thisArg, disposables);
            },
            registerPortAttributesProvider: (portSelector, provider) => {
                checkProposedApiEnabled(extension, 'portsAttributes');
                return extHostTunnelService.registerPortsAttributesProvider(portSelector, provider);
            },
            registerTunnelProvider: (tunnelProvider, information) => {
                checkProposedApiEnabled(extension, 'tunnelFactory');
                return extHostTunnelService.registerTunnelProvider(tunnelProvider, information);
            },
            registerTimelineProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'timeline');
                return extHostTimeline.registerTimelineProvider(scheme, provider, extension.identifier, extHostCommands.converter);
            },
            get isTrusted() {
                return extHostWorkspace.trusted;
            },
            requestWorkspaceTrust: (options) => {
                checkProposedApiEnabled(extension, 'workspaceTrust');
                return extHostWorkspace.requestWorkspaceTrust(options);
            },
            onDidGrantWorkspaceTrust: (listener, thisArgs, disposables) => {
                return _asExtensionEvent(extHostWorkspace.onDidGrantWorkspaceTrust)(listener, thisArgs, disposables);
            },
            registerEditSessionIdentityProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'editSessionIdentityProvider');
                return extHostWorkspace.registerEditSessionIdentityProvider(scheme, provider);
            },
            onWillCreateEditSessionIdentity: (listener, thisArgs, disposables) => {
                checkProposedApiEnabled(extension, 'editSessionIdentityProvider');
                return _asExtensionEvent(extHostWorkspace.getOnWillCreateEditSessionIdentityEvent(extension))(listener, thisArgs, disposables);
            },
            registerCanonicalUriProvider: (scheme, provider) => {
                checkProposedApiEnabled(extension, 'canonicalUriProvider');
                return extHostWorkspace.registerCanonicalUriProvider(scheme, provider);
            },
            getCanonicalUri: (uri, options, token) => {
                checkProposedApiEnabled(extension, 'canonicalUriProvider');
                return extHostWorkspace.provideCanonicalUri(uri, options, token);
            },
            decode(content, options) {
                return extHostWorkspace.decode(content, options);
            },
            encode(content, options) {
                return extHostWorkspace.encode(content, options);
            },
        };
        const scm = {
            get inputBox() {
                extHostApiDeprecation.report('scm.inputBox', extension, `Use 'SourceControl.inputBox' instead`);
                return extHostSCM.getLastInputBox(extension);
            },
            createSourceControl(id, label, rootUri, iconPath, isHidden, parent) {
                if (iconPath || isHidden || parent) {
                    checkProposedApiEnabled(extension, 'scmProviderOptions');
                }
                return extHostSCM.createSourceControl(extension, id, label, rootUri, iconPath, isHidden, parent);
            }
        };
        const comments = {
            createCommentController(id, label) {
                return extHostComment.createCommentController(extension, id, label);
            }
        };
        const debug = {
            get activeDebugSession() {
                return extHostDebugService.activeDebugSession;
            },
            get activeDebugConsole() {
                return extHostDebugService.activeDebugConsole;
            },
            get breakpoints() {
                return extHostDebugService.breakpoints;
            },
            get activeStackItem() {
                return extHostDebugService.activeStackItem;
            },
            registerDebugVisualizationProvider(id, provider) {
                checkProposedApiEnabled(extension, 'debugVisualization');
                return extHostDebugService.registerDebugVisualizationProvider(extension, id, provider);
            },
            registerDebugVisualizationTreeProvider(id, provider) {
                checkProposedApiEnabled(extension, 'debugVisualization');
                return extHostDebugService.registerDebugVisualizationTree(extension, id, provider);
            },
            onDidStartDebugSession(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostDebugService.onDidStartDebugSession)(listener, thisArg, disposables);
            },
            onDidTerminateDebugSession(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostDebugService.onDidTerminateDebugSession)(listener, thisArg, disposables);
            },
            onDidChangeActiveDebugSession(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostDebugService.onDidChangeActiveDebugSession)(listener, thisArg, disposables);
            },
            onDidReceiveDebugSessionCustomEvent(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostDebugService.onDidReceiveDebugSessionCustomEvent)(listener, thisArg, disposables);
            },
            onDidChangeBreakpoints(listener, thisArgs, disposables) {
                return _asExtensionEvent(extHostDebugService.onDidChangeBreakpoints)(listener, thisArgs, disposables);
            },
            onDidChangeActiveStackItem(listener, thisArg, disposables) {
                return _asExtensionEvent(extHostDebugService.onDidChangeActiveStackItem)(listener, thisArg, disposables);
            },
            registerDebugConfigurationProvider(debugType, provider, triggerKind) {
                return extHostDebugService.registerDebugConfigurationProvider(debugType, provider, triggerKind || DebugConfigurationProviderTriggerKind.Initial);
            },
            registerDebugAdapterDescriptorFactory(debugType, factory) {
                return extHostDebugService.registerDebugAdapterDescriptorFactory(extension, debugType, factory);
            },
            registerDebugAdapterTrackerFactory(debugType, factory) {
                return extHostDebugService.registerDebugAdapterTrackerFactory(debugType, factory);
            },
            startDebugging(folder, nameOrConfig, parentSessionOrOptions) {
                if (!parentSessionOrOptions || (typeof parentSessionOrOptions === 'object' && 'configuration' in parentSessionOrOptions)) {
                    return extHostDebugService.startDebugging(folder, nameOrConfig, { parentSession: parentSessionOrOptions });
                }
                return extHostDebugService.startDebugging(folder, nameOrConfig, parentSessionOrOptions || {});
            },
            stopDebugging(session) {
                return extHostDebugService.stopDebugging(session);
            },
            addBreakpoints(breakpoints) {
                return extHostDebugService.addBreakpoints(breakpoints);
            },
            removeBreakpoints(breakpoints) {
                return extHostDebugService.removeBreakpoints(breakpoints);
            },
            asDebugSourceUri(source, session) {
                return extHostDebugService.asDebugSourceUri(source, session);
            }
        };
        const tasks = {
            registerTaskProvider: (type, provider) => {
                return extHostTask.registerTaskProvider(extension, type, provider);
            },
            fetchTasks: (filter) => {
                return extHostTask.fetchTasks(filter);
            },
            executeTask: (task) => {
                return extHostTask.executeTask(extension, task);
            },
            get taskExecutions() {
                return extHostTask.taskExecutions;
            },
            onDidStartTask: (listener, thisArgs, disposables) => {
                const wrappedListener = (event) => {
                    if (!isProposedApiEnabled(extension, 'taskExecutionTerminal')) {
                        if (event?.execution?.terminal !== undefined) {
                            event.execution.terminal = undefined;
                        }
                    }
                    const eventWithExecution = {
                        ...event,
                        execution: event.execution
                    };
                    return listener.call(thisArgs, eventWithExecution);
                };
                return _asExtensionEvent(extHostTask.onDidStartTask)(wrappedListener, thisArgs, disposables);
            },
            onDidEndTask: (listeners, thisArgs, disposables) => {
                return _asExtensionEvent(extHostTask.onDidEndTask)(listeners, thisArgs, disposables);
            },
            onDidStartTaskProcess: (listeners, thisArgs, disposables) => {
                return _asExtensionEvent(extHostTask.onDidStartTaskProcess)(listeners, thisArgs, disposables);
            },
            onDidEndTaskProcess: (listeners, thisArgs, disposables) => {
                return _asExtensionEvent(extHostTask.onDidEndTaskProcess)(listeners, thisArgs, disposables);
            },
            onDidStartTaskProblemMatchers: (listeners, thisArgs, disposables) => {
                checkProposedApiEnabled(extension, 'taskProblemMatcherStatus');
                return _asExtensionEvent(extHostTask.onDidStartTaskProblemMatchers)(listeners, thisArgs, disposables);
            },
            onDidEndTaskProblemMatchers: (listeners, thisArgs, disposables) => {
                checkProposedApiEnabled(extension, 'taskProblemMatcherStatus');
                return _asExtensionEvent(extHostTask.onDidEndTaskProblemMatchers)(listeners, thisArgs, disposables);
            }
        };
        const notebooks = {
            createNotebookController(id, notebookType, label, handler, rendererScripts) {
                return extHostNotebookKernels.createNotebookController(extension, id, notebookType, label, handler, isProposedApiEnabled(extension, 'notebookMessaging') ? rendererScripts : undefined);
            },
            registerNotebookCellStatusBarItemProvider: (notebookType, provider) => {
                return extHostNotebook.registerNotebookCellStatusBarItemProvider(extension, notebookType, provider);
            },
            createRendererMessaging(rendererId) {
                return extHostNotebookRenderers.createRendererMessaging(extension, rendererId);
            },
            createNotebookControllerDetectionTask(notebookType) {
                checkProposedApiEnabled(extension, 'notebookKernelSource');
                return extHostNotebookKernels.createNotebookControllerDetectionTask(extension, notebookType);
            },
            registerKernelSourceActionProvider(notebookType, provider) {
                checkProposedApiEnabled(extension, 'notebookKernelSource');
                return extHostNotebookKernels.registerKernelSourceActionProvider(extension, notebookType, provider);
            },
        };
        const l10n = {
            t(...params) {
                if (typeof params[0] === 'string') {
                    const key = params.shift();
                    const argsFormatted = !params || typeof params[0] !== 'object' ? params : params[0];
                    return extHostLocalization.getMessage(extension.identifier.value, { message: key, args: argsFormatted });
                }
                return extHostLocalization.getMessage(extension.identifier.value, params[0]);
            },
            get bundle() {
                return extHostLocalization.getBundle(extension.identifier.value);
            },
            get uri() {
                return extHostLocalization.getBundleUri(extension.identifier.value);
            }
        };
        const interactive = {
            transferActiveChat(toWorkspace) {
                checkProposedApiEnabled(extension, 'interactive');
                return extHostChatAgents2.transferActiveChat(toWorkspace);
            }
        };
        const ai = {
            getRelatedInformation(query, types) {
                checkProposedApiEnabled(extension, 'aiRelatedInformation');
                return extHostAiRelatedInformation.getRelatedInformation(extension, query, types);
            },
            registerRelatedInformationProvider(type, provider) {
                checkProposedApiEnabled(extension, 'aiRelatedInformation');
                return extHostAiRelatedInformation.registerRelatedInformationProvider(extension, type, provider);
            },
            registerEmbeddingVectorProvider(model, provider) {
                checkProposedApiEnabled(extension, 'aiRelatedInformation');
                return extHostAiEmbeddingVector.registerEmbeddingVectorProvider(extension, model, provider);
            },
            registerSettingsSearchProvider(provider) {
                checkProposedApiEnabled(extension, 'aiSettingsSearch');
                return extHostAiSettingsSearch.registerSettingsSearchProvider(extension, provider);
            }
        };
        const chat = {
            registerMappedEditsProvider(_selector, _provider) {
                checkProposedApiEnabled(extension, 'mappedEditsProvider');
                return { dispose() { } };
            },
            registerMappedEditsProvider2(provider) {
                checkProposedApiEnabled(extension, 'mappedEditsProvider');
                return extHostCodeMapper.registerMappedEditsProvider(extension, provider);
            },
            createChatParticipant(id, handler) {
                return extHostChatAgents2.createChatAgent(extension, id, handler);
            },
            createDynamicChatParticipant(id, dynamicProps, handler) {
                checkProposedApiEnabled(extension, 'chatParticipantPrivate');
                return extHostChatAgents2.createDynamicChatAgent(extension, id, dynamicProps, handler);
            },
            registerChatParticipantDetectionProvider(provider) {
                checkProposedApiEnabled(extension, 'chatParticipantPrivate');
                return extHostChatAgents2.registerChatParticipantDetectionProvider(extension, provider);
            },
            registerRelatedFilesProvider(provider, metadata) {
                checkProposedApiEnabled(extension, 'chatEditing');
                return extHostChatAgents2.registerRelatedFilesProvider(extension, provider, metadata);
            },
            onDidDisposeChatSession: (listeners, thisArgs, disposables) => {
                checkProposedApiEnabled(extension, 'chatParticipantPrivate');
                return _asExtensionEvent(extHostChatAgents2.onDidDisposeChatSession)(listeners, thisArgs, disposables);
            },
            registerChatSessionItemProvider: (chatSessionType, provider) => {
                checkProposedApiEnabled(extension, 'chatSessionsProvider');
                return extHostChatSessions.registerChatSessionItemProvider(extension, chatSessionType, provider);
            },
            registerChatSessionContentProvider(scheme, provider, chatParticipant, capabilities) {
                checkProposedApiEnabled(extension, 'chatSessionsProvider');
                return extHostChatSessions.registerChatSessionContentProvider(extension, scheme, chatParticipant, provider, capabilities);
            },
            registerChatOutputRenderer: (viewType, renderer) => {
                checkProposedApiEnabled(extension, 'chatOutputRenderer');
                return extHostChatOutputRenderer.registerChatOutputRenderer(extension, viewType, renderer);
            },
            registerChatContextProvider(selector, id, provider) {
                checkProposedApiEnabled(extension, 'chatContextProvider');
                return extHostChatContext.registerChatContextProvider(selector ? checkSelector(selector) : undefined, `${extension.id}-${id}`, provider);
            },
            registerCustomAgentsProvider(provider) {
                checkProposedApiEnabled(extension, 'chatParticipantPrivate');
                return extHostChatAgents2.registerCustomAgentsProvider(extension, provider);
            },
        };
        const lm = {
            selectChatModels: (selector) => {
                return extHostLanguageModels.selectLanguageModels(extension, selector ?? {});
            },
            onDidChangeChatModels: (listener, thisArgs, disposables) => {
                return extHostLanguageModels.onDidChangeProviders(listener, thisArgs, disposables);
            },
            registerLanguageModelChatProvider: (vendor, provider) => {
                return extHostLanguageModels.registerLanguageModelChatProvider(extension, vendor, provider);
            },
            get isModelProxyAvailable() {
                checkProposedApiEnabled(extension, 'languageModelProxy');
                return extHostLanguageModels.isModelProxyAvailable;
            },
            onDidChangeModelProxyAvailability: (listener, thisArgs, disposables) => {
                checkProposedApiEnabled(extension, 'languageModelProxy');
                return extHostLanguageModels.onDidChangeModelProxyAvailability(listener, thisArgs, disposables);
            },
            getModelProxy: () => {
                checkProposedApiEnabled(extension, 'languageModelProxy');
                return extHostLanguageModels.getModelProxy(extension);
            },
            registerLanguageModelProxyProvider: (provider) => {
                checkProposedApiEnabled(extension, 'chatParticipantPrivate');
                return extHostLanguageModels.registerLanguageModelProxyProvider(extension, provider);
            },
            get embeddingModels() {
                checkProposedApiEnabled(extension, 'embeddings');
                return extHostEmbeddings.embeddingsModels;
            },
            onDidChangeEmbeddingModels: (listener, thisArgs, disposables) => {
                checkProposedApiEnabled(extension, 'embeddings');
                return extHostEmbeddings.onDidChange(listener, thisArgs, disposables);
            },
            registerEmbeddingsProvider(embeddingsModel, provider) {
                checkProposedApiEnabled(extension, 'embeddings');
                return extHostEmbeddings.registerEmbeddingsProvider(extension, embeddingsModel, provider);
            },
            async computeEmbeddings(embeddingsModel, input, token) {
                checkProposedApiEnabled(extension, 'embeddings');
                if (typeof input === 'string') {
                    return extHostEmbeddings.computeEmbeddings(embeddingsModel, input, token);
                }
                else {
                    return extHostEmbeddings.computeEmbeddings(embeddingsModel, input, token);
                }
            },
            registerTool(name, tool) {
                return extHostLanguageModelTools.registerTool(extension, name, tool);
            },
            invokeTool(name, parameters, token) {
                return extHostLanguageModelTools.invokeTool(extension, name, parameters, token);
            },
            get tools() {
                return extHostLanguageModelTools.getTools(extension);
            },
            fileIsIgnored(uri, token) {
                return extHostLanguageModels.fileIsIgnored(extension, uri, token);
            },
            registerIgnoredFileProvider(provider) {
                return extHostLanguageModels.registerIgnoredFileProvider(extension, provider);
            },
            registerMcpServerDefinitionProvider(id, provider) {
                return extHostMcp.registerMcpConfigurationProvider(extension, id, provider);
            },
            onDidChangeChatRequestTools(...args) {
                checkProposedApiEnabled(extension, 'chatParticipantAdditions');
                return _asExtensionEvent(extHostChatAgents2.onDidChangeChatRequestTools)(...args);
            }
        };
        const speech = {
            registerSpeechProvider(id, provider) {
                checkProposedApiEnabled(extension, 'speech');
                return extHostSpeech.registerProvider(extension.identifier, id, provider);
            }
        };
        return {
            version: initData.version,
            ai,
            authentication,
            commands,
            comments,
            chat,
            debug,
            env,
            extensions,
            interactive,
            l10n,
            languages,
            lm,
            notebooks,
            scm,
            speech,
            tasks,
            tests,
            window,
            workspace,
            Breakpoint: Breakpoint,
            TerminalOutputAnchor: TerminalOutputAnchor,
            ChatResultFeedbackKind: ChatResultFeedbackKind,
            ChatVariableLevel: ChatVariableLevel,
            ChatCompletionItem: ChatCompletionItem,
            ChatReferenceDiagnostic: ChatReferenceDiagnostic,
            CallHierarchyIncomingCall: CallHierarchyIncomingCall,
            CallHierarchyItem: CallHierarchyItem,
            CallHierarchyOutgoingCall: CallHierarchyOutgoingCall,
            CancellationError: CancellationError,
            CancellationTokenSource: CancellationTokenSource,
            CandidatePortSource: CandidatePortSource,
            CodeAction: CodeAction,
            CodeActionKind: CodeActionKind,
            CodeActionTriggerKind: CodeActionTriggerKind,
            CodeLens: CodeLens,
            Color: Color,
            ColorInformation: ColorInformation,
            ColorPresentation: ColorPresentation,
            ColorThemeKind: ColorThemeKind,
            CommentMode: CommentMode,
            CommentState: CommentState,
            CommentThreadCollapsibleState: CommentThreadCollapsibleState,
            CommentThreadState: CommentThreadState,
            CommentThreadApplicability: CommentThreadApplicability,
            CommentThreadFocus: CommentThreadFocus,
            CompletionItem: CompletionItem,
            CompletionItemKind: CompletionItemKind,
            CompletionItemTag: CompletionItemTag,
            CompletionList: CompletionList,
            CompletionTriggerKind: CompletionTriggerKind,
            ConfigurationTarget: ConfigurationTarget,
            CustomExecution: CustomExecution,
            DebugAdapterExecutable: DebugAdapterExecutable,
            DebugAdapterInlineImplementation: DebugAdapterInlineImplementation,
            DebugAdapterNamedPipeServer: DebugAdapterNamedPipeServer,
            DebugAdapterServer: DebugAdapterServer,
            DebugConfigurationProviderTriggerKind: DebugConfigurationProviderTriggerKind,
            DebugConsoleMode: DebugConsoleMode,
            DebugVisualization: DebugVisualization,
            DecorationRangeBehavior: DecorationRangeBehavior,
            Diagnostic: Diagnostic,
            DiagnosticRelatedInformation: DiagnosticRelatedInformation,
            DiagnosticSeverity: DiagnosticSeverity,
            DiagnosticTag: DiagnosticTag,
            Disposable: Disposable,
            DocumentHighlight: DocumentHighlight,
            DocumentHighlightKind: DocumentHighlightKind,
            MultiDocumentHighlight: MultiDocumentHighlight,
            DocumentLink: DocumentLink,
            DocumentSymbol: DocumentSymbol,
            EndOfLine: EndOfLine,
            EnvironmentVariableMutatorType: EnvironmentVariableMutatorType,
            EvaluatableExpression: EvaluatableExpression,
            InlineValueText: InlineValueText,
            InlineValueVariableLookup: InlineValueVariableLookup,
            InlineValueEvaluatableExpression: InlineValueEvaluatableExpression,
            InlineCompletionTriggerKind: InlineCompletionTriggerKind,
            InlineCompletionsDisposeReasonKind: InlineCompletionsDisposeReasonKind,
            EventEmitter: Emitter,
            ExtensionKind: ExtensionKind,
            ExtensionMode: ExtensionMode,
            ExternalUriOpenerPriority: ExternalUriOpenerPriority,
            FileChangeType: FileChangeType,
            FileDecoration: FileDecoration,
            FileDecoration2: FileDecoration,
            FileSystemError: FileSystemError,
            FileType: FileType,
            FilePermission: FilePermission,
            FoldingRange: FoldingRange,
            FoldingRangeKind: FoldingRangeKind,
            FunctionBreakpoint: FunctionBreakpoint,
            InlineCompletionItem: InlineSuggestion,
            InlineCompletionList: InlineSuggestionList,
            Hover: Hover,
            VerboseHover: VerboseHover,
            HoverVerbosityAction: HoverVerbosityAction,
            IndentAction: IndentAction,
            Location: Location,
            MarkdownString: MarkdownString,
            OverviewRulerLane: OverviewRulerLane,
            ParameterInformation: ParameterInformation,
            PortAutoForwardAction: PortAutoForwardAction,
            Position: Position,
            ProcessExecution: ProcessExecution,
            ProgressLocation: ProgressLocation,
            QuickInputButtonLocation: QuickInputButtonLocation,
            QuickInputButtons: QuickInputButtons,
            Range: Range,
            RelativePattern: RelativePattern,
            Selection: Selection,
            SelectionRange: SelectionRange,
            SemanticTokens: SemanticTokens,
            SemanticTokensBuilder: SemanticTokensBuilder,
            SemanticTokensEdit: SemanticTokensEdit,
            SemanticTokensEdits: SemanticTokensEdits,
            SemanticTokensLegend: SemanticTokensLegend,
            ShellExecution: ShellExecution,
            ShellQuoting: ShellQuoting,
            SignatureHelp: SignatureHelp,
            SignatureHelpTriggerKind: SignatureHelpTriggerKind,
            SignatureInformation: SignatureInformation,
            SnippetString: SnippetString,
            SourceBreakpoint: SourceBreakpoint,
            StandardTokenType: StandardTokenType,
            StatusBarAlignment: StatusBarAlignment,
            SymbolInformation: SymbolInformation,
            SymbolKind: SymbolKind,
            SymbolTag: SymbolTag,
            Task: Task,
            TaskEventKind: TaskEventKind,
            TaskGroup: TaskGroup,
            TaskPanelKind: TaskPanelKind,
            TaskRevealKind: TaskRevealKind,
            TaskScope: TaskScope,
            TerminalLink: TerminalLink,
            TerminalQuickFixTerminalCommand: TerminalQuickFixCommand,
            TerminalQuickFixOpener: TerminalQuickFixOpener,
            TerminalLocation: TerminalLocation,
            TerminalProfile: TerminalProfile,
            TerminalExitReason: TerminalExitReason,
            TerminalShellExecutionCommandLineConfidence: TerminalShellExecutionCommandLineConfidence,
            TerminalCompletionItem: TerminalCompletionItem,
            TerminalCompletionItemKind: TerminalCompletionItemKind,
            TerminalCompletionList: TerminalCompletionList,
            TerminalShellType: TerminalShellType,
            TextDocumentSaveReason: TextDocumentSaveReason,
            TextEdit: TextEdit,
            SnippetTextEdit: SnippetTextEdit,
            TextEditorCursorStyle: TextEditorCursorStyle,
            TextEditorChangeKind: TextEditorChangeKind,
            TextEditorLineNumbersStyle: TextEditorLineNumbersStyle,
            TextEditorRevealType: TextEditorRevealType,
            TextEditorSelectionChangeKind: TextEditorSelectionChangeKind,
            SyntaxTokenType: SyntaxTokenType,
            TextDocumentChangeReason: TextDocumentChangeReason,
            ThemeColor: ThemeColor,
            ThemeIcon: ThemeIcon,
            TreeItem: TreeItem,
            TreeItemCheckboxState: TreeItemCheckboxState,
            TreeItemCollapsibleState: TreeItemCollapsibleState,
            TypeHierarchyItem: TypeHierarchyItem,
            UIKind: UIKind,
            Uri: URI,
            ViewColumn: ViewColumn,
            WorkspaceEdit: WorkspaceEdit,
            DocumentPasteTriggerKind: DocumentPasteTriggerKind,
            DocumentDropEdit: DocumentDropEdit,
            DocumentDropOrPasteEditKind: DocumentDropOrPasteEditKind,
            DocumentPasteEdit: DocumentPasteEdit,
            InlayHint: InlayHint,
            InlayHintLabelPart: InlayHintLabelPart,
            InlayHintKind: InlayHintKind,
            RemoteAuthorityResolverError: RemoteAuthorityResolverError,
            ResolvedAuthority: ResolvedAuthority,
            ManagedResolvedAuthority: ManagedResolvedAuthority,
            SourceControlInputBoxValidationType: SourceControlInputBoxValidationType,
            ExtensionRuntime: ExtensionRuntime,
            TimelineItem: TimelineItem,
            NotebookRange: NotebookRange,
            NotebookCellKind: NotebookCellKind,
            NotebookCellExecutionState: NotebookCellExecutionState,
            NotebookCellData: NotebookCellData,
            NotebookData: NotebookData,
            NotebookRendererScript: NotebookRendererScript,
            NotebookCellStatusBarAlignment: NotebookCellStatusBarAlignment,
            NotebookEditorRevealType: NotebookEditorRevealType,
            NotebookCellOutput: NotebookCellOutput,
            NotebookCellOutputItem: NotebookCellOutputItem,
            CellErrorStackFrame: CellErrorStackFrame,
            NotebookCellStatusBarItem: NotebookCellStatusBarItem,
            NotebookControllerAffinity: NotebookControllerAffinity,
            NotebookControllerAffinity2: NotebookControllerAffinity2,
            NotebookEdit: NotebookEdit,
            NotebookKernelSourceAction: NotebookKernelSourceAction,
            NotebookVariablesRequestKind: NotebookVariablesRequestKind,
            PortAttributes: PortAttributes,
            LinkedEditingRanges: LinkedEditingRanges,
            TestResultState: TestResultState,
            TestRunRequest: TestRunRequest,
            TestMessage: TestMessage,
            TestMessageStackFrame: TestMessageStackFrame,
            TestTag: TestTag,
            TestRunProfileKind: TestRunProfileKind,
            TextSearchCompleteMessageType: TextSearchCompleteMessageType,
            DataTransfer: DataTransfer,
            DataTransferItem: DataTransferItem,
            TestCoverageCount: TestCoverageCount,
            FileCoverage: FileCoverage,
            StatementCoverage: StatementCoverage,
            BranchCoverage: BranchCoverage,
            DeclarationCoverage: DeclarationCoverage,
            WorkspaceTrustState: WorkspaceTrustState,
            LanguageStatusSeverity: LanguageStatusSeverity,
            QuickPickItemKind: QuickPickItemKind,
            InputBoxValidationSeverity: InputBoxValidationSeverity,
            TabInputText: TextTabInput,
            TabInputTextDiff: TextDiffTabInput,
            TabInputTextMerge: TextMergeTabInput,
            TabInputCustom: CustomEditorTabInput,
            TabInputNotebook: NotebookEditorTabInput,
            TabInputNotebookDiff: NotebookDiffEditorTabInput,
            TabInputWebview: WebviewEditorTabInput,
            TabInputTerminal: TerminalEditorTabInput,
            TabInputInteractiveWindow: InteractiveWindowInput,
            TabInputChat: ChatEditorTabInput,
            TabInputTextMultiDiff: TextMultiDiffTabInput,
            TelemetryTrustedValue: TelemetryTrustedValue,
            LogLevel: LogLevel,
            EditSessionIdentityMatch: EditSessionIdentityMatch,
            InteractiveSessionVoteDirection: InteractiveSessionVoteDirection,
            ChatCopyKind: ChatCopyKind,
            ChatSessionChangedFile: ChatSessionChangedFile,
            ChatEditingSessionActionOutcome: ChatEditingSessionActionOutcome,
            InteractiveEditorResponseFeedbackKind: InteractiveEditorResponseFeedbackKind,
            DebugStackFrame: DebugStackFrame,
            DebugThread: DebugThread,
            RelatedInformationType: RelatedInformationType,
            SpeechToTextStatus: SpeechToTextStatus,
            TextToSpeechStatus: TextToSpeechStatus,
            PartialAcceptTriggerKind: PartialAcceptTriggerKind,
            InlineCompletionEndOfLifeReasonKind: InlineCompletionEndOfLifeReasonKind,
            InlineCompletionDisplayLocationKind: InlineCompletionDisplayLocationKind,
            KeywordRecognitionStatus: KeywordRecognitionStatus,
            ChatImageMimeType: ChatImageMimeType,
            ChatResponseMarkdownPart: ChatResponseMarkdownPart,
            ChatResponseFileTreePart: ChatResponseFileTreePart,
            ChatResponseAnchorPart: ChatResponseAnchorPart,
            ChatResponseProgressPart: ChatResponseProgressPart,
            ChatResponseProgressPart2: ChatResponseProgressPart2,
            ChatResponseThinkingProgressPart: ChatResponseThinkingProgressPart,
            ChatResponseReferencePart: ChatResponseReferencePart,
            ChatResponseReferencePart2: ChatResponseReferencePart,
            ChatResponseCodeCitationPart: ChatResponseCodeCitationPart,
            ChatResponseCodeblockUriPart: ChatResponseCodeblockUriPart,
            ChatResponseWarningPart: ChatResponseWarningPart,
            ChatResponseTextEditPart: ChatResponseTextEditPart,
            ChatResponseNotebookEditPart: ChatResponseNotebookEditPart,
            ChatResponseMarkdownWithVulnerabilitiesPart: ChatResponseMarkdownWithVulnerabilitiesPart,
            ChatResponseCommandButtonPart: ChatResponseCommandButtonPart,
            ChatResponseConfirmationPart: ChatResponseConfirmationPart,
            ChatResponseMovePart: ChatResponseMovePart,
            ChatResponseExtensionsPart: ChatResponseExtensionsPart,
            ChatResponseExternalEditPart: ChatResponseExternalEditPart,
            ChatResponsePullRequestPart: ChatResponsePullRequestPart,
            ChatPrepareToolInvocationPart: ChatPrepareToolInvocationPart,
            ChatResponseMultiDiffPart: ChatResponseMultiDiffPart,
            ChatResponseReferencePartStatusKind: ChatResponseReferencePartStatusKind,
            ChatResponseClearToPreviousToolInvocationReason: ChatResponseClearToPreviousToolInvocationReason,
            ChatRequestTurn: ChatRequestTurn,
            ChatRequestTurn2: ChatRequestTurn,
            ChatResponseTurn: ChatResponseTurn,
            ChatResponseTurn2: ChatResponseTurn2,
            ChatToolInvocationPart: ChatToolInvocationPart,
            ChatLocation: ChatLocation,
            ChatSessionStatus: ChatSessionStatus,
            ChatRequestEditorData: ChatRequestEditorData,
            ChatRequestNotebookData: ChatRequestNotebookData,
            ChatReferenceBinaryData: ChatReferenceBinaryData,
            ChatRequestEditedFileEventKind: ChatRequestEditedFileEventKind,
            LanguageModelChatMessageRole: LanguageModelChatMessageRole,
            LanguageModelChatMessage: LanguageModelChatMessage,
            LanguageModelChatMessage2: LanguageModelChatMessage2,
            LanguageModelToolResultPart: LanguageModelToolResultPart,
            LanguageModelToolResultPart2: LanguageModelToolResultPart,
            LanguageModelTextPart: LanguageModelTextPart,
            LanguageModelTextPart2: LanguageModelTextPart,
            LanguageModelPartAudience: LanguageModelPartAudience,
            ToolResultAudience: LanguageModelPartAudience,
            LanguageModelToolCallPart: LanguageModelToolCallPart,
            LanguageModelThinkingPart: LanguageModelThinkingPart,
            LanguageModelError: LanguageModelError,
            LanguageModelToolResult: LanguageModelToolResult,
            LanguageModelToolResult2: LanguageModelToolResult2,
            LanguageModelDataPart: LanguageModelDataPart,
            LanguageModelDataPart2: LanguageModelDataPart,
            LanguageModelToolExtensionSource: LanguageModelToolExtensionSource,
            LanguageModelToolMCPSource: LanguageModelToolMCPSource,
            ExtendedLanguageModelToolResult: ExtendedLanguageModelToolResult,
            LanguageModelChatToolMode: LanguageModelChatToolMode,
            LanguageModelPromptTsxPart: LanguageModelPromptTsxPart,
            NewSymbolName: NewSymbolName,
            NewSymbolNameTag: NewSymbolNameTag,
            NewSymbolNameTriggerKind: NewSymbolNameTriggerKind,
            ExcludeSettingOptions: ExcludeSettingOptions,
            TextSearchContext2: TextSearchContext2,
            TextSearchMatch2: TextSearchMatch2,
            AISearchKeyword: AISearchKeyword,
            TextSearchCompleteMessageTypeNew: TextSearchCompleteMessageType,
            ChatErrorLevel: ChatErrorLevel,
            McpHttpServerDefinition: McpHttpServerDefinition,
            McpHttpServerDefinition2: McpHttpServerDefinition,
            McpStdioServerDefinition: McpStdioServerDefinition,
            McpStdioServerDefinition2: McpStdioServerDefinition,
            McpToolAvailability: McpToolAvailability,
            SettingsSearchResultKind: SettingsSearchResultKind,
        };
    };
}

export { createApiFactoryAndRegisterActors };
