// @codingame/monaco-vscode-api/vscode/src/vs/editor/contrib/inlineCompletions/browser/controller/commands.js

import { KeyMod as KeyMod$1, KeyCode } from '../../../../../base/common/keyCodes.js';
import '../../../../../base/common/observableInternal/index.js';
import { splitLines } from '../../../../../base/common/strings.js';
import { vUnion, vObj, vOptionalProp, vBoolean, vWithJsonSchemaRef, vString, vUndefined } from '../../../../../base/common/validation.js';
import { localize2, localize } from '../../../../../nls.js';
import { CONTEXT_ACCESSIBILITY_MODE_ENABLED } from '../../../../../platform/accessibility/common/accessibility.js';
import { MenuId, Action2 } from '../../../../../platform/actions/common/actions.js';
import { IClipboardService } from '../../../../../platform/clipboard/common/clipboardService.service.js';
import { IConfigurationService } from '../../../../../platform/configuration/common/configuration.service.js';
import { ContextKeyExpr } from '../../../../../platform/contextkey/common/contextkey.js';
import { KeybindingWeight, KeybindingsRegistry } from '../../../../../platform/keybinding/common/keybindingsRegistry.js';
import '../../../../../platform/notification/common/notification.js';
import { INotificationService } from '../../../../../platform/notification/common/notification.service.js';
import { EditorAction } from '../../../../browser/editorExtensions.js';
import { EditorContextKeys } from '../../../../common/editorContextKeys.js';
import { ILanguageFeaturesService } from '../../../../common/services/languageFeatures.service.js';
import { Context } from '../../../suggest/browser/suggest.js';
import { showNextInlineSuggestionActionId, showPreviousInlineSuggestionActionId, inlineSuggestCommitId, inlineSuggestCommitAlternativeActionId, jumpToNextInlineEditId, hideInlineCompletionId, toggleShowCollapsedId } from './commandIds.js';
import { InlineCompletionContextKeys } from './inlineCompletionContextKeys.js';
import { InlineCompletionsController } from './inlineCompletionsController.js';
import { asyncTransaction, transaction } from '../../../../../base/common/observableInternal/transaction.js';
import Severity$1 from '../../../../../base/common/severity.js';

class ShowNextInlineSuggestionAction extends EditorAction {
    static { this.ID = showNextInlineSuggestionActionId; }
    constructor() {
        super({
            id: ShowNextInlineSuggestionAction.ID,
            label: ( localize2(1274, "Show Next Inline Suggestion")),
            precondition: ( ContextKeyExpr.and(
                EditorContextKeys.writable,
                InlineCompletionContextKeys.inlineSuggestionVisible
            )),
            kbOpts: {
                weight: 100,
                primary: KeyMod$1.Alt | KeyCode.BracketRight,
            },
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.get(editor);
        controller?.model.get()?.next();
    }
}
class ShowPreviousInlineSuggestionAction extends EditorAction {
    static { this.ID = showPreviousInlineSuggestionActionId; }
    constructor() {
        super({
            id: ShowPreviousInlineSuggestionAction.ID,
            label: ( localize2(1275, "Show Previous Inline Suggestion")),
            precondition: ( ContextKeyExpr.and(
                EditorContextKeys.writable,
                InlineCompletionContextKeys.inlineSuggestionVisible
            )),
            kbOpts: {
                weight: 100,
                primary: KeyMod$1.Alt | KeyCode.BracketLeft,
            },
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.get(editor);
        controller?.model.get()?.previous();
    }
}
const providerIdSchemaUri = 'vscode://schemas/inlineCompletionProviderIdArgs';
function inlineCompletionProviderGetMatcher(provider) {
    const result = [];
    if (provider.providerId) {
        result.push(provider.providerId.toStringWithoutVersion());
        result.push(provider.providerId.extensionId + ':*');
    }
    return result;
}
const argsValidator = vUnion(vObj({
    showNoResultNotification: vOptionalProp(vBoolean()),
    providerId: vOptionalProp(vWithJsonSchemaRef(providerIdSchemaUri, vString())),
    explicit: vOptionalProp(vBoolean()),
}), vUndefined());
class TriggerInlineSuggestionAction extends EditorAction {
    constructor() {
        super({
            id: 'editor.action.inlineSuggest.trigger',
            label: ( localize2(1276, "Trigger Inline Suggestion")),
            precondition: EditorContextKeys.writable,
            metadata: {
                description: ( localize(1277, "Triggers an inline suggestion in the editor.")),
                args: [{
                        name: 'args',
                        description: ( localize(1278, "Options for triggering inline suggestions.")),
                        isOptional: true,
                        schema: argsValidator.getJSONSchema(),
                    }]
            }
        });
    }
    async run(accessor, editor, args) {
        const notificationService = accessor.get(INotificationService);
        const languageFeaturesService = accessor.get(ILanguageFeaturesService);
        const controller = InlineCompletionsController.get(editor);
        const validatedArgs = argsValidator.validateOrThrow(args);
        const provider = validatedArgs?.providerId ?
            languageFeaturesService.inlineCompletionsProvider.all(editor.getModel())
                .find(p => ( inlineCompletionProviderGetMatcher(p).some(m => m === validatedArgs.providerId)))
            : undefined;
        await asyncTransaction(async (tx) => {
            await controller?.model.get()?.trigger(tx, {
                provider: provider,
                explicit: validatedArgs?.explicit ?? true,
            });
            controller?.playAccessibilitySignal(tx);
        });
        if (validatedArgs?.showNoResultNotification) {
            if (!controller?.model.get()?.state.get()) {
                notificationService.notify({
                    severity: Severity$1.Info,
                    message: ( localize(1279, "No inline suggestion is available."))
                });
            }
        }
    }
}
class AcceptNextWordOfInlineCompletion extends EditorAction {
    constructor() {
        super({
            id: 'editor.action.inlineSuggest.acceptNextWord',
            label: ( localize2(1280, "Accept Next Word Of Inline Suggestion")),
            precondition: ( ContextKeyExpr.and(
                EditorContextKeys.writable,
                InlineCompletionContextKeys.inlineSuggestionVisible
            )),
            kbOpts: {
                weight: KeybindingWeight.EditorContrib + 1,
                primary: KeyMod$1.CtrlCmd | KeyCode.RightArrow,
                kbExpr: ( ContextKeyExpr.and(
                    EditorContextKeys.writable,
                    InlineCompletionContextKeys.inlineSuggestionVisible,
                    InlineCompletionContextKeys.cursorBeforeGhostText,
                    ( CONTEXT_ACCESSIBILITY_MODE_ENABLED.negate())
                )),
            },
            menuOpts: [{
                    menuId: MenuId.InlineSuggestionToolbar,
                    title: ( localize(1281, 'Accept Word')),
                    group: 'primary',
                    order: 2,
                }],
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.get(editor);
        await controller?.model.get()?.acceptNextWord();
    }
}
class AcceptNextLineOfInlineCompletion extends EditorAction {
    constructor() {
        super({
            id: 'editor.action.inlineSuggest.acceptNextLine',
            label: ( localize2(1282, "Accept Next Line Of Inline Suggestion")),
            precondition: ( ContextKeyExpr.and(
                EditorContextKeys.writable,
                InlineCompletionContextKeys.inlineSuggestionVisible
            )),
            kbOpts: {
                weight: KeybindingWeight.EditorContrib + 1,
            },
            menuOpts: [{
                    menuId: MenuId.InlineSuggestionToolbar,
                    title: ( localize(1283, 'Accept Line')),
                    group: 'secondary',
                    order: 2,
                }],
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.get(editor);
        await controller?.model.get()?.acceptNextLine();
    }
}
class AcceptInlineCompletion extends EditorAction {
    constructor() {
        super({
            id: inlineSuggestCommitId,
            label: ( localize2(1284, "Accept Inline Suggestion")),
            precondition: ( ContextKeyExpr.or(
                InlineCompletionContextKeys.inlineSuggestionVisible,
                InlineCompletionContextKeys.inlineEditVisible
            )),
            menuOpts: [{
                    menuId: MenuId.InlineSuggestionToolbar,
                    title: ( localize(1285, "Accept")),
                    group: 'primary',
                    order: 2,
                }, {
                    menuId: MenuId.InlineEditsActions,
                    title: ( localize(1285, "Accept")),
                    group: 'primary',
                    order: 2,
                }],
            kbOpts: [
                {
                    primary: KeyCode.Tab,
                    weight: 200,
                    kbExpr: ( ContextKeyExpr.or(( ContextKeyExpr.and(InlineCompletionContextKeys.inlineSuggestionVisible, ( EditorContextKeys.tabMovesFocus.toNegated()), ( Context.Visible.toNegated()), ( EditorContextKeys.hoverFocused.toNegated()), ( InlineCompletionContextKeys.hasSelection.toNegated()), InlineCompletionContextKeys.inlineSuggestionHasIndentationLessThanTabSize)), ( ContextKeyExpr.and(InlineCompletionContextKeys.inlineEditVisible, ( EditorContextKeys.tabMovesFocus.toNegated()), ( Context.Visible.toNegated()), ( EditorContextKeys.hoverFocused.toNegated()), InlineCompletionContextKeys.tabShouldAcceptInlineEdit)))),
                }
            ],
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.getInFocusedEditorOrParent(accessor);
        if (controller) {
            controller.model.get()?.accept(controller.editor);
            controller.editor.focus();
        }
    }
}
KeybindingsRegistry.registerKeybindingRule({
    id: inlineSuggestCommitId,
    weight: 202,
    primary: KeyCode.Tab,
    when: ( ContextKeyExpr.and(InlineCompletionContextKeys.inInlineEditsPreviewEditor))
});
class AcceptInlineCompletionAlternativeAction extends EditorAction {
    constructor() {
        super({
            id: inlineSuggestCommitAlternativeActionId,
            label: ( localize2(1286, "Accept Inline Suggestion Alternative Action")),
            precondition: ( ContextKeyExpr.and(
                InlineCompletionContextKeys.inlineSuggestionAlternativeActionVisible,
                InlineCompletionContextKeys.inlineEditVisible
            )),
            menuOpts: [],
            kbOpts: [
                {
                    primary: KeyMod$1.Shift | KeyCode.Tab,
                    weight: 203,
                }
            ],
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.getInFocusedEditorOrParent(accessor);
        if (controller) {
            controller.model.get()?.accept(controller.editor, true);
            controller.editor.focus();
        }
    }
}
KeybindingsRegistry.registerKeybindingRule({
    id: inlineSuggestCommitAlternativeActionId,
    weight: 203,
    primary: KeyMod$1.Shift | KeyCode.Tab,
    when: ( ContextKeyExpr.and(InlineCompletionContextKeys.inInlineEditsPreviewEditor))
});
class JumpToNextInlineEdit extends EditorAction {
    constructor() {
        super({
            id: jumpToNextInlineEditId,
            label: ( localize2(1287, "Jump to next inline edit")),
            precondition: InlineCompletionContextKeys.inlineEditVisible,
            menuOpts: [{
                    menuId: MenuId.InlineEditsActions,
                    title: ( localize(1288, "Jump")),
                    group: 'primary',
                    order: 1,
                    when: ( InlineCompletionContextKeys.cursorAtInlineEdit.toNegated()),
                }],
            kbOpts: {
                primary: KeyCode.Tab,
                weight: 201,
                kbExpr: ( ContextKeyExpr.and(InlineCompletionContextKeys.inlineEditVisible, ( EditorContextKeys.tabMovesFocus.toNegated()), ( Context.Visible.toNegated()), ( EditorContextKeys.hoverFocused.toNegated()), InlineCompletionContextKeys.tabShouldJumpToInlineEdit)),
            }
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.get(editor);
        if (controller) {
            controller.jump();
        }
    }
}
class HideInlineCompletion extends EditorAction {
    static { this.ID = hideInlineCompletionId; }
    constructor() {
        super({
            id: HideInlineCompletion.ID,
            label: ( localize2(1289, "Hide Inline Suggestion")),
            precondition: ( ContextKeyExpr.or(
                InlineCompletionContextKeys.inlineSuggestionVisible,
                InlineCompletionContextKeys.inlineEditVisible
            )),
            kbOpts: {
                weight: KeybindingWeight.EditorContrib + 90,
                primary: KeyCode.Escape,
            },
            menuOpts: [{
                    menuId: MenuId.InlineEditsActions,
                    title: ( localize(1290, "Reject")),
                    group: 'primary',
                    order: 3,
                }]
        });
    }
    async run(accessor, editor) {
        const controller = InlineCompletionsController.getInFocusedEditorOrParent(accessor);
        transaction(tx => {
            controller?.model.get()?.stop('explicitCancel', tx);
        });
        controller?.editor.focus();
    }
}
class ToggleInlineCompletionShowCollapsed extends EditorAction {
    static { this.ID = toggleShowCollapsedId; }
    constructor() {
        super({
            id: ToggleInlineCompletionShowCollapsed.ID,
            label: ( localize2(1291, "Toggle Inline Suggestions Show Collapsed")),
            precondition: ContextKeyExpr.true(),
        });
    }
    async run(accessor, editor) {
        const configurationService = accessor.get(IConfigurationService);
        const showCollapsed = configurationService.getValue('editor.inlineSuggest.edits.showCollapsed');
        configurationService.updateValue('editor.inlineSuggest.edits.showCollapsed', !showCollapsed);
    }
}
KeybindingsRegistry.registerKeybindingRule({
    id: HideInlineCompletion.ID,
    weight: -1,
    primary: KeyCode.Escape,
    secondary: [KeyMod$1.Shift | KeyCode.Escape],
    when: ( ContextKeyExpr.and(InlineCompletionContextKeys.inInlineEditsPreviewEditor))
});
class ToggleAlwaysShowInlineSuggestionToolbar extends Action2 {
    static { this.ID = 'editor.action.inlineSuggest.toggleAlwaysShowToolbar'; }
    constructor() {
        super({
            id: ToggleAlwaysShowInlineSuggestionToolbar.ID,
            title: ( localize(1292, "Always Show Toolbar")),
            f1: false,
            precondition: undefined,
            menu: [{
                    id: MenuId.InlineSuggestionToolbar,
                    group: 'secondary',
                    order: 10,
                }],
            toggled: ( ContextKeyExpr.equals('config.editor.inlineSuggest.showToolbar', 'always'))
        });
    }
    async run(accessor) {
        const configService = accessor.get(IConfigurationService);
        const currentValue = configService.getValue('editor.inlineSuggest.showToolbar');
        const newValue = currentValue === 'always' ? 'onHover' : 'always';
        configService.updateValue('editor.inlineSuggest.showToolbar', newValue);
    }
}
class DevExtractReproSample extends EditorAction {
    constructor() {
        super({
            id: 'editor.action.inlineSuggest.dev.extractRepro',
            label: ( localize(1293, "Developer: Extract Inline Suggest State")),
            alias: 'Developer: Inline Suggest Extract Repro',
            precondition: ( ContextKeyExpr.or(
                InlineCompletionContextKeys.inlineEditVisible,
                InlineCompletionContextKeys.inlineSuggestionVisible
            )),
        });
    }
    async run(accessor, editor) {
        const clipboardService = accessor.get(IClipboardService);
        const controller = InlineCompletionsController.get(editor);
        const m = controller?.model.get();
        if (!m) {
            return;
        }
        const repro = m.extractReproSample();
        const inlineCompletionLines = splitLines(JSON.stringify({ inlineCompletion: repro.inlineCompletion }, null, 4));
        const json = ( inlineCompletionLines.map(l => '// ' + l)).join('\n');
        const reproStr = `${repro.documentValue}\n\n// <json>\n${json}\n// </json>\n`;
        await clipboardService.writeText(reproStr);
        return { reproCase: reproStr };
    }
}

export { AcceptInlineCompletion, AcceptInlineCompletionAlternativeAction, AcceptNextLineOfInlineCompletion, AcceptNextWordOfInlineCompletion, DevExtractReproSample, HideInlineCompletion, JumpToNextInlineEdit, ShowNextInlineSuggestionAction, ShowPreviousInlineSuggestionAction, ToggleAlwaysShowInlineSuggestionToolbar, ToggleInlineCompletionShowCollapsed, TriggerInlineSuggestionAction, inlineCompletionProviderGetMatcher, providerIdSchemaUri };
