// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/mcp/common/mcpConfiguration.js

import { MarkdownString } from '../../../../base/common/htmlContent.js';
import { Disposable } from '../../../../base/common/lifecycle.js';
import { localize } from '../../../../nls.js';
import { SyncDescriptor } from '../../../../platform/instantiation/common/descriptors.js';
import { Registry } from '../../../../platform/registry/common/platform.js';
import { mcpSchemaId } from '../../../services/configuration/common/configuration.js';
import { inputsSchema } from '../../../services/configurationResolver/common/configurationResolverSchema.js';
import { Extensions } from '../../../services/extensionManagement/common/extensionFeatures.js';

const mcpActivationEventPrefix = 'onMcpCollection:';
const mcpActivationEvent = (contributedCollectionId) => mcpActivationEventPrefix + contributedCollectionId;
var DiscoverySource;
(function (DiscoverySource) {
    DiscoverySource["ClaudeDesktop"] = "claude-desktop";
    DiscoverySource["Windsurf"] = "windsurf";
    DiscoverySource["CursorGlobal"] = "cursor-global";
    DiscoverySource["CursorWorkspace"] = "cursor-workspace";
})(DiscoverySource || (DiscoverySource = {}));
const allDiscoverySources = ( Object.keys({
    [DiscoverySource.ClaudeDesktop]: true,
    [DiscoverySource.Windsurf]: true,
    [DiscoverySource.CursorGlobal]: true,
    [DiscoverySource.CursorWorkspace]: true,
}));
const discoverySourceLabel = {
    [DiscoverySource.ClaudeDesktop]: ( localize(8989, "Claude Desktop")),
    [DiscoverySource.Windsurf]: ( localize(8990, "Windsurf")),
    [DiscoverySource.CursorGlobal]: ( localize(8991, "Cursor (Global)")),
    [DiscoverySource.CursorWorkspace]: ( localize(8992, "Cursor (Workspace)")),
};
const discoverySourceSettingsLabel = {
    [DiscoverySource.ClaudeDesktop]: ( localize(8993, "Claude Desktop configuration (`claude_desktop_config.json`)")),
    [DiscoverySource.Windsurf]: ( localize(8994, "Windsurf configurations (`~/.codeium/windsurf/mcp_config.json`)")),
    [DiscoverySource.CursorGlobal]: ( localize(8995, "Cursor global configuration (`~/.cursor/mcp.json`)")),
    [DiscoverySource.CursorWorkspace]: ( localize(8996, "Cursor workspace configuration (`.cursor/mcp.json`)")),
};
const mcpConfigurationSection = 'mcp';
const mcpDiscoverySection = 'chat.mcp.discovery.enabled';
const mcpServerSamplingSection = 'chat.mcp.serverSampling';
const mcpSchemaExampleServers = {
    'mcp-server-time': {
        command: 'python',
        args: ['-m', 'mcp_server_time', '--local-timezone=America/Los_Angeles'],
        env: {},
    }
};
const httpSchemaExamples = {
    'my-mcp-server': {
        url: 'http://localhost:3001/mcp',
        headers: {},
    }
};
const mcpDevModeProps = (stdio) => ({
    dev: {
        type: 'object',
        markdownDescription: ( localize(
            8997,
            'Enabled development mode for the server. When present, the server will be started eagerly and output will be included in its output. Properties inside the `dev` object can configure additional behavior.'
        )),
        examples: [{ watch: 'src/**/*.ts', debug: { type: 'node' } }],
        properties: {
            watch: {
                description: ( localize(
                    8998,
                    'A glob pattern or list of glob patterns relative to the workspace folder to watch. The MCP server will be restarted when these files change.'
                )),
                examples: ['src/**/*.ts'],
                oneOf: [{ type: 'string' }, { type: 'array', items: { type: 'string' } }],
            },
            ...(stdio && {
                debug: {
                    markdownDescription: ( localize(
                        8999,
                        'If set, debugs the MCP server using the given runtime as it\'s started.'
                    )),
                    oneOf: [
                        {
                            type: 'object',
                            required: ['type'],
                            properties: {
                                type: {
                                    type: 'string',
                                    enum: ['node'],
                                    description: ( localize(9000, "Debug the MCP server using Node.js."))
                                }
                            },
                            additionalProperties: false
                        },
                        {
                            type: 'object',
                            required: ['type'],
                            properties: {
                                type: {
                                    type: 'string',
                                    enum: ['debugpy'],
                                    description: ( localize(9001, "Debug the MCP server using Python and debugpy."))
                                },
                                debugpyPath: {
                                    type: 'string',
                                    description: ( localize(9002, "Path to the debugpy executable."))
                                },
                            },
                            additionalProperties: false
                        }
                    ]
                }
            })
        }
    }
});
const mcpStdioServerSchema = {
    type: 'object',
    additionalProperties: false,
    examples: [mcpSchemaExampleServers['mcp-server-time']],
    properties: {
        type: {
            type: 'string',
            enum: ['stdio'],
            description: ( localize(9003, "The type of the server."))
        },
        command: {
            type: 'string',
            description: ( localize(9004, "The command to run the server."))
        },
        cwd: {
            type: 'string',
            description: ( localize(
                9005,
                "The working directory for the server command. Defaults to the workspace folder when run in a workspace."
            )),
            examples: ['${workspaceFolder}'],
        },
        args: {
            type: 'array',
            description: ( localize(9006, "Arguments passed to the server.")),
            items: {
                type: 'string'
            },
        },
        envFile: {
            type: 'string',
            description: ( localize(9007, "Path to a file containing environment variables for the server.")),
            examples: ['${workspaceFolder}/.env'],
        },
        env: {
            description: ( localize(9008, "Environment variables passed to the server.")),
            additionalProperties: {
                anyOf: [
                    { type: 'null' },
                    { type: 'string' },
                    { type: 'number' },
                ]
            }
        },
        ...mcpDevModeProps(true),
    }
};
const mcpServerSchema = {
    id: mcpSchemaId,
    type: 'object',
    title: ( localize(9009, "Model Context Protocol Servers")),
    allowTrailingCommas: true,
    allowComments: true,
    additionalProperties: false,
    properties: {
        servers: {
            examples: [
                mcpSchemaExampleServers,
                httpSchemaExamples,
            ],
            additionalProperties: {
                oneOf: [
                    mcpStdioServerSchema, {
                        type: 'object',
                        additionalProperties: false,
                        required: ['url'],
                        examples: [httpSchemaExamples['my-mcp-server']],
                        properties: {
                            type: {
                                type: 'string',
                                enum: ['http', 'sse'],
                                description: ( localize(9003, "The type of the server."))
                            },
                            url: {
                                type: 'string',
                                format: 'uri',
                                pattern: '^https?:\\/\\/.+',
                                patternErrorMessage: ( localize(9010, "The URL must start with 'http://' or 'https://'.")),
                                description: ( localize(9011, "The URL of the Streamable HTTP or SSE endpoint."))
                            },
                            headers: {
                                type: 'object',
                                description: ( localize(9012, "Additional headers sent to the server.")),
                                additionalProperties: { type: 'string' },
                            },
                            ...mcpDevModeProps(false),
                        }
                    },
                ]
            }
        },
        inputs: inputsSchema.definitions.inputs
    }
};
const mcpContributionPoint = {
    extensionPoint: 'mcpServerDefinitionProviders',
    activationEventsGenerator: function* (contribs) {
        for (const contrib of contribs) {
            if (contrib.id) {
                yield mcpActivationEvent(contrib.id);
            }
        }
    },
    jsonSchema: {
        description: ( localize(
            9013,
            'Contributes Model Context Protocol servers. Users of this should also use `vscode.lm.registerMcpServerDefinitionProvider`.'
        )),
        type: 'array',
        defaultSnippets: [{ body: [{ id: '', label: '' }] }],
        items: {
            additionalProperties: false,
            type: 'object',
            defaultSnippets: [{ body: { id: '', label: '' } }],
            properties: {
                id: {
                    description: ( localize(9014, "Unique ID for the collection.")),
                    type: 'string'
                },
                label: {
                    description: ( localize(9015, "Display name for the collection.")),
                    type: 'string'
                },
                when: {
                    description: ( localize(9016, "Condition which must be true to enable this collection.")),
                    type: 'string'
                }
            }
        }
    }
};
class McpServerDefinitionsProviderRenderer extends Disposable {
    constructor() {
        super(...arguments);
        this.type = 'table';
    }
    shouldRender(manifest) {
        return !!manifest.contributes?.mcpServerDefinitionProviders && Array.isArray(manifest.contributes.mcpServerDefinitionProviders) && manifest.contributes.mcpServerDefinitionProviders.length > 0;
    }
    render(manifest) {
        const mcpServerDefinitionProviders = manifest.contributes?.mcpServerDefinitionProviders ?? [];
        const headers = [( localize(9017, "ID")), ( localize(9018, "Name"))];
        const rows = ( mcpServerDefinitionProviders
            .map(mcpServerDefinitionProvider => {
            return [
                ( new MarkdownString()).appendMarkdown(`\`${mcpServerDefinitionProvider.id}\``),
                mcpServerDefinitionProvider.label
            ];
        }));
        return {
            data: {
                headers,
                rows
            },
            dispose: () => { }
        };
    }
}
( Registry.as(Extensions.ExtensionFeaturesRegistry)).registerExtensionFeature({
    id: mcpConfigurationSection,
    label: ( localize(9019, "MCP Servers")),
    access: {
        canToggle: false
    },
    renderer: ( new SyncDescriptor(McpServerDefinitionsProviderRenderer)),
});

export { DiscoverySource, allDiscoverySources, discoverySourceLabel, discoverySourceSettingsLabel, mcpActivationEvent, mcpConfigurationSection, mcpContributionPoint, mcpDiscoverySection, mcpSchemaExampleServers, mcpServerSamplingSection, mcpServerSchema, mcpStdioServerSchema };
