// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminalContrib/chatAgentTools/common/terminalChatAgentToolsConfiguration.js

import { localize } from '../../../../../nls.js';
import { TerminalSettingId } from '../../../../../platform/terminal/common/terminal.js';
import product from '../../../../../platform/product/common/product.js';
import { terminalProfileBaseProperties } from '../../../../../platform/terminal/common/terminalPlatformConfiguration.js';
import { PolicyCategory } from '../../../../../base/common/policy.js';

var TerminalChatAgentToolsSettingId;
(function (TerminalChatAgentToolsSettingId) {
    TerminalChatAgentToolsSettingId["EnableAutoApprove"] = "chat.tools.terminal.enableAutoApprove";
    TerminalChatAgentToolsSettingId["AutoApprove"] = "chat.tools.terminal.autoApprove";
    TerminalChatAgentToolsSettingId["AutoApproveWorkspaceNpmScripts"] = "chat.tools.terminal.autoApproveWorkspaceNpmScripts";
    TerminalChatAgentToolsSettingId["IgnoreDefaultAutoApproveRules"] = "chat.tools.terminal.ignoreDefaultAutoApproveRules";
    TerminalChatAgentToolsSettingId["BlockDetectedFileWrites"] = "chat.tools.terminal.blockDetectedFileWrites";
    TerminalChatAgentToolsSettingId["ShellIntegrationTimeout"] = "chat.tools.terminal.shellIntegrationTimeout";
    TerminalChatAgentToolsSettingId["AutoReplyToPrompts"] = "chat.tools.terminal.autoReplyToPrompts";
    TerminalChatAgentToolsSettingId["OutputLocation"] = "chat.tools.terminal.outputLocation";
    TerminalChatAgentToolsSettingId["PreventShellHistory"] = "chat.tools.terminal.preventShellHistory";
    TerminalChatAgentToolsSettingId["TerminalProfileLinux"] = "chat.tools.terminal.terminalProfile.linux";
    TerminalChatAgentToolsSettingId["TerminalProfileMacOs"] = "chat.tools.terminal.terminalProfile.osx";
    TerminalChatAgentToolsSettingId["TerminalProfileWindows"] = "chat.tools.terminal.terminalProfile.windows";
    TerminalChatAgentToolsSettingId["DeprecatedAutoApproveCompatible"] = "chat.agent.terminal.autoApprove";
    TerminalChatAgentToolsSettingId["DeprecatedAutoApprove1"] = "chat.agent.terminal.allowList";
    TerminalChatAgentToolsSettingId["DeprecatedAutoApprove2"] = "chat.agent.terminal.denyList";
    TerminalChatAgentToolsSettingId["DeprecatedAutoApprove3"] = "github.copilot.chat.agent.terminal.allowList";
    TerminalChatAgentToolsSettingId["DeprecatedAutoApprove4"] = "github.copilot.chat.agent.terminal.denyList";
})(TerminalChatAgentToolsSettingId || (TerminalChatAgentToolsSettingId = {}));
const autoApproveBoolean = {
    type: 'boolean',
    enum: [
        true,
        false,
    ],
    enumDescriptions: [
        ( localize(12096, "Automatically approve the pattern.")),
        ( localize(12097, "Require explicit approval for the pattern.")),
    ],
    description: ( localize(
        12098,
        "The start of a command to match against. A regular expression can be provided by wrapping the string in `/` characters."
    )),
};
const terminalChatAgentProfileSchema = {
    type: 'object',
    required: ['path'],
    properties: {
        path: {
            description: ( localize(12099, "A path to a shell executable.")),
            type: 'string',
        },
        ...terminalProfileBaseProperties,
    }
};
const terminalChatAgentToolsConfiguration = {
    [TerminalChatAgentToolsSettingId.EnableAutoApprove]: {
        description: ( localize(
            12100,
            "Controls whether to allow auto approval in the run in terminal tool."
        )),
        type: 'boolean',
        default: true,
        policy: {
            name: 'ChatToolsTerminalEnableAutoApprove',
            category: PolicyCategory.IntegratedTerminal,
            minimumVersion: '1.104',
            localization: {
                description: {
                    key: 'autoApproveMode.description',
                    value: ( localize(
                        12100,
                        "Controls whether to allow auto approval in the run in terminal tool."
                    )),
                }
            }
        }
    },
    [TerminalChatAgentToolsSettingId.AutoApprove]: {
        markdownDescription: [
            ( localize(
            12101,
            "A list of commands or regular expressions that control whether the run in terminal tool commands require explicit approval. These will be matched against the start of a command. A regular expression can be provided by wrapping the string in {0} characters followed by optional flags such as {1} for case-insensitivity.",
            '`/`',
            '`i`'
        )),
            ( localize(
            12102,
            "Set to {0} to automatically approve commands, {1} to always require explicit approval or {2} to unset the value.",
            '`true`',
            '`false`',
            '`null`'
        )),
            ( localize(
            12103,
            "Note that these commands and regular expressions are evaluated for every _sub-command_ within the full _command line_, so {0} for example will need both {1} and {2} to match a {3} entry and must not match a {4} entry in order to auto approve. Inline commands such as {5} (process substitution) should also be detected.",
            '`foo && bar`',
            '`foo`',
            '`bar`',
            '`true`',
            '`false`',
            '`<(foo)`'
        )),
            ( localize(
            12104,
            "An object can be used to match against the full command line instead of matching sub-commands and inline commands, for example {0}. In order to be auto approved _both_ the sub-command and command line must not be explicitly denied, then _either_ all sub-commands or command line needs to be approved.",
            '`{ approve: false, matchCommandLine: true }`'
        )),
            ( localize(
            12105,
            "Note that there's a default set of rules to allow and also deny commands. Consider setting {0} to {1} to ignore all default rules to ensure there are no conflicts with your own rules. Do this at your own risk, the default denial rules are designed to protect you against running dangerous commands.",
            `\`#${TerminalChatAgentToolsSettingId.IgnoreDefaultAutoApproveRules}#\``,
            '`true`'
        )),
            [
                ( localize(12106, 'Examples:')),
                `|${( localize(12107, "Value"))}|${( localize(12108, "Description"))}|`,
                '|---|---|',
                '| `\"mkdir\": true` | ' + ( localize(12109, "Allow all commands starting with {0}", '`mkdir`')),
                '| `\"npm run build\": true` | ' + ( localize(12110, "Allow all commands starting with {0}", '`npm run build`')),
                '| `\"bin/test.sh\": true` | ' + ( localize(
                    12111,
                    "Allow all commands that match the path {0} ({1}, {2}, etc.)",
                    '`bin/test.sh`',
                    '`bin\\test.sh`',
                    '`./bin/test.sh`'
                )),
                '| `\"/^git (status\\|show\\\\b.*)$/\": true` | ' + ( localize(
                    12112,
                    "Allow {0} and all commands starting with {1}",
                    '`git status`',
                    '`git show`'
                )),
                '| `\"/^Get-ChildItem\\\\b/i\": true` | ' + ( localize(12113, "will allow {0} commands regardless of casing", '`Get-ChildItem`')),
                '| `\"/.*/\": true` | ' + ( localize(12114, "Allow all commands (denied commands still require approval)")),
                '| `\"rm\": false` | ' + ( localize(
                    12115,
                    "Require explicit approval for all commands starting with {0}",
                    '`rm`'
                )),
                '| `\"/\\\\.ps1/i\": { approve: false, matchCommandLine: true }` | ' + ( localize(
                    12116,
                    "Require explicit approval for any _command line_ that contains {0} regardless of casing",
                    '`".ps1"`'
                )),
                '| `\"rm\": null` | ' + ( localize(12117, "Unset the default {0} value for {1}", '`false`', '`rm`')),
            ].join('\n'),
        ].join('\n\n'),
        type: 'object',
        additionalProperties: {
            anyOf: [
                autoApproveBoolean,
                {
                    type: 'object',
                    properties: {
                        approve: autoApproveBoolean,
                        matchCommandLine: {
                            type: 'boolean',
                            enum: [
                                true,
                                false,
                            ],
                            enumDescriptions: [
                                ( localize(12118, "Match against the full command line, eg. `foo && bar`.")),
                                ( localize(
                                12119,
                                "Match against sub-commands and inline commands, eg. `foo && bar` will need both `foo` and `bar` to match."
                            )),
                            ],
                            description: ( localize(
                                12120,
                                "Whether to match against the full command line, as opposed to splitting by sub-commands and inline commands."
                            )),
                        }
                    },
                    required: ['approve']
                },
                {
                    type: 'null',
                    description: ( localize(
                        12121,
                        "Ignore the pattern, this is useful for unsetting the same pattern set at a higher scope."
                    )),
                },
            ]
        },
        default: {
            cd: true,
            echo: true,
            ls: true,
            pwd: true,
            cat: true,
            head: true,
            tail: true,
            findstr: true,
            wc: true,
            tr: true,
            cut: true,
            cmp: true,
            which: true,
            basename: true,
            dirname: true,
            realpath: true,
            readlink: true,
            stat: true,
            file: true,
            du: true,
            df: true,
            sleep: true,
            nl: true,
            grep: true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+status\\b/': true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+log\\b/': true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+show\\b/': true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+diff\\b/': true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+ls-files\\b/': true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+grep\\b/': true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+branch\\b/': true,
            '/^git(\\s+(-C\\s+\\S+|--no-pager))*\\s+branch\\b.*-(d|D|m|M|-delete|-force)\\b/': false,
            'Get-ChildItem': true,
            'Get-Content': true,
            'Get-Date': true,
            'Get-Random': true,
            'Get-Location': true,
            'Write-Host': true,
            'Write-Output': true,
            'Out-String': true,
            'Split-Path': true,
            'Join-Path': true,
            'Start-Sleep': true,
            'Where-Object': true,
            '/^Select-[a-z0-9]/i': true,
            '/^Measure-[a-z0-9]/i': true,
            '/^Compare-[a-z0-9]/i': true,
            '/^Format-[a-z0-9]/i': true,
            '/^Sort-[a-z0-9]/i': true,
            column: true,
            '/^column\\b.*-c\\s+[0-9]{4,}/': false,
            date: true,
            '/^date\\b.*(-s|--set)\\b/': false,
            find: true,
            '/^find\\b.*-(delete|exec|execdir|fprint|fprintf|fls|ok|okdir)\\b/': false,
            rg: true,
            '/^rg\\b.*(--pre|--hostname-bin)\\b/': false,
            sed: true,
            '/^sed\\b.*(-[a-zA-Z]*(e|i|I|f)[a-zA-Z]*|--expression|--file|--in-place)\\b/': false,
            '/^sed\\b.*(\/e|\/w|;W)/': false,
            sort: true,
            '/^sort\\b.*-(o|S)\\b/': false,
            tree: true,
            '/^tree\\b.*-o\\b/': false,
            rm: false,
            rmdir: false,
            del: false,
            'Remove-Item': false,
            ri: false,
            rd: false,
            erase: false,
            dd: false,
            kill: false,
            ps: false,
            top: false,
            'Stop-Process': false,
            spps: false,
            taskkill: false,
            'taskkill.exe': false,
            curl: false,
            wget: false,
            'Invoke-RestMethod': false,
            'Invoke-WebRequest': false,
            'irm': false,
            'iwr': false,
            chmod: false,
            chown: false,
            'Set-ItemProperty': false,
            'sp': false,
            'Set-Acl': false,
            jq: false,
            xargs: false,
            eval: false,
            'Invoke-Expression': false,
            iex: false,
        },
    },
    [TerminalChatAgentToolsSettingId.IgnoreDefaultAutoApproveRules]: {
        type: 'boolean',
        default: false,
        tags: ['experimental'],
        markdownDescription: ( localize(
            12122,
            "Whether to ignore the built-in default auto-approve rules used by the run in terminal tool as defined in {0}. When this setting is enabled, the run in terminal tool will ignore any rule that comes from the default set but still follow rules defined in the user, remote and workspace settings. Use this setting at your own risk; the default auto-approve rules are designed to protect you against running dangerous commands.",
            `\`#${TerminalChatAgentToolsSettingId.AutoApprove}#\``
        )),
    },
    [TerminalChatAgentToolsSettingId.AutoApproveWorkspaceNpmScripts]: {
        restricted: true,
        type: 'boolean',
        default: true,
        tags: ['experimental'],
        markdownDescription: ( localize(
            12123,
            "Whether to automatically approve npm, yarn, and pnpm run commands when the script is defined in a workspace package.json file. Since the workspace is trusted, scripts defined in package.json are considered safe to run without explicit approval."
        )),
    },
    [TerminalChatAgentToolsSettingId.BlockDetectedFileWrites]: {
        type: 'string',
        enum: ['never', 'outsideWorkspace', 'all'],
        enumDescriptions: [
            ( localize(12124, "Allow all detected file writes.")),
            ( localize(
            12125,
            "Block file writes detected outside the workspace. This depends on the shell integration feature working correctly to determine the current working directory of the terminal."
        )),
            ( localize(12126, "Block all detected file writes.")),
        ],
        default: 'outsideWorkspace',
        tags: ['experimental'],
        markdownDescription: ( localize(
            12127,
            "Controls whether detected file write operations are blocked in the run in terminal tool. When detected, this will require explicit approval regardless of whether the command would normally be auto approved. Note that this cannot detect all possible methods of writing files, this is what is currently detected:\n\n- File redirection (detected via the bash or PowerShell tree sitter grammar)"
        )),
    },
    [TerminalChatAgentToolsSettingId.ShellIntegrationTimeout]: {
        markdownDescription: ( localize(
            12128,
            "Configures the duration in milliseconds to wait for shell integration to be detected when the run in terminal tool launches a new terminal. Set to `0` to wait the minimum time, the default value `-1` means the wait time is variable based on the value of {0} and whether it's a remote window. A large value can be useful if your shell starts very slowly and a low value if you're intentionally not using shell integration.",
            `\`#${TerminalSettingId.ShellIntegrationEnabled}#\``
        )),
        type: 'integer',
        minimum: -1,
        maximum: 60000,
        default: -1,
        markdownDeprecationMessage: ( localize(
            12129,
            'Use {0} instead',
            `\`#${TerminalSettingId.ShellIntegrationTimeout}#\``
        ))
    },
    [TerminalChatAgentToolsSettingId.TerminalProfileLinux]: {
        restricted: true,
        markdownDescription: ( localize(
            12130,
            "The terminal profile to use on Linux for chat agent's run in terminal tool."
        )),
        type: ['object', 'null'],
        default: null,
        'anyOf': [
            { type: 'null' },
            terminalChatAgentProfileSchema
        ],
        defaultSnippets: [
            {
                body: {
                    path: '${1}'
                }
            }
        ]
    },
    [TerminalChatAgentToolsSettingId.TerminalProfileMacOs]: {
        restricted: true,
        markdownDescription: ( localize(
            12131,
            "The terminal profile to use on macOS for chat agent's run in terminal tool."
        )),
        type: ['object', 'null'],
        default: null,
        'anyOf': [
            { type: 'null' },
            terminalChatAgentProfileSchema
        ],
        defaultSnippets: [
            {
                body: {
                    path: '${1}'
                }
            }
        ]
    },
    [TerminalChatAgentToolsSettingId.TerminalProfileWindows]: {
        restricted: true,
        markdownDescription: ( localize(
            12132,
            "The terminal profile to use on Windows for chat agent's run in terminal tool."
        )),
        type: ['object', 'null'],
        default: null,
        'anyOf': [
            { type: 'null' },
            terminalChatAgentProfileSchema
        ],
        defaultSnippets: [
            {
                body: {
                    path: '${1}'
                }
            }
        ]
    },
    [TerminalChatAgentToolsSettingId.AutoReplyToPrompts]: {
        type: 'boolean',
        default: false,
        tags: ['experimental'],
        markdownDescription: ( localize(
            12133,
            "Whether to automatically respond to prompts in the terminal such as `Confirm? y/n`. This is an experimental feature and may not work in all scenarios."
        )),
    },
    [TerminalChatAgentToolsSettingId.OutputLocation]: {
        markdownDescription: ( localize(12134, "Where to show the output from the run in terminal tool session.")),
        type: 'string',
        enum: ['terminal', 'none'],
        enumDescriptions: [
            ( localize(12135, "Reveal the terminal when running the command.")),
            ( localize(12136, "Do not reveal the terminal automatically.")),
        ],
        default: product.quality !== 'stable' ? 'none' : 'terminal',
        tags: ['experimental'],
        experiment: {
            mode: 'auto'
        }
    },
    [TerminalChatAgentToolsSettingId.PreventShellHistory]: {
        type: 'boolean',
        default: true,
        markdownDescription: [
            ( localize(
            12137,
            "Whether to exclude commands run by the terminal tool from the shell history. See below for the supported shells and the method used for each:"
        )),
            `- \`bash\`: ${( localize(
                12138,
                "Sets `HISTCONTROL=ignorespace` and prepends the command with space"
            ))}`,
            `- \`zsh\`: ${( localize(
                12139,
                "Sets `HIST_IGNORE_SPACE` option and prepends the command with space"
            ))}`,
            `- \`fish\`: ${( localize(
                12140,
                "Sets `fish_private_mode` to prevent any command from entering history"
            ))}`,
            `- \`pwsh\`: ${( localize(
                12141,
                "Sets a custom history handler via PSReadLine's `AddToHistoryHandler` to prevent any command from entering history"
            ))}`,
        ].join('\n'),
    }
};
for (const id of [
    TerminalChatAgentToolsSettingId.DeprecatedAutoApprove1,
    TerminalChatAgentToolsSettingId.DeprecatedAutoApprove2,
    TerminalChatAgentToolsSettingId.DeprecatedAutoApprove3,
    TerminalChatAgentToolsSettingId.DeprecatedAutoApprove4,
    TerminalChatAgentToolsSettingId.DeprecatedAutoApproveCompatible,
]) {
    terminalChatAgentToolsConfiguration[id] = {
        deprecated: true,
        markdownDeprecationMessage: ( localize(
            12142,
            'Use {0} instead',
            `\`#${TerminalChatAgentToolsSettingId.AutoApprove}#\``
        ))
    };
}

export { TerminalChatAgentToolsSettingId, terminalChatAgentToolsConfiguration };
