// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminalContrib/autoReplies/common/terminalAutoRepliesConfiguration.js

import { localize } from '../../../../../nls.js';

var TerminalAutoRepliesSettingId;
(function (TerminalAutoRepliesSettingId) {
    TerminalAutoRepliesSettingId["AutoReplies"] = "terminal.integrated.autoReplies";
})(TerminalAutoRepliesSettingId || (TerminalAutoRepliesSettingId = {}));
const terminalAutoRepliesConfiguration = {
    [TerminalAutoRepliesSettingId.AutoReplies]: {
        markdownDescription: ( localize(
            11966,
            "A set of messages that, when encountered in the terminal, will be automatically responded to. Provided the message is specific enough, this can help automate away common responses.\n\nRemarks:\n\n- Use {0} to automatically respond to the terminate batch job prompt on Windows.\n- The message includes escape sequences so the reply might not happen with styled text.\n- Each reply can only happen once every second.\n- Use {1} in the reply to mean the enter key.\n- To unset a default key, set the value to null.\n- Restart VS Code if new don't apply.",
            '`"Terminate batch job (Y/N)": "Y\\r"`',
            '`"\\r"`'
        )),
        type: 'object',
        additionalProperties: {
            oneOf: [{
                    type: 'string',
                    description: ( localize(11967, "The reply to send to the process."))
                },
                { type: 'null' }]
        },
        default: {}
    },
};

export { TerminalAutoRepliesSettingId, terminalAutoRepliesConfiguration };
