// monaco-languageclient/lib/worker/index.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) 2025 TypeFox and others.
 * Licensed under the MIT License. See LICENSE in the package root for license information.
 * ------------------------------------------------------------------------------------------ */
import { getEnhancedMonacoEnvironment } from 'monaco-languageclient/vscodeApiWrapper';
export class Worker {
    url;
    options;
    constructor(url, options) {
        this.url = url;
        this.options = options;
    }
}
export const useWorkerFactory = (config) => {
    const envEnhanced = getEnhancedMonacoEnvironment();
    envEnhanced.getWorkerUrl = (workerId, label) => {
        config.logger?.info(`getWorkerUrl: workerId: ${workerId} label: ${label}`);
        return config.workerLoaders?.[label]?.().url.toString();
    };
    envEnhanced.getWorkerOptions = (moduleId, label) => {
        config.logger?.info(`getWorkerOptions: moduleId: ${moduleId} label: ${label}`);
        return config.workerLoaders?.[label]?.().options;
    };
};
export const defineDefaultWorkerLoaders = () => {
    const defaultEditorWorkerService = () => new Worker(new URL('@codingame/monaco-vscode-editor-api/esm/vs/editor/editor.worker.js', import.meta.url), { type: 'module' });
    const defaultExtensionHostWorkerMain = () => new Worker(new URL('@codingame/monaco-vscode-api/workers/extensionHost.worker', import.meta.url), { type: 'module' });
    const defaultTextMateWorker = () => new Worker(new URL('@codingame/monaco-vscode-textmate-service-override/worker', import.meta.url), { type: 'module' });
    return {
        // if you import monaco api as 'monaco-editor': monaco-editor/esm/vs/editor/editor.worker.js
        editorWorkerService: defaultEditorWorkerService,
        extensionHostWorkerMain: defaultExtensionHostWorkerMain,
        TextMateWorker: defaultTextMateWorker,
        // these are other possible workers not configured by default
        OutputLinkDetectionWorker: undefined,
        LanguageDetectionWorker: undefined,
        NotebookEditorWorker: undefined,
        LocalFileSearchWorker: undefined
    };
};
export const configureDefaultWorkerFactory = (logger) => {
    useWorkerFactory({
        workerLoaders: defineDefaultWorkerLoaders(),
        logger
    });
};
