// @codingame/monaco-vscode-editor-api/esm/vs/editor/editor.api.js

import { languages, editor } from '../../../vscode/src/vs/editor/editor.api.js';
export { CancellationTokenSource, Emitter, KeyCode, KeyMod, MarkerSeverity, MarkerTag, Position, Range, Selection, SelectionDirection, Token, Uri } from '../../../vscode/src/vs/editor/editor.api.js';
import { createConfiguredEditor, createConfiguredDiffEditor, createModelReference, writeFile } from '@codingame/monaco-vscode-api/monaco';
import { withReadyServices } from '@codingame/monaco-vscode-api/services';

const originalOnLanguage = languages.onLanguage;
languages.onLanguage = (languageId, callback) => withReadyServices(() => originalOnLanguage(languageId, callback));
const originalOnLanguageEncountered = languages.onLanguage;
languages.onLanguageEncountered = (languageId, callback) => withReadyServices(() => originalOnLanguageEncountered(languageId, callback));
editor.create = createConfiguredEditor;
editor.createDiffEditor = createConfiguredDiffEditor;
editor.createModelReference = createModelReference;
editor.writeFile = writeFile;

export { editor, languages };
